// @wormhole-foundation/sdk-definitions/dist/cjs/protocols/circleBridge/automaticCircleBridgeLayout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.automaticCircleBridgeNamedPayloads = exports.circleConnectPayload = exports.depositWithPayloadLayout = void 0;
const sdk_base_1 = require("@wormhole-foundation/sdk-base");
const index_js_1 = require("./../../layout-items/index.js");
const index_js_2 = require("./../../vaa/index.js");
//from here: https://github.com/wormhole-foundation/wormhole-circle-integration/blob/105ad59bad687416527003e0241dee4020889341/evm/src/circle_integration/CircleIntegrationMessages.sol#L25
const depositWithPayloadLayout = (customPayload) => [
    (0, index_js_1.payloadIdItem)(1),
    {
        name: "token",
        binary: "bytes",
        layout: [
            { name: "address", ...index_js_1.universalAddressItem },
            { name: "amount", ...index_js_1.amountItem },
        ],
    },
    { name: "sourceDomain", ...index_js_1.circleDomainItem },
    { name: "targetDomain", ...index_js_1.circleDomainItem },
    { name: "nonce", ...index_js_1.circleNonceItem },
    { name: "caller", ...index_js_1.universalAddressItem },
    { name: "mintRecipient", ...index_js_1.universalAddressItem },
    sdk_base_1.layout.customizableBytes({ name: "payload", lengthSize: 2 }, customPayload),
];
exports.depositWithPayloadLayout = depositWithPayloadLayout;
//from here:
//  https://github.com/wormhole-foundation/example-circle-relayer/blob/189becd8d3935decb17383bd2e61b4909cbddc89/evm/src/circle-relayer/CircleRelayerMessages.sol#L16
exports.circleConnectPayload = [
    (0, index_js_1.payloadIdItem)(1),
    { name: "targetRelayerFee", ...index_js_1.amountItem },
    { name: "toNativeTokenAmount", ...index_js_1.amountItem },
    { name: "targetRecipient", ...index_js_1.universalAddressItem },
];
exports.automaticCircleBridgeNamedPayloads = [
    ["DepositWithPayload", (0, exports.depositWithPayloadLayout)()],
    ["TransferWithRelay", (0, exports.depositWithPayloadLayout)(exports.circleConnectPayload)],
];
// factory registration:
require("../../registry.js");
(0, index_js_2.registerPayloadTypes)("AutomaticCircleBridge", exports.automaticCircleBridgeNamedPayloads);
