// @injectivelabs/sdk-ts/dist/cjs/client/indexer/grpc/IndexerGrpcArchiverApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexerGrpcArchiverApi = void 0;
const exceptions_1 = require("@injectivelabs/exceptions");
const indexer_proto_ts_1 = require("@injectivelabs/indexer-proto-ts");
const BaseIndexerGrpcConsumer_1 = __importDefault(require("../../base/BaseIndexerGrpcConsumer"));
const types_1 = require("../types");
const transformers_1 = require("../transformers");
/**
 * @category Indexer Grpc API
 */
class IndexerGrpcArchiverApi extends BaseIndexerGrpcConsumer_1.default {
    constructor(endpoint) {
        super(endpoint);
        this.module = types_1.IndexerModule.Archiver;
        this.client = new indexer_proto_ts_1.InjectiveArchiverRPC.InjectiveArchiverRPCClientImpl(this.getGrpcWebImpl(endpoint));
    }
    fetchHistoricalBalance(_a) {
        return __awaiter(this, arguments, void 0, function* ({ account, resolution, }) {
            const request = indexer_proto_ts_1.InjectiveArchiverRPC.BalanceRequest.create();
            request.account = account;
            request.resolution = resolution;
            try {
                const response = yield this.retry(() => this.client.Balance(request));
                return transformers_1.IndexerGrpcArchiverTransformer.grpcHistoricalBalanceResponseToHistoricalBalances(response);
            }
            catch (e) {
                if (e instanceof indexer_proto_ts_1.InjectiveArchiverRPC.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        context: 'Historical Balance',
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: 'Historical Balance',
                    contextModule: this.module,
                });
            }
        });
    }
    fetchHistoricalRpnl(_a) {
        return __awaiter(this, arguments, void 0, function* ({ account, resolution, }) {
            const request = indexer_proto_ts_1.InjectiveArchiverRPC.RpnlRequest.create();
            request.account = account;
            request.resolution = resolution;
            try {
                const response = yield this.retry(() => this.client.Rpnl(request));
                return transformers_1.IndexerGrpcArchiverTransformer.grpcHistoricalRPNLResponseToHistoricalRPNL(response);
            }
            catch (e) {
                if (e instanceof indexer_proto_ts_1.InjectiveArchiverRPC.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        context: 'Historical Rpnl',
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: 'Historical Rpnl',
                    contextModule: this.module,
                });
            }
        });
    }
    fetchHistoricalVolumes(_a) {
        return __awaiter(this, arguments, void 0, function* ({ account, resolution, }) {
            const request = indexer_proto_ts_1.InjectiveArchiverRPC.VolumesRequest.create();
            request.account = account;
            request.resolution = resolution;
            try {
                const response = yield this.retry(() => this.client.Volumes(request));
                return transformers_1.IndexerGrpcArchiverTransformer.grpcHistoricalVolumesResponseToHistoricalVolumes(response);
            }
            catch (e) {
                if (e instanceof indexer_proto_ts_1.InjectiveArchiverRPC.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        context: 'Historical Volumes',
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: 'Historical Volumes',
                    contextModule: this.module,
                });
            }
        });
    }
}
exports.IndexerGrpcArchiverApi = IndexerGrpcArchiverApi;
