// @wormhole-foundation/sdk-connect/dist/cjs/routes/resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouteResolver = void 0;
const sdk_definitions_1 = require("@wormhole-foundation/sdk-definitions");
const route_js_1 = require("./route.js");
const token_js_1 = require("./token.js");
class RouteResolver {
    wh;
    routeConstructors;
    inputTokenList;
    constructor(wh, routeConstructors) {
        this.wh = wh;
        this.routeConstructors = routeConstructors;
    }
    async supportedSourceTokens(chain) {
        if (this.inputTokenList)
            return this.inputTokenList;
        const itl = await Promise.all(this.routeConstructors.map(async (rc) => {
            try {
                return await rc.supportedSourceTokens(chain);
            }
            catch (e) {
                console.error(`Failed to get supported source tokens for ${rc.meta.name}: `, e);
                return [];
            }
        }));
        this.inputTokenList = (0, token_js_1.uniqueTokens)(itl.flat());
        return this.inputTokenList;
    }
    async supportedDestinationTokens(inputToken, fromChain, toChain) {
        const [, inputTokenId] = (0, sdk_definitions_1.resolveWrappedToken)(fromChain.network, fromChain.chain, inputToken);
        const tokens = await Promise.all(this.routeConstructors.map(async (rc) => {
            try {
                return await rc.supportedDestinationTokens(inputTokenId, fromChain, toChain);
            }
            catch (e) {
                return [];
            }
        }));
        return (0, token_js_1.uniqueTokens)(tokens.flat());
    }
    async findRoutes(request) {
        // First we find all routes which support the request inputs (network, chains, and tokens)
        const supportedRoutes = await Promise.all(this.routeConstructors.map(async (rc) => {
            try {
                const protocolSupported = rc.supportedNetworks().includes(this.wh.network) &&
                    rc.supportedChains(this.wh.network).includes(request.toChain.chain) &&
                    rc.supportedChains(this.wh.network).includes(request.fromChain.chain) &&
                    rc.isProtocolSupported(request.fromChain) &&
                    rc.isProtocolSupported(request.toChain);
                const sourceTokenAddress = (0, sdk_definitions_1.canonicalAddress)((0, sdk_definitions_1.isNative)(request.source.id.address) ? request.source.wrapped : request.source.id);
                const sourceTokenSupported = (await rc.supportedSourceTokens(request.fromChain)).filter((tokenId) => {
                    return (0, sdk_definitions_1.canonicalAddress)(tokenId) === sourceTokenAddress;
                }).length > 0;
                const dstTokenAddress = (0, sdk_definitions_1.canonicalAddress)((0, sdk_definitions_1.isNative)(request.destination.id.address)
                    ? request.destination.wrapped
                    : request.destination.id);
                const destinationTokenSupported = (await rc.supportedDestinationTokens(request.source.id, request.fromChain, request.toChain)).filter((tokenId) => {
                    return (0, sdk_definitions_1.canonicalAddress)(tokenId) === dstTokenAddress;
                }).length > 0;
                return protocolSupported && sourceTokenSupported && destinationTokenSupported;
            }
            catch (e) {
                return false;
            }
        })).then((routesSupported) => this.routeConstructors.filter((_, index) => routesSupported[index]));
        // Next, we make sure all supported routes are available. For relayed routes, this will ping
        // the relayer to make sure it's online.
        return await Promise.all(supportedRoutes.map(async (rc) => {
            const route = new rc(this.wh);
            try {
                const available = (0, route_js_1.isAutomatic)(route) ? await route.isAvailable(request) : true;
                return [route, available];
            }
            catch (e) {
                console.error(`failed to check if route is available for ${rc.meta.name}: `, e);
                return [route, false];
            }
        }))
            .then((availableRoutes) => availableRoutes.filter(([_, available]) => available))
            .then((availableRoutes) => availableRoutes.map(([route, _]) => route));
    }
}
exports.RouteResolver = RouteResolver;
