// @wormhole-foundation/sdk-base/dist/cjs/constants/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.guardians = exports.contracts = exports.circle = exports.nativeChainIds = exports.rpc = exports.explorer = exports.decimals = exports.finality = exports.network = exports.chain = exports.platform = void 0;
__exportStar(require("./networks.js"), exports);
__exportStar(require("./chains.js"), exports);
__exportStar(require("./platforms.js"), exports);
__exportStar(require("./tokens.js"), exports);
exports.platform = __importStar(require("./platforms.js"));
exports.chain = __importStar(require("./chains.js"));
exports.network = __importStar(require("./networks.js"));
exports.finality = __importStar(require("./finality.js"));
exports.decimals = __importStar(require("./decimals.js"));
exports.explorer = __importStar(require("./explorer.js"));
exports.rpc = __importStar(require("./rpc.js"));
exports.nativeChainIds = __importStar(require("./nativeChainIds.js"));
exports.circle = __importStar(require("./circle.js"));
exports.contracts = __importStar(require("./contracts/index.js"));
exports.guardians = __importStar(require("./guardians.js"));
