// @injectivelabs/core-proto-ts/cjs/injective/auction/v1beta1/genesis.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisState = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const auction_1 = require("./auction");
function createBaseGenesisState() {
    return { params: undefined, auctionRound: "0", highestBid: undefined, auctionEndingTimestamp: "0" };
}
exports.GenesisState = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.params !== undefined) {
            auction_1.Params.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        if (message.auctionRound !== "0") {
            writer.uint32(16).uint64(message.auctionRound);
        }
        if (message.highestBid !== undefined) {
            auction_1.Bid.encode(message.highestBid, writer.uint32(26).fork()).ldelim();
        }
        if (message.auctionEndingTimestamp !== "0") {
            writer.uint32(32).int64(message.auctionEndingTimestamp);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.params = auction_1.Params.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.auctionRound = longToString(reader.uint64());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.highestBid = auction_1.Bid.decode(reader, reader.uint32());
                    continue;
                case 4:
                    if (tag !== 32) {
                        break;
                    }
                    message.auctionEndingTimestamp = longToString(reader.int64());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            params: isSet(object.params) ? auction_1.Params.fromJSON(object.params) : undefined,
            auctionRound: isSet(object.auctionRound) ? globalThis.String(object.auctionRound) : "0",
            highestBid: isSet(object.highestBid) ? auction_1.Bid.fromJSON(object.highestBid) : undefined,
            auctionEndingTimestamp: isSet(object.auctionEndingTimestamp)
                ? globalThis.String(object.auctionEndingTimestamp)
                : "0",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.params !== undefined) {
            obj.params = auction_1.Params.toJSON(message.params);
        }
        if (message.auctionRound !== "0") {
            obj.auctionRound = message.auctionRound;
        }
        if (message.highestBid !== undefined) {
            obj.highestBid = auction_1.Bid.toJSON(message.highestBid);
        }
        if (message.auctionEndingTimestamp !== "0") {
            obj.auctionEndingTimestamp = message.auctionEndingTimestamp;
        }
        return obj;
    },
    create(base) {
        return exports.GenesisState.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseGenesisState();
        message.params = (object.params !== undefined && object.params !== null)
            ? auction_1.Params.fromPartial(object.params)
            : undefined;
        message.auctionRound = (_a = object.auctionRound) !== null && _a !== void 0 ? _a : "0";
        message.highestBid = (object.highestBid !== undefined && object.highestBid !== null)
            ? auction_1.Bid.fromPartial(object.highestBid)
            : undefined;
        message.auctionEndingTimestamp = (_b = object.auctionEndingTimestamp) !== null && _b !== void 0 ? _b : "0";
        return message;
    },
};
function longToString(long) {
    return long.toString();
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
