// @wormhole-foundation/sdk-connect/dist/cjs/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isFailed = exports.isCompleted = exports.isDestinationQueued = exports.isRedeemed = exports.isAttested = exports.isSourceFinalized = exports.isSourceInitiated = exports.TransferState = void 0;
// Transfer state machine states
var TransferState;
(function (TransferState) {
    TransferState[TransferState["Failed"] = -1] = "Failed";
    TransferState[TransferState["Created"] = 0] = "Created";
    TransferState[TransferState["SourceInitiated"] = 1] = "SourceInitiated";
    TransferState[TransferState["SourceFinalized"] = 2] = "SourceFinalized";
    TransferState[TransferState["Attested"] = 3] = "Attested";
    TransferState[TransferState["DestinationInitiated"] = 4] = "DestinationInitiated";
    TransferState[TransferState["DestinationQueued"] = 5] = "DestinationQueued";
    TransferState[TransferState["DestinationFinalized"] = 6] = "DestinationFinalized";
})(TransferState || (exports.TransferState = TransferState = {}));
function isSourceInitiated(receipt) {
    return receipt.state === TransferState.SourceInitiated;
}
exports.isSourceInitiated = isSourceInitiated;
function isSourceFinalized(receipt) {
    return receipt.state === TransferState.SourceFinalized;
}
exports.isSourceFinalized = isSourceFinalized;
function isAttested(receipt) {
    return receipt.state === TransferState.Attested;
}
exports.isAttested = isAttested;
function isRedeemed(receipt) {
    return receipt.state === TransferState.DestinationInitiated;
}
exports.isRedeemed = isRedeemed;
function isDestinationQueued(receipt) {
    return receipt.state === TransferState.DestinationQueued;
}
exports.isDestinationQueued = isDestinationQueued;
function isCompleted(receipt) {
    return receipt.state === TransferState.DestinationFinalized;
}
exports.isCompleted = isCompleted;
function isFailed(receipt) {
    return receipt.state < 0;
}
exports.isFailed = isFailed;
