// @injectivelabs/sdk-ts/dist/cjs/core/modules/ibc/msgs/MsgTransferCosmjs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// import { MsgTransferEncodeObject } from '@cosmjs/stargate'
const snakecase_keys_1 = __importDefault(require("snakecase-keys"));
const tx_1 = require("cosmjs-types/ibc/applications/transfer/v1/tx");
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
/**
 * @category Messages
 *
 * @deprecated use MsgTransfer with SIGN_DIRECT and a Cosmos wallet
 */
class MsgTransferCosmjs {
    constructor(params) {
        this.params = params;
    }
    static fromJSON(params) {
        return new MsgTransferCosmjs(params);
    }
    toProto() {
        const { params } = this;
        const token = core_proto_ts_1.CosmosBaseV1Beta1Coin.Coin.create();
        token.denom = params.amount.denom;
        token.amount = params.amount.amount;
        const message = core_proto_ts_1.IbcApplicationsTransferV1Tx.MsgTransfer.create();
        message.receiver = params.receiver;
        message.sender = params.sender;
        message.sourceChannel = params.channelId;
        message.sourcePort = params.port;
        message.token = token;
        if (params.height) {
            const timeoutHeight = core_proto_ts_1.IbcCoreClientV1Client.Height.create();
            timeoutHeight.revisionHeight = params.height.revisionHeight.toString();
            timeoutHeight.revisionNumber = params.height.revisionNumber.toString();
            message.timeoutHeight = timeoutHeight;
        }
        if (params.timeout) {
            message.timeoutTimestamp = params.timeout.toString();
        }
        return tx_1.MsgTransfer.fromJSON(message);
    }
    toData() {
        throw new Error('Method not implemented.');
    }
    toAmino() {
        const proto = this.toProto();
        const message = Object.assign({}, (0, snakecase_keys_1.default)(proto));
        return {
            type: 'cosmos-sdk/MsgTransfer',
            value: Object.assign({}, message),
        };
    }
    toWeb3() {
        const amino = this.toAmino();
        const { value } = amino;
        return Object.assign({ '@type': '/ibc.applications.transfer.v1.MsgTransfer' }, value);
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/ibc.applications.transfer.v1.MsgTransfer',
            message: proto,
        };
    }
    toBinary() {
        return tx_1.MsgTransfer.encode(this.toProto()).finish();
    }
}
exports.default = MsgTransferCosmjs;
