// @wormhole-foundation/sdk-solana-core/dist/cjs/utils/coder/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WormholeCoder = void 0;
const accounts_js_1 = require("./accounts.js");
const events_js_1 = require("./events.js");
const instruction_js_1 = require("./instruction.js");
const state_js_1 = require("./state.js");
const types_js_1 = require("./types.js");
__exportStar(require("./instruction.js"), exports);
class WormholeCoder {
    instruction;
    accounts;
    state;
    events;
    types;
    constructor(idl) {
        this.instruction = new instruction_js_1.WormholeInstructionCoder(idl);
        this.accounts = new accounts_js_1.WormholeAccountsCoder(idl);
        this.state = new state_js_1.WormholeStateCoder(idl);
        this.events = new events_js_1.WormholeEventsCoder(idl);
        this.types = new types_js_1.WormholeTypesCoder(idl);
    }
}
exports.WormholeCoder = WormholeCoder;
