// @injectivelabs/core-proto-ts/esm/injective/peggy/v1/params.js
/* eslint-disable */
import Long from "long";
import _m0 from "protobufjs/minimal";
import { Coin } from "../../../cosmos/base/v1beta1/coin";
function createBaseParams() {
    return {
        peggyId: "",
        contractSourceHash: "",
        bridgeEthereumAddress: "",
        bridgeChainId: "0",
        signedValsetsWindow: "0",
        signedBatchesWindow: "0",
        signedClaimsWindow: "0",
        targetBatchTimeout: "0",
        averageBlockTime: "0",
        averageEthereumBlockTime: "0",
        slashFractionValset: new Uint8Array(0),
        slashFractionBatch: new Uint8Array(0),
        slashFractionClaim: new Uint8Array(0),
        slashFractionConflictingClaim: new Uint8Array(0),
        unbondSlashingValsetsWindow: "0",
        slashFractionBadEthSignature: new Uint8Array(0),
        cosmosCoinDenom: "",
        cosmosCoinErc20Contract: "",
        claimSlashingEnabled: false,
        bridgeContractStartHeight: "0",
        valsetReward: undefined,
    };
}
export const Params = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.peggyId !== "") {
            writer.uint32(10).string(message.peggyId);
        }
        if (message.contractSourceHash !== "") {
            writer.uint32(18).string(message.contractSourceHash);
        }
        if (message.bridgeEthereumAddress !== "") {
            writer.uint32(26).string(message.bridgeEthereumAddress);
        }
        if (message.bridgeChainId !== "0") {
            writer.uint32(32).uint64(message.bridgeChainId);
        }
        if (message.signedValsetsWindow !== "0") {
            writer.uint32(40).uint64(message.signedValsetsWindow);
        }
        if (message.signedBatchesWindow !== "0") {
            writer.uint32(48).uint64(message.signedBatchesWindow);
        }
        if (message.signedClaimsWindow !== "0") {
            writer.uint32(56).uint64(message.signedClaimsWindow);
        }
        if (message.targetBatchTimeout !== "0") {
            writer.uint32(64).uint64(message.targetBatchTimeout);
        }
        if (message.averageBlockTime !== "0") {
            writer.uint32(72).uint64(message.averageBlockTime);
        }
        if (message.averageEthereumBlockTime !== "0") {
            writer.uint32(80).uint64(message.averageEthereumBlockTime);
        }
        if (message.slashFractionValset.length !== 0) {
            writer.uint32(90).bytes(message.slashFractionValset);
        }
        if (message.slashFractionBatch.length !== 0) {
            writer.uint32(98).bytes(message.slashFractionBatch);
        }
        if (message.slashFractionClaim.length !== 0) {
            writer.uint32(106).bytes(message.slashFractionClaim);
        }
        if (message.slashFractionConflictingClaim.length !== 0) {
            writer.uint32(114).bytes(message.slashFractionConflictingClaim);
        }
        if (message.unbondSlashingValsetsWindow !== "0") {
            writer.uint32(120).uint64(message.unbondSlashingValsetsWindow);
        }
        if (message.slashFractionBadEthSignature.length !== 0) {
            writer.uint32(130).bytes(message.slashFractionBadEthSignature);
        }
        if (message.cosmosCoinDenom !== "") {
            writer.uint32(138).string(message.cosmosCoinDenom);
        }
        if (message.cosmosCoinErc20Contract !== "") {
            writer.uint32(146).string(message.cosmosCoinErc20Contract);
        }
        if (message.claimSlashingEnabled === true) {
            writer.uint32(152).bool(message.claimSlashingEnabled);
        }
        if (message.bridgeContractStartHeight !== "0") {
            writer.uint32(160).uint64(message.bridgeContractStartHeight);
        }
        if (message.valsetReward !== undefined) {
            Coin.encode(message.valsetReward, writer.uint32(170).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.peggyId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.contractSourceHash = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.bridgeEthereumAddress = reader.string();
                    continue;
                case 4:
                    if (tag !== 32) {
                        break;
                    }
                    message.bridgeChainId = longToString(reader.uint64());
                    continue;
                case 5:
                    if (tag !== 40) {
                        break;
                    }
                    message.signedValsetsWindow = longToString(reader.uint64());
                    continue;
                case 6:
                    if (tag !== 48) {
                        break;
                    }
                    message.signedBatchesWindow = longToString(reader.uint64());
                    continue;
                case 7:
                    if (tag !== 56) {
                        break;
                    }
                    message.signedClaimsWindow = longToString(reader.uint64());
                    continue;
                case 8:
                    if (tag !== 64) {
                        break;
                    }
                    message.targetBatchTimeout = longToString(reader.uint64());
                    continue;
                case 9:
                    if (tag !== 72) {
                        break;
                    }
                    message.averageBlockTime = longToString(reader.uint64());
                    continue;
                case 10:
                    if (tag !== 80) {
                        break;
                    }
                    message.averageEthereumBlockTime = longToString(reader.uint64());
                    continue;
                case 11:
                    if (tag !== 90) {
                        break;
                    }
                    message.slashFractionValset = reader.bytes();
                    continue;
                case 12:
                    if (tag !== 98) {
                        break;
                    }
                    message.slashFractionBatch = reader.bytes();
                    continue;
                case 13:
                    if (tag !== 106) {
                        break;
                    }
                    message.slashFractionClaim = reader.bytes();
                    continue;
                case 14:
                    if (tag !== 114) {
                        break;
                    }
                    message.slashFractionConflictingClaim = reader.bytes();
                    continue;
                case 15:
                    if (tag !== 120) {
                        break;
                    }
                    message.unbondSlashingValsetsWindow = longToString(reader.uint64());
                    continue;
                case 16:
                    if (tag !== 130) {
                        break;
                    }
                    message.slashFractionBadEthSignature = reader.bytes();
                    continue;
                case 17:
                    if (tag !== 138) {
                        break;
                    }
                    message.cosmosCoinDenom = reader.string();
                    continue;
                case 18:
                    if (tag !== 146) {
                        break;
                    }
                    message.cosmosCoinErc20Contract = reader.string();
                    continue;
                case 19:
                    if (tag !== 152) {
                        break;
                    }
                    message.claimSlashingEnabled = reader.bool();
                    continue;
                case 20:
                    if (tag !== 160) {
                        break;
                    }
                    message.bridgeContractStartHeight = longToString(reader.uint64());
                    continue;
                case 21:
                    if (tag !== 170) {
                        break;
                    }
                    message.valsetReward = Coin.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            peggyId: isSet(object.peggyId) ? globalThis.String(object.peggyId) : "",
            contractSourceHash: isSet(object.contractSourceHash) ? globalThis.String(object.contractSourceHash) : "",
            bridgeEthereumAddress: isSet(object.bridgeEthereumAddress) ? globalThis.String(object.bridgeEthereumAddress) : "",
            bridgeChainId: isSet(object.bridgeChainId) ? globalThis.String(object.bridgeChainId) : "0",
            signedValsetsWindow: isSet(object.signedValsetsWindow) ? globalThis.String(object.signedValsetsWindow) : "0",
            signedBatchesWindow: isSet(object.signedBatchesWindow) ? globalThis.String(object.signedBatchesWindow) : "0",
            signedClaimsWindow: isSet(object.signedClaimsWindow) ? globalThis.String(object.signedClaimsWindow) : "0",
            targetBatchTimeout: isSet(object.targetBatchTimeout) ? globalThis.String(object.targetBatchTimeout) : "0",
            averageBlockTime: isSet(object.averageBlockTime) ? globalThis.String(object.averageBlockTime) : "0",
            averageEthereumBlockTime: isSet(object.averageEthereumBlockTime)
                ? globalThis.String(object.averageEthereumBlockTime)
                : "0",
            slashFractionValset: isSet(object.slashFractionValset)
                ? bytesFromBase64(object.slashFractionValset)
                : new Uint8Array(0),
            slashFractionBatch: isSet(object.slashFractionBatch)
                ? bytesFromBase64(object.slashFractionBatch)
                : new Uint8Array(0),
            slashFractionClaim: isSet(object.slashFractionClaim)
                ? bytesFromBase64(object.slashFractionClaim)
                : new Uint8Array(0),
            slashFractionConflictingClaim: isSet(object.slashFractionConflictingClaim)
                ? bytesFromBase64(object.slashFractionConflictingClaim)
                : new Uint8Array(0),
            unbondSlashingValsetsWindow: isSet(object.unbondSlashingValsetsWindow)
                ? globalThis.String(object.unbondSlashingValsetsWindow)
                : "0",
            slashFractionBadEthSignature: isSet(object.slashFractionBadEthSignature)
                ? bytesFromBase64(object.slashFractionBadEthSignature)
                : new Uint8Array(0),
            cosmosCoinDenom: isSet(object.cosmosCoinDenom) ? globalThis.String(object.cosmosCoinDenom) : "",
            cosmosCoinErc20Contract: isSet(object.cosmosCoinErc20Contract)
                ? globalThis.String(object.cosmosCoinErc20Contract)
                : "",
            claimSlashingEnabled: isSet(object.claimSlashingEnabled)
                ? globalThis.Boolean(object.claimSlashingEnabled)
                : false,
            bridgeContractStartHeight: isSet(object.bridgeContractStartHeight)
                ? globalThis.String(object.bridgeContractStartHeight)
                : "0",
            valsetReward: isSet(object.valsetReward) ? Coin.fromJSON(object.valsetReward) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.peggyId !== "") {
            obj.peggyId = message.peggyId;
        }
        if (message.contractSourceHash !== "") {
            obj.contractSourceHash = message.contractSourceHash;
        }
        if (message.bridgeEthereumAddress !== "") {
            obj.bridgeEthereumAddress = message.bridgeEthereumAddress;
        }
        if (message.bridgeChainId !== "0") {
            obj.bridgeChainId = message.bridgeChainId;
        }
        if (message.signedValsetsWindow !== "0") {
            obj.signedValsetsWindow = message.signedValsetsWindow;
        }
        if (message.signedBatchesWindow !== "0") {
            obj.signedBatchesWindow = message.signedBatchesWindow;
        }
        if (message.signedClaimsWindow !== "0") {
            obj.signedClaimsWindow = message.signedClaimsWindow;
        }
        if (message.targetBatchTimeout !== "0") {
            obj.targetBatchTimeout = message.targetBatchTimeout;
        }
        if (message.averageBlockTime !== "0") {
            obj.averageBlockTime = message.averageBlockTime;
        }
        if (message.averageEthereumBlockTime !== "0") {
            obj.averageEthereumBlockTime = message.averageEthereumBlockTime;
        }
        if (message.slashFractionValset.length !== 0) {
            obj.slashFractionValset = base64FromBytes(message.slashFractionValset);
        }
        if (message.slashFractionBatch.length !== 0) {
            obj.slashFractionBatch = base64FromBytes(message.slashFractionBatch);
        }
        if (message.slashFractionClaim.length !== 0) {
            obj.slashFractionClaim = base64FromBytes(message.slashFractionClaim);
        }
        if (message.slashFractionConflictingClaim.length !== 0) {
            obj.slashFractionConflictingClaim = base64FromBytes(message.slashFractionConflictingClaim);
        }
        if (message.unbondSlashingValsetsWindow !== "0") {
            obj.unbondSlashingValsetsWindow = message.unbondSlashingValsetsWindow;
        }
        if (message.slashFractionBadEthSignature.length !== 0) {
            obj.slashFractionBadEthSignature = base64FromBytes(message.slashFractionBadEthSignature);
        }
        if (message.cosmosCoinDenom !== "") {
            obj.cosmosCoinDenom = message.cosmosCoinDenom;
        }
        if (message.cosmosCoinErc20Contract !== "") {
            obj.cosmosCoinErc20Contract = message.cosmosCoinErc20Contract;
        }
        if (message.claimSlashingEnabled === true) {
            obj.claimSlashingEnabled = message.claimSlashingEnabled;
        }
        if (message.bridgeContractStartHeight !== "0") {
            obj.bridgeContractStartHeight = message.bridgeContractStartHeight;
        }
        if (message.valsetReward !== undefined) {
            obj.valsetReward = Coin.toJSON(message.valsetReward);
        }
        return obj;
    },
    create(base) {
        return Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
        const message = createBaseParams();
        message.peggyId = (_a = object.peggyId) !== null && _a !== void 0 ? _a : "";
        message.contractSourceHash = (_b = object.contractSourceHash) !== null && _b !== void 0 ? _b : "";
        message.bridgeEthereumAddress = (_c = object.bridgeEthereumAddress) !== null && _c !== void 0 ? _c : "";
        message.bridgeChainId = (_d = object.bridgeChainId) !== null && _d !== void 0 ? _d : "0";
        message.signedValsetsWindow = (_e = object.signedValsetsWindow) !== null && _e !== void 0 ? _e : "0";
        message.signedBatchesWindow = (_f = object.signedBatchesWindow) !== null && _f !== void 0 ? _f : "0";
        message.signedClaimsWindow = (_g = object.signedClaimsWindow) !== null && _g !== void 0 ? _g : "0";
        message.targetBatchTimeout = (_h = object.targetBatchTimeout) !== null && _h !== void 0 ? _h : "0";
        message.averageBlockTime = (_j = object.averageBlockTime) !== null && _j !== void 0 ? _j : "0";
        message.averageEthereumBlockTime = (_k = object.averageEthereumBlockTime) !== null && _k !== void 0 ? _k : "0";
        message.slashFractionValset = (_l = object.slashFractionValset) !== null && _l !== void 0 ? _l : new Uint8Array(0);
        message.slashFractionBatch = (_m = object.slashFractionBatch) !== null && _m !== void 0 ? _m : new Uint8Array(0);
        message.slashFractionClaim = (_o = object.slashFractionClaim) !== null && _o !== void 0 ? _o : new Uint8Array(0);
        message.slashFractionConflictingClaim = (_p = object.slashFractionConflictingClaim) !== null && _p !== void 0 ? _p : new Uint8Array(0);
        message.unbondSlashingValsetsWindow = (_q = object.unbondSlashingValsetsWindow) !== null && _q !== void 0 ? _q : "0";
        message.slashFractionBadEthSignature = (_r = object.slashFractionBadEthSignature) !== null && _r !== void 0 ? _r : new Uint8Array(0);
        message.cosmosCoinDenom = (_s = object.cosmosCoinDenom) !== null && _s !== void 0 ? _s : "";
        message.cosmosCoinErc20Contract = (_t = object.cosmosCoinErc20Contract) !== null && _t !== void 0 ? _t : "";
        message.claimSlashingEnabled = (_u = object.claimSlashingEnabled) !== null && _u !== void 0 ? _u : false;
        message.bridgeContractStartHeight = (_v = object.bridgeContractStartHeight) !== null && _v !== void 0 ? _v : "0";
        message.valsetReward = (object.valsetReward !== undefined && object.valsetReward !== null)
            ? Coin.fromPartial(object.valsetReward)
            : undefined;
        return message;
    },
};
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function longToString(long) {
    return long.toString();
}
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
