// @wormhole-foundation/sdk-base/dist/cjs/constants/contracts/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.porticoContractChains = exports.circleContractChains = exports.tokenBridgeRelayerChains = exports.tokenBridgeChains = exports.rollupContracts = exports.circleContracts = exports.portico = exports.translator = exports.gateway = exports.relayer = exports.nftBridge = exports.tokenBridgeRelayer = exports.tokenBridge = exports.coreBridge = void 0;
const core = __importStar(require("./core.js"));
const tb = __importStar(require("./tokenBridge.js"));
const tbr = __importStar(require("./tokenBridgeRelayer.js"));
const nb = __importStar(require("./nftBridge.js"));
const r = __importStar(require("./relayer.js"));
const circle = __importStar(require("./circle.js"));
const g = __importStar(require("./cosmos.js"));
const rollup = __importStar(require("./rollupCheckpoint.js"));
const p = __importStar(require("./portico.js"));
const index_js_1 = require("./../../utils/index.js");
exports.coreBridge = (0, index_js_1.constMap)(core.coreBridgeContracts);
exports.tokenBridge = (0, index_js_1.constMap)(tb.tokenBridgeContracts);
exports.tokenBridgeRelayer = (0, index_js_1.constMap)(tbr.tokenBridgeRelayerContracts);
exports.nftBridge = (0, index_js_1.constMap)(nb.nftBridgeContracts);
exports.relayer = (0, index_js_1.constMap)(r.relayerContracts);
exports.gateway = (0, index_js_1.constMap)(g.gatewayContracts);
exports.translator = (0, index_js_1.constMap)(g.translatorContracts);
exports.portico = (0, index_js_1.constMap)(p.porticoContracts);
exports.circleContracts = (0, index_js_1.constMap)(circle.circleContracts);
exports.rollupContracts = (0, index_js_1.constMap)(rollup.rollupContractAddresses);
exports.tokenBridgeChains = (0, index_js_1.constMap)(tb.tokenBridgeContracts, [0, 1]);
exports.tokenBridgeRelayerChains = (0, index_js_1.constMap)(tbr.tokenBridgeRelayerContracts, [0, 1]);
exports.circleContractChains = (0, index_js_1.constMap)(circle.circleContracts, [0, 1]);
exports.porticoContractChains = (0, index_js_1.constMap)(p.porticoContracts, [0, 1]);
