// @wormhole-foundation/sdk-solana-core/dist/cjs/utils/cpi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPostMessageCpiAccounts = exports.getWormholeDerivedAccounts = void 0;
const index_js_1 = require("./accounts/index.js");
const index_js_2 = require("./instructions/index.js");
/**
 * Generate Wormhole PDAs.
 *
 * @param cpiProgramId
 * @param wormholeProgramId
 * @returns
 */
function getWormholeDerivedAccounts(cpiProgramId, wormholeProgramId) {
    const { emitter: wormholeEmitter, sequence: wormholeSequence } = (0, index_js_1.getEmitterKeys)(cpiProgramId, wormholeProgramId);
    return {
        wormholeBridge: (0, index_js_1.deriveWormholeBridgeDataKey)(wormholeProgramId),
        wormholeEmitter,
        wormholeSequence,
        wormholeFeeCollector: (0, index_js_1.deriveFeeCollectorKey)(wormholeProgramId),
    };
}
exports.getWormholeDerivedAccounts = getWormholeDerivedAccounts;
/**
 * Generate accounts needed to perform `post_message` instruction
 * as cross-program invocation.
 *
 * @param cpiProgramId
 * @param wormholeProgramId
 * @param payer
 * @param message
 * @returns
 */
function getPostMessageCpiAccounts(cpiProgramId, wormholeProgramId, payer, message) {
    const accounts = (0, index_js_2.getPostMessageAccounts)(wormholeProgramId, payer, message, cpiProgramId);
    return {
        payer: accounts.payer,
        wormholeBridge: accounts.bridge,
        wormholeMessage: accounts.message,
        wormholeEmitter: accounts.emitter,
        wormholeSequence: accounts.sequence,
        wormholeFeeCollector: accounts.feeCollector,
        clock: accounts.clock,
        rent: accounts.rent,
        systemProgram: accounts.systemProgram,
    };
}
exports.getPostMessageCpiAccounts = getPostMessageCpiAccounts;
