// @wormhole-foundation/sdk-base/dist/esm/utils/layout/fixedDynamic.js
import { isPrimitiveType, isLayoutItem, isFixedPrimitiveConversion } from './utils.js';
function filterItem(item, fixed) {
    switch (item.binary) {
        // @ts-ignore - fallthrough is intentional
        case "bytes": {
            if ("layout" in item) {
                const { custom } = item;
                if (custom === undefined) {
                    const { layout } = item;
                    if (isLayoutItem(layout))
                        return filterItem(layout, fixed);
                    const filteredItems = internalFilterItemsOfProperLayout(layout, fixed);
                    return (filteredItems.length > 0) ? { ...item, layout: filteredItems } : null;
                }
                const isFixedItem = typeof custom.from !== "function";
                return (fixed && isFixedItem || !fixed && !isFixedItem) ? item : null;
            }
        }
        case "int":
        case "uint": {
            const { custom } = item;
            const isFixedItem = isPrimitiveType(custom) || isFixedPrimitiveConversion(custom);
            return (fixed && isFixedItem || !fixed && !isFixedItem) ? item : null;
        }
        case "array": {
            const filtered = internalFilterItemsOfLayout(item.layout, fixed);
            return (filtered !== null) ? { ...item, layout: filtered } : null;
        }
        case "switch": {
            const filteredIdLayoutPairs = item.layouts.reduce((acc, [idOrConversionId, idLayout]) => {
                const filteredItems = internalFilterItemsOfProperLayout(idLayout, fixed);
                return filteredItems.length > 0
                    ? [...acc, [idOrConversionId, filteredItems]]
                    : acc;
            }, []);
            return { ...item, layouts: filteredIdLayoutPairs };
        }
    }
}
function internalFilterItemsOfProperLayout(proper, fixed) {
    return proper.reduce((acc, item) => {
        const filtered = filterItem(item, fixed);
        return filtered !== null ? [...acc, filtered] : acc;
    }, []);
}
function internalFilterItemsOfLayout(layout, fixed) {
    return (Array.isArray(layout)
        ? internalFilterItemsOfProperLayout(layout, fixed)
        : filterItem(layout, fixed));
}
function filterItemsOfLayout(layout, fixed) {
    return internalFilterItemsOfLayout(layout, fixed);
}
export const fixedItemsOfLayout = (layout) => filterItemsOfLayout(layout, true);
export const dynamicItemsOfLayout = (layout) => filterItemsOfLayout(layout, false);
function internalAddFixedValuesItem(item, dynamicValue) {
    switch (item.binary) {
        // @ts-ignore - fallthrough is intentional
        case "bytes": {
            if ("layout" in item) {
                const { custom } = item;
                if (custom === undefined || typeof custom.from !== "function")
                    return internalAddFixedValues(item.layout, custom ? custom.from : dynamicValue);
                return dynamicValue;
            }
        }
        case "int":
        case "uint": {
            const { custom } = item;
            return item?.omit
                ? undefined
                : isPrimitiveType(custom)
                    ? custom
                    : isFixedPrimitiveConversion(custom)
                        ? custom.to
                        : dynamicValue;
        }
        case "array":
            return Array.isArray(dynamicValue)
                ? dynamicValue.map(element => internalAddFixedValues(item.layout, element))
                : undefined;
        case "switch": {
            const id = dynamicValue[item.idTag ?? "id"];
            const [_, idLayout] = item.layouts.find(([idOrConversionId]) => (Array.isArray(idOrConversionId) ? idOrConversionId[1] : idOrConversionId) == id);
            return {
                [item.idTag ?? "id"]: id,
                ...internalAddFixedValues(idLayout, dynamicValue)
            };
        }
    }
}
function internalAddFixedValues(layout, dynamicValues) {
    dynamicValues = dynamicValues ?? {};
    if (isLayoutItem(layout))
        return internalAddFixedValuesItem(layout, dynamicValues);
    const ret = {};
    for (const item of layout) {
        const fixedVals = internalAddFixedValuesItem(item, dynamicValues[item.name] ?? {});
        if (fixedVals !== undefined)
            ret[item.name] = fixedVals;
    }
    return ret;
}
export function addFixedValues(layout, dynamicValues) {
    return internalAddFixedValues(layout, dynamicValues);
}
