// @injectivelabs/sdk-ts/dist/esm/core/modules/wasm/exec-args/ExecArgCreatePerpGridStrategy.js
import { dataToExecData, ExecArgBase, } from '../ExecArgBase';
import { ExitType } from '../types';
/**
 * @category Contract Exec Arguments
 */
export default class ExecArgCreatePerpGridStrategy extends ExecArgBase {
    static fromJSON(params) {
        return new ExecArgCreatePerpGridStrategy(params);
    }
    toData() {
        const { params } = this;
        return {
            subaccount_id: params.subaccountId,
            bounds: [params.lowerBound, params.upperBound],
            levels: params.levels,
            slippage: params.slippage,
            strategy_type: {
                perpetual: {
                    margin_ratio: params.marginRatio,
                },
            },
            stop_loss: params.stopLoss
                ? {
                    exit_type: ExitType.Default,
                    exit_price: params.stopLoss,
                }
                : undefined,
            take_profit: params.takeProfit
                ? {
                    exit_type: ExitType.Default,
                    exit_price: params.takeProfit,
                }
                : undefined,
        };
    }
    toExecData() {
        return dataToExecData('create_strategy', this.toData());
    }
}
