// @wormhole-foundation/sdk/dist/cjs/platforms/evm.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_connect_1 = require("@wormhole-foundation/sdk-connect");
const _evm = __importStar(require("@wormhole-foundation/sdk-evm"));
/** Platform and protocol definitions for Evm */
const evm = {
    Address: _evm.EvmAddress,
    Platform: _evm.EvmPlatform,
    getSigner: _evm.getEvmSigner,
    protocols: {
        WormholeCore: () => Promise.resolve().then(() => __importStar(require("@wormhole-foundation/sdk-evm-core"))),
        TokenBridge: () => Promise.resolve().then(() => __importStar(require("@wormhole-foundation/sdk-evm-tokenbridge"))),
        PorticoBridge: () => Promise.resolve().then(() => __importStar(require("@wormhole-foundation/sdk-evm-portico"))),
        CircleBridge: () => Promise.resolve().then(() => __importStar(require("@wormhole-foundation/sdk-evm-cctp"))),
    },
    getChain: (network, chain, overrides) => new _evm.EvmChain(chain, new _evm.EvmPlatform(network, (0, sdk_connect_1.applyChainsConfigConfigOverrides)(network, _evm._platform, {
        [chain]: overrides,
    }))),
};
exports.default = evm;
