// @injectivelabs/sdk-ts/dist/cjs/client/indexer/transformers/IndexerGrpcArchiverTransformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexerGrpcArchiverTransformer = void 0;
/**
 * @category Indexer Grpc Transformer
 */
class IndexerGrpcArchiverTransformer {
    static grpcHistoricalBalanceToHistoricalBalance(historicalBalance) {
        return {
            t: historicalBalance.t,
            v: historicalBalance.v,
        };
    }
    static grpcHistoricalRPNLToHistoricalRPNL(historicalRPNL) {
        return {
            t: historicalRPNL.t,
            v: historicalRPNL.v,
        };
    }
    static grpcHistoricalVolumesToHistoricalVolumes(historicalVolumes) {
        return {
            t: historicalVolumes.t,
            v: historicalVolumes.v,
        };
    }
    static grpcHistoricalBalanceResponseToHistoricalBalances(response) {
        if (!response.historicalBalance) {
            return { t: [], v: [] };
        }
        return IndexerGrpcArchiverTransformer.grpcHistoricalBalanceToHistoricalBalance(response.historicalBalance);
    }
    static grpcHistoricalRPNLResponseToHistoricalRPNL(response) {
        if (!response.historicalRpnl) {
            return { t: [], v: [] };
        }
        return IndexerGrpcArchiverTransformer.grpcHistoricalRPNLToHistoricalRPNL(response.historicalRpnl);
    }
    static grpcHistoricalVolumesResponseToHistoricalVolumes(response) {
        if (!response.historicalVolumes) {
            return { t: [], v: [] };
        }
        return IndexerGrpcArchiverTransformer.grpcHistoricalVolumesToHistoricalVolumes(response.historicalVolumes);
    }
}
exports.IndexerGrpcArchiverTransformer = IndexerGrpcArchiverTransformer;
