// @injectivelabs/core-proto-ts/esm/injective/wasmx/v1/proposal.js
/* eslint-disable */
import Long from "long";
import _m0 from "protobufjs/minimal";
import { StoreCodeProposal } from "../../../cosmwasm/wasm/v1/proposal_legacy";
export var FundingMode;
(function (FundingMode) {
    FundingMode[FundingMode["Unspecified"] = 0] = "Unspecified";
    FundingMode[FundingMode["SelfFunded"] = 1] = "SelfFunded";
    FundingMode[FundingMode["GrantOnly"] = 2] = "GrantOnly";
    FundingMode[FundingMode["Dual"] = 3] = "Dual";
    FundingMode[FundingMode["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(FundingMode || (FundingMode = {}));
export function fundingModeFromJSON(object) {
    switch (object) {
        case 0:
        case "Unspecified":
            return FundingMode.Unspecified;
        case 1:
        case "SelfFunded":
            return FundingMode.SelfFunded;
        case 2:
        case "GrantOnly":
            return FundingMode.GrantOnly;
        case 3:
        case "Dual":
            return FundingMode.Dual;
        case -1:
        case "UNRECOGNIZED":
        default:
            return FundingMode.UNRECOGNIZED;
    }
}
export function fundingModeToJSON(object) {
    switch (object) {
        case FundingMode.Unspecified:
            return "Unspecified";
        case FundingMode.SelfFunded:
            return "SelfFunded";
        case FundingMode.GrantOnly:
            return "GrantOnly";
        case FundingMode.Dual:
            return "Dual";
        case FundingMode.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBaseContractRegistrationRequestProposal() {
    return { title: "", description: "", contractRegistrationRequest: undefined };
}
export const ContractRegistrationRequestProposal = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.title !== "") {
            writer.uint32(10).string(message.title);
        }
        if (message.description !== "") {
            writer.uint32(18).string(message.description);
        }
        if (message.contractRegistrationRequest !== undefined) {
            ContractRegistrationRequest.encode(message.contractRegistrationRequest, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseContractRegistrationRequestProposal();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.title = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.description = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.contractRegistrationRequest = ContractRegistrationRequest.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            title: isSet(object.title) ? globalThis.String(object.title) : "",
            description: isSet(object.description) ? globalThis.String(object.description) : "",
            contractRegistrationRequest: isSet(object.contractRegistrationRequest)
                ? ContractRegistrationRequest.fromJSON(object.contractRegistrationRequest)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.title !== "") {
            obj.title = message.title;
        }
        if (message.description !== "") {
            obj.description = message.description;
        }
        if (message.contractRegistrationRequest !== undefined) {
            obj.contractRegistrationRequest = ContractRegistrationRequest.toJSON(message.contractRegistrationRequest);
        }
        return obj;
    },
    create(base) {
        return ContractRegistrationRequestProposal.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseContractRegistrationRequestProposal();
        message.title = (_a = object.title) !== null && _a !== void 0 ? _a : "";
        message.description = (_b = object.description) !== null && _b !== void 0 ? _b : "";
        message.contractRegistrationRequest =
            (object.contractRegistrationRequest !== undefined && object.contractRegistrationRequest !== null)
                ? ContractRegistrationRequest.fromPartial(object.contractRegistrationRequest)
                : undefined;
        return message;
    },
};
function createBaseBatchContractRegistrationRequestProposal() {
    return { title: "", description: "", contractRegistrationRequests: [] };
}
export const BatchContractRegistrationRequestProposal = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.title !== "") {
            writer.uint32(10).string(message.title);
        }
        if (message.description !== "") {
            writer.uint32(18).string(message.description);
        }
        for (const v of message.contractRegistrationRequests) {
            ContractRegistrationRequest.encode(v, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseBatchContractRegistrationRequestProposal();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.title = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.description = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.contractRegistrationRequests.push(ContractRegistrationRequest.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            title: isSet(object.title) ? globalThis.String(object.title) : "",
            description: isSet(object.description) ? globalThis.String(object.description) : "",
            contractRegistrationRequests: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.contractRegistrationRequests)
                ? object.contractRegistrationRequests.map((e) => ContractRegistrationRequest.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.title !== "") {
            obj.title = message.title;
        }
        if (message.description !== "") {
            obj.description = message.description;
        }
        if ((_a = message.contractRegistrationRequests) === null || _a === void 0 ? void 0 : _a.length) {
            obj.contractRegistrationRequests = message.contractRegistrationRequests.map((e) => ContractRegistrationRequest.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return BatchContractRegistrationRequestProposal.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseBatchContractRegistrationRequestProposal();
        message.title = (_a = object.title) !== null && _a !== void 0 ? _a : "";
        message.description = (_b = object.description) !== null && _b !== void 0 ? _b : "";
        message.contractRegistrationRequests =
            ((_c = object.contractRegistrationRequests) === null || _c === void 0 ? void 0 : _c.map((e) => ContractRegistrationRequest.fromPartial(e))) || [];
        return message;
    },
};
function createBaseBatchContractDeregistrationProposal() {
    return { title: "", description: "", contracts: [] };
}
export const BatchContractDeregistrationProposal = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.title !== "") {
            writer.uint32(10).string(message.title);
        }
        if (message.description !== "") {
            writer.uint32(18).string(message.description);
        }
        for (const v of message.contracts) {
            writer.uint32(26).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseBatchContractDeregistrationProposal();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.title = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.description = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.contracts.push(reader.string());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            title: isSet(object.title) ? globalThis.String(object.title) : "",
            description: isSet(object.description) ? globalThis.String(object.description) : "",
            contracts: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.contracts)
                ? object.contracts.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.title !== "") {
            obj.title = message.title;
        }
        if (message.description !== "") {
            obj.description = message.description;
        }
        if ((_a = message.contracts) === null || _a === void 0 ? void 0 : _a.length) {
            obj.contracts = message.contracts;
        }
        return obj;
    },
    create(base) {
        return BatchContractDeregistrationProposal.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseBatchContractDeregistrationProposal();
        message.title = (_a = object.title) !== null && _a !== void 0 ? _a : "";
        message.description = (_b = object.description) !== null && _b !== void 0 ? _b : "";
        message.contracts = ((_c = object.contracts) === null || _c === void 0 ? void 0 : _c.map((e) => e)) || [];
        return message;
    },
};
function createBaseContractRegistrationRequest() {
    return {
        contractAddress: "",
        gasLimit: "0",
        gasPrice: "0",
        shouldPinContract: false,
        isMigrationAllowed: false,
        codeId: "0",
        adminAddress: "",
        granterAddress: "",
        fundingMode: 0,
    };
}
export const ContractRegistrationRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.contractAddress !== "") {
            writer.uint32(10).string(message.contractAddress);
        }
        if (message.gasLimit !== "0") {
            writer.uint32(16).uint64(message.gasLimit);
        }
        if (message.gasPrice !== "0") {
            writer.uint32(24).uint64(message.gasPrice);
        }
        if (message.shouldPinContract === true) {
            writer.uint32(32).bool(message.shouldPinContract);
        }
        if (message.isMigrationAllowed === true) {
            writer.uint32(40).bool(message.isMigrationAllowed);
        }
        if (message.codeId !== "0") {
            writer.uint32(48).uint64(message.codeId);
        }
        if (message.adminAddress !== "") {
            writer.uint32(58).string(message.adminAddress);
        }
        if (message.granterAddress !== "") {
            writer.uint32(66).string(message.granterAddress);
        }
        if (message.fundingMode !== 0) {
            writer.uint32(72).int32(message.fundingMode);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseContractRegistrationRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.contractAddress = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.gasLimit = longToString(reader.uint64());
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.gasPrice = longToString(reader.uint64());
                    continue;
                case 4:
                    if (tag !== 32) {
                        break;
                    }
                    message.shouldPinContract = reader.bool();
                    continue;
                case 5:
                    if (tag !== 40) {
                        break;
                    }
                    message.isMigrationAllowed = reader.bool();
                    continue;
                case 6:
                    if (tag !== 48) {
                        break;
                    }
                    message.codeId = longToString(reader.uint64());
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.adminAddress = reader.string();
                    continue;
                case 8:
                    if (tag !== 66) {
                        break;
                    }
                    message.granterAddress = reader.string();
                    continue;
                case 9:
                    if (tag !== 72) {
                        break;
                    }
                    message.fundingMode = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            contractAddress: isSet(object.contractAddress) ? globalThis.String(object.contractAddress) : "",
            gasLimit: isSet(object.gasLimit) ? globalThis.String(object.gasLimit) : "0",
            gasPrice: isSet(object.gasPrice) ? globalThis.String(object.gasPrice) : "0",
            shouldPinContract: isSet(object.shouldPinContract) ? globalThis.Boolean(object.shouldPinContract) : false,
            isMigrationAllowed: isSet(object.isMigrationAllowed) ? globalThis.Boolean(object.isMigrationAllowed) : false,
            codeId: isSet(object.codeId) ? globalThis.String(object.codeId) : "0",
            adminAddress: isSet(object.adminAddress) ? globalThis.String(object.adminAddress) : "",
            granterAddress: isSet(object.granterAddress) ? globalThis.String(object.granterAddress) : "",
            fundingMode: isSet(object.fundingMode) ? fundingModeFromJSON(object.fundingMode) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.contractAddress !== "") {
            obj.contractAddress = message.contractAddress;
        }
        if (message.gasLimit !== "0") {
            obj.gasLimit = message.gasLimit;
        }
        if (message.gasPrice !== "0") {
            obj.gasPrice = message.gasPrice;
        }
        if (message.shouldPinContract === true) {
            obj.shouldPinContract = message.shouldPinContract;
        }
        if (message.isMigrationAllowed === true) {
            obj.isMigrationAllowed = message.isMigrationAllowed;
        }
        if (message.codeId !== "0") {
            obj.codeId = message.codeId;
        }
        if (message.adminAddress !== "") {
            obj.adminAddress = message.adminAddress;
        }
        if (message.granterAddress !== "") {
            obj.granterAddress = message.granterAddress;
        }
        if (message.fundingMode !== 0) {
            obj.fundingMode = fundingModeToJSON(message.fundingMode);
        }
        return obj;
    },
    create(base) {
        return ContractRegistrationRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        const message = createBaseContractRegistrationRequest();
        message.contractAddress = (_a = object.contractAddress) !== null && _a !== void 0 ? _a : "";
        message.gasLimit = (_b = object.gasLimit) !== null && _b !== void 0 ? _b : "0";
        message.gasPrice = (_c = object.gasPrice) !== null && _c !== void 0 ? _c : "0";
        message.shouldPinContract = (_d = object.shouldPinContract) !== null && _d !== void 0 ? _d : false;
        message.isMigrationAllowed = (_e = object.isMigrationAllowed) !== null && _e !== void 0 ? _e : false;
        message.codeId = (_f = object.codeId) !== null && _f !== void 0 ? _f : "0";
        message.adminAddress = (_g = object.adminAddress) !== null && _g !== void 0 ? _g : "";
        message.granterAddress = (_h = object.granterAddress) !== null && _h !== void 0 ? _h : "";
        message.fundingMode = (_j = object.fundingMode) !== null && _j !== void 0 ? _j : 0;
        return message;
    },
};
function createBaseBatchStoreCodeProposal() {
    return { title: "", description: "", proposals: [] };
}
export const BatchStoreCodeProposal = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.title !== "") {
            writer.uint32(10).string(message.title);
        }
        if (message.description !== "") {
            writer.uint32(18).string(message.description);
        }
        for (const v of message.proposals) {
            StoreCodeProposal.encode(v, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseBatchStoreCodeProposal();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.title = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.description = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.proposals.push(StoreCodeProposal.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            title: isSet(object.title) ? globalThis.String(object.title) : "",
            description: isSet(object.description) ? globalThis.String(object.description) : "",
            proposals: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.proposals)
                ? object.proposals.map((e) => StoreCodeProposal.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.title !== "") {
            obj.title = message.title;
        }
        if (message.description !== "") {
            obj.description = message.description;
        }
        if ((_a = message.proposals) === null || _a === void 0 ? void 0 : _a.length) {
            obj.proposals = message.proposals.map((e) => StoreCodeProposal.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return BatchStoreCodeProposal.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseBatchStoreCodeProposal();
        message.title = (_a = object.title) !== null && _a !== void 0 ? _a : "";
        message.description = (_b = object.description) !== null && _b !== void 0 ? _b : "";
        message.proposals = ((_c = object.proposals) === null || _c === void 0 ? void 0 : _c.map((e) => StoreCodeProposal.fromPartial(e))) || [];
        return message;
    },
};
function longToString(long) {
    return long.toString();
}
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
