// @injectivelabs/core-proto-ts/esm/ibc/applications/interchain_accounts/controller/v1/controller.js
/* eslint-disable */
import _m0 from "protobufjs/minimal";
function createBaseParams() {
    return { controllerEnabled: false };
}
export const Params = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.controllerEnabled === true) {
            writer.uint32(8).bool(message.controllerEnabled);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.controllerEnabled = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            controllerEnabled: isSet(object.controllerEnabled) ? globalThis.Boolean(object.controllerEnabled) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.controllerEnabled === true) {
            obj.controllerEnabled = message.controllerEnabled;
        }
        return obj;
    },
    create(base) {
        return Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseParams();
        message.controllerEnabled = (_a = object.controllerEnabled) !== null && _a !== void 0 ? _a : false;
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
