// @injectivelabs/sdk-ts/dist/esm/core/modules/exchange/msgs/MsgAdminUpdateBinaryOptionsMarket.js
import { InjectiveExchangeV1Beta1Tx } from '@injectivelabs/core-proto-ts';
import { MsgBase } from '../../MsgBase';
import snakecaseKeys from 'snakecase-keys';
/**
 * @category Messages
 */
export default class MsgAdminUpdateBinaryOptionsMarket extends MsgBase {
    static fromJSON(params) {
        return new MsgAdminUpdateBinaryOptionsMarket(params);
    }
    toProto() {
        const { params } = this;
        const message = InjectiveExchangeV1Beta1Tx.MsgAdminUpdateBinaryOptionsMarket.create();
        message.sender = params.sender;
        message.marketId = params.marketId;
        message.settlementPrice = params.settlementPrice;
        message.settlementTimestamp = params.settlementTimestamp;
        message.expirationTimestamp = params.expirationTimestamp;
        message.status = params.status;
        return InjectiveExchangeV1Beta1Tx.MsgAdminUpdateBinaryOptionsMarket.fromPartial(message);
    }
    toData() {
        const proto = this.toProto();
        return {
            '@type': '/injective.exchange.v1beta1.MsgAdminUpdateBinaryOptionsMarket',
            ...proto,
        };
    }
    toAmino() {
        const proto = this.toProto();
        const message = {
            ...snakecaseKeys(proto),
        };
        return {
            type: 'exchange/MsgAdminUpdateBinaryOptionsMarket',
            value: message,
        };
    }
    toWeb3() {
        const amino = this.toAmino();
        const { value } = amino;
        return {
            '@type': '/injective.exchange.v1beta1.MsgAdminUpdateBinaryOptionsMarket',
            ...value,
        };
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/injective.exchange.v1beta1.MsgAdminUpdateBinaryOptionsMarket',
            message: proto,
        };
    }
    toBinary() {
        return InjectiveExchangeV1Beta1Tx.MsgAdminUpdateBinaryOptionsMarket.encode(this.toProto()).finish();
    }
}
