// @wormhole-foundation/sdk-solana-tokenbridge/dist/cjs/utils/tokenBridge/instructions/attestToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAttestTokenAccounts = exports.createAttestTokenInstruction = void 0;
const web3_js_1 = require("@solana/web3.js");
const program_js_1 = require("../program.js");
const sdk_solana_core_1 = require("@wormhole-foundation/sdk-solana-core");
const splMetadata_js_1 = require("../../splMetadata.js");
const index_js_1 = require("./../accounts/index.js");
function createAttestTokenInstruction(connection, tokenBridgeProgramId, wormholeProgramId, payer, mint, message, nonce) {
    const methods = (0, program_js_1.createReadOnlyTokenBridgeProgramInterface)(tokenBridgeProgramId, connection).methods.attestToken(nonce);
    console.log(getAttestTokenAccounts(tokenBridgeProgramId, wormholeProgramId, payer, mint, message));
    // @ts-ignore
    return methods._ixFn(...methods._args, {
        accounts: getAttestTokenAccounts(tokenBridgeProgramId, wormholeProgramId, payer, mint, message),
        signers: undefined,
        remainingAccounts: undefined,
        preInstructions: undefined,
        postInstructions: undefined,
    });
}
exports.createAttestTokenInstruction = createAttestTokenInstruction;
function getAttestTokenAccounts(tokenBridgeProgramId, wormholeProgramId, payer, mint, message) {
    const { bridge: wormholeBridge, emitter: wormholeEmitter, sequence: wormholeSequence, feeCollector: wormholeFeeCollector, clock, rent, systemProgram, } = sdk_solana_core_1.utils.getPostMessageAccounts(wormholeProgramId, payer, message, tokenBridgeProgramId);
    return {
        payer: new web3_js_1.PublicKey(payer),
        config: (0, index_js_1.deriveTokenBridgeConfigKey)(tokenBridgeProgramId),
        mint: new web3_js_1.PublicKey(mint),
        wrappedMeta: (0, index_js_1.deriveWrappedMetaKey)(tokenBridgeProgramId, mint),
        splMetadata: (0, splMetadata_js_1.deriveSplTokenMetadataKey)(mint),
        wormholeBridge,
        wormholeMessage: new web3_js_1.PublicKey(message),
        wormholeEmitter,
        wormholeSequence,
        wormholeFeeCollector,
        clock,
        rent,
        systemProgram,
        wormholeProgram: new web3_js_1.PublicKey(wormholeProgramId),
    };
}
exports.getAttestTokenAccounts = getAttestTokenAccounts;
