// @wormhole-foundation/sdk-solana-tokenbridge/dist/cjs/utils/tokenBridge/instructions/createWrapped.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCreateWrappedAccounts = exports.createCreateWrappedInstruction = void 0;
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const sdk_connect_1 = require("@wormhole-foundation/sdk-connect");
const sdk_solana_core_1 = require("@wormhole-foundation/sdk-solana-core");
const splMetadata_js_1 = require("../../splMetadata.js");
const index_js_1 = require("./../accounts/index.js");
const program_js_1 = require("../program.js");
function createCreateWrappedInstruction(connection, tokenBridgeProgramId, wormholeProgramId, payer, vaa) {
    const methods = (0, program_js_1.createReadOnlyTokenBridgeProgramInterface)(tokenBridgeProgramId, connection).methods.createWrapped();
    // @ts-ignore
    return methods._ixFn(...methods._args, {
        accounts: getCreateWrappedAccounts(tokenBridgeProgramId, wormholeProgramId, payer, vaa),
        signers: undefined,
        remainingAccounts: undefined,
        preInstructions: undefined,
        postInstructions: undefined,
    });
}
exports.createCreateWrappedInstruction = createCreateWrappedInstruction;
function getCreateWrappedAccounts(tokenBridgeProgramId, wormholeProgramId, payer, vaa) {
    const mint = (0, index_js_1.deriveWrappedMintKey)(tokenBridgeProgramId, (0, sdk_connect_1.toChainId)(vaa.payload.token.chain), vaa.payload.token.address.toUint8Array());
    return {
        payer: new web3_js_1.PublicKey(payer),
        config: (0, index_js_1.deriveTokenBridgeConfigKey)(tokenBridgeProgramId),
        endpoint: (0, index_js_1.deriveEndpointKey)(tokenBridgeProgramId, (0, sdk_connect_1.toChainId)(vaa.emitterChain), vaa.emitterAddress.toUint8Array()),
        vaa: sdk_solana_core_1.utils.derivePostedVaaKey(wormholeProgramId, Buffer.from(vaa.hash)),
        claim: sdk_solana_core_1.utils.deriveClaimKey(tokenBridgeProgramId, vaa.emitterAddress.toUint8Array(), (0, sdk_connect_1.toChainId)(vaa.emitterChain), vaa.sequence),
        mint,
        wrappedMeta: (0, index_js_1.deriveWrappedMetaKey)(tokenBridgeProgramId, mint),
        splMetadata: (0, splMetadata_js_1.deriveSplTokenMetadataKey)(mint),
        mintAuthority: (0, index_js_1.deriveMintAuthorityKey)(tokenBridgeProgramId),
        rent: web3_js_1.SYSVAR_RENT_PUBKEY,
        systemProgram: web3_js_1.SystemProgram.programId,
        tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
        splMetadataProgram: splMetadata_js_1.SplTokenMetadataProgram.programId,
        wormholeProgram: new web3_js_1.PublicKey(wormholeProgramId),
    };
}
exports.getCreateWrappedAccounts = getCreateWrappedAccounts;
