// @wormhole-foundation/sdk-solana-tokenbridge/dist/esm/utils/tokenBridge/instructions/attestToken.js
import { PublicKey } from '@solana/web3.js';
import { createReadOnlyTokenBridgeProgramInterface } from '../program.js';
import { utils as coreUtils } from '@wormhole-foundation/sdk-solana-core';
import { deriveSplTokenMetadataKey } from '../../splMetadata.js';
import { deriveTokenBridgeConfigKey, deriveWrappedMetaKey, } from './../accounts/index.js';
export function createAttestTokenInstruction(connection, tokenBridgeProgramId, wormholeProgramId, payer, mint, message, nonce) {
    const methods = createReadOnlyTokenBridgeProgramInterface(tokenBridgeProgramId, connection).methods.attestToken(nonce);
    console.log(getAttestTokenAccounts(tokenBridgeProgramId, wormholeProgramId, payer, mint, message));
    // @ts-ignore
    return methods._ixFn(...methods._args, {
        accounts: getAttestTokenAccounts(tokenBridgeProgramId, wormholeProgramId, payer, mint, message),
        signers: undefined,
        remainingAccounts: undefined,
        preInstructions: undefined,
        postInstructions: undefined,
    });
}
export function getAttestTokenAccounts(tokenBridgeProgramId, wormholeProgramId, payer, mint, message) {
    const { bridge: wormholeBridge, emitter: wormholeEmitter, sequence: wormholeSequence, feeCollector: wormholeFeeCollector, clock, rent, systemProgram, } = coreUtils.getPostMessageAccounts(wormholeProgramId, payer, message, tokenBridgeProgramId);
    return {
        payer: new PublicKey(payer),
        config: deriveTokenBridgeConfigKey(tokenBridgeProgramId),
        mint: new PublicKey(mint),
        wrappedMeta: deriveWrappedMetaKey(tokenBridgeProgramId, mint),
        splMetadata: deriveSplTokenMetadataKey(mint),
        wormholeBridge,
        wormholeMessage: new PublicKey(message),
        wormholeEmitter,
        wormholeSequence,
        wormholeFeeCollector,
        clock,
        rent,
        systemProgram,
        wormholeProgram: new PublicKey(wormholeProgramId),
    };
}
