// @injectivelabs/core-proto-ts/cjs/testpb/bank_query.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcWebError = exports.GrpcWebImpl = exports.BankQueryServiceListSupplyDesc = exports.BankQueryServiceGetSupplyDesc = exports.BankQueryServiceListBalanceDesc = exports.BankQueryServiceGetBalanceDesc = exports.BankQueryServiceDesc = exports.BankQueryServiceClientImpl = exports.ListSupplyResponse = exports.ListSupplyRequest_RangeQuery = exports.ListSupplyRequest_IndexKey_Denom = exports.ListSupplyRequest_IndexKey = exports.ListSupplyRequest = exports.GetSupplyResponse = exports.GetSupplyRequest = exports.ListBalanceResponse = exports.ListBalanceRequest_RangeQuery = exports.ListBalanceRequest_IndexKey_Denom = exports.ListBalanceRequest_IndexKey_AddressDenom = exports.ListBalanceRequest_IndexKey = exports.ListBalanceRequest = exports.GetBalanceResponse = exports.GetBalanceRequest = void 0;
/* eslint-disable */
const grpc_web_1 = require("@injectivelabs/grpc-web");
const browser_headers_1 = require("browser-headers");
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const pagination_1 = require("../cosmos/base/query/v1beta1/pagination");
const bank_1 = require("./bank");
function createBaseGetBalanceRequest() {
    return { address: "", denom: "" };
}
exports.GetBalanceRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.address !== "") {
            writer.uint32(10).string(message.address);
        }
        if (message.denom !== "") {
            writer.uint32(18).string(message.denom);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetBalanceRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.address = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.denom = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            address: isSet(object.address) ? globalThis.String(object.address) : "",
            denom: isSet(object.denom) ? globalThis.String(object.denom) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.address !== "") {
            obj.address = message.address;
        }
        if (message.denom !== "") {
            obj.denom = message.denom;
        }
        return obj;
    },
    create(base) {
        return exports.GetBalanceRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseGetBalanceRequest();
        message.address = (_a = object.address) !== null && _a !== void 0 ? _a : "";
        message.denom = (_b = object.denom) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBaseGetBalanceResponse() {
    return { value: undefined };
}
exports.GetBalanceResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== undefined) {
            bank_1.Balance.encode(message.value, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetBalanceResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = bank_1.Balance.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { value: isSet(object.value) ? bank_1.Balance.fromJSON(object.value) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.value !== undefined) {
            obj.value = bank_1.Balance.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.GetBalanceResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGetBalanceResponse();
        message.value = (object.value !== undefined && object.value !== null)
            ? bank_1.Balance.fromPartial(object.value)
            : undefined;
        return message;
    },
};
function createBaseListBalanceRequest() {
    return { prefixQuery: undefined, rangeQuery: undefined, pagination: undefined };
}
exports.ListBalanceRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.prefixQuery !== undefined) {
            exports.ListBalanceRequest_IndexKey.encode(message.prefixQuery, writer.uint32(10).fork()).ldelim();
        }
        if (message.rangeQuery !== undefined) {
            exports.ListBalanceRequest_RangeQuery.encode(message.rangeQuery, writer.uint32(18).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListBalanceRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.prefixQuery = exports.ListBalanceRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.rangeQuery = exports.ListBalanceRequest_RangeQuery.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            prefixQuery: isSet(object.prefixQuery) ? exports.ListBalanceRequest_IndexKey.fromJSON(object.prefixQuery) : undefined,
            rangeQuery: isSet(object.rangeQuery) ? exports.ListBalanceRequest_RangeQuery.fromJSON(object.rangeQuery) : undefined,
            pagination: isSet(object.pagination) ? pagination_1.PageRequest.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.prefixQuery !== undefined) {
            obj.prefixQuery = exports.ListBalanceRequest_IndexKey.toJSON(message.prefixQuery);
        }
        if (message.rangeQuery !== undefined) {
            obj.rangeQuery = exports.ListBalanceRequest_RangeQuery.toJSON(message.rangeQuery);
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListBalanceRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListBalanceRequest();
        message.prefixQuery = (object.prefixQuery !== undefined && object.prefixQuery !== null)
            ? exports.ListBalanceRequest_IndexKey.fromPartial(object.prefixQuery)
            : undefined;
        message.rangeQuery = (object.rangeQuery !== undefined && object.rangeQuery !== null)
            ? exports.ListBalanceRequest_RangeQuery.fromPartial(object.rangeQuery)
            : undefined;
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageRequest.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
function createBaseListBalanceRequest_IndexKey() {
    return { addressDenom: undefined, denom: undefined };
}
exports.ListBalanceRequest_IndexKey = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.addressDenom !== undefined) {
            exports.ListBalanceRequest_IndexKey_AddressDenom.encode(message.addressDenom, writer.uint32(10).fork()).ldelim();
        }
        if (message.denom !== undefined) {
            exports.ListBalanceRequest_IndexKey_Denom.encode(message.denom, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListBalanceRequest_IndexKey();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.addressDenom = exports.ListBalanceRequest_IndexKey_AddressDenom.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.denom = exports.ListBalanceRequest_IndexKey_Denom.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            addressDenom: isSet(object.addressDenom)
                ? exports.ListBalanceRequest_IndexKey_AddressDenom.fromJSON(object.addressDenom)
                : undefined,
            denom: isSet(object.denom) ? exports.ListBalanceRequest_IndexKey_Denom.fromJSON(object.denom) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.addressDenom !== undefined) {
            obj.addressDenom = exports.ListBalanceRequest_IndexKey_AddressDenom.toJSON(message.addressDenom);
        }
        if (message.denom !== undefined) {
            obj.denom = exports.ListBalanceRequest_IndexKey_Denom.toJSON(message.denom);
        }
        return obj;
    },
    create(base) {
        return exports.ListBalanceRequest_IndexKey.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListBalanceRequest_IndexKey();
        message.addressDenom = (object.addressDenom !== undefined && object.addressDenom !== null)
            ? exports.ListBalanceRequest_IndexKey_AddressDenom.fromPartial(object.addressDenom)
            : undefined;
        message.denom = (object.denom !== undefined && object.denom !== null)
            ? exports.ListBalanceRequest_IndexKey_Denom.fromPartial(object.denom)
            : undefined;
        return message;
    },
};
function createBaseListBalanceRequest_IndexKey_AddressDenom() {
    return { address: undefined, denom: undefined };
}
exports.ListBalanceRequest_IndexKey_AddressDenom = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.address !== undefined) {
            writer.uint32(10).string(message.address);
        }
        if (message.denom !== undefined) {
            writer.uint32(18).string(message.denom);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListBalanceRequest_IndexKey_AddressDenom();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.address = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.denom = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            address: isSet(object.address) ? globalThis.String(object.address) : undefined,
            denom: isSet(object.denom) ? globalThis.String(object.denom) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.address !== undefined) {
            obj.address = message.address;
        }
        if (message.denom !== undefined) {
            obj.denom = message.denom;
        }
        return obj;
    },
    create(base) {
        return exports.ListBalanceRequest_IndexKey_AddressDenom.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseListBalanceRequest_IndexKey_AddressDenom();
        message.address = (_a = object.address) !== null && _a !== void 0 ? _a : undefined;
        message.denom = (_b = object.denom) !== null && _b !== void 0 ? _b : undefined;
        return message;
    },
};
function createBaseListBalanceRequest_IndexKey_Denom() {
    return { denom: undefined };
}
exports.ListBalanceRequest_IndexKey_Denom = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.denom !== undefined) {
            writer.uint32(10).string(message.denom);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListBalanceRequest_IndexKey_Denom();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.denom = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { denom: isSet(object.denom) ? globalThis.String(object.denom) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.denom !== undefined) {
            obj.denom = message.denom;
        }
        return obj;
    },
    create(base) {
        return exports.ListBalanceRequest_IndexKey_Denom.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListBalanceRequest_IndexKey_Denom();
        message.denom = (_a = object.denom) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseListBalanceRequest_RangeQuery() {
    return { from: undefined, to: undefined };
}
exports.ListBalanceRequest_RangeQuery = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.from !== undefined) {
            exports.ListBalanceRequest_IndexKey.encode(message.from, writer.uint32(10).fork()).ldelim();
        }
        if (message.to !== undefined) {
            exports.ListBalanceRequest_IndexKey.encode(message.to, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListBalanceRequest_RangeQuery();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.from = exports.ListBalanceRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.to = exports.ListBalanceRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            from: isSet(object.from) ? exports.ListBalanceRequest_IndexKey.fromJSON(object.from) : undefined,
            to: isSet(object.to) ? exports.ListBalanceRequest_IndexKey.fromJSON(object.to) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.from !== undefined) {
            obj.from = exports.ListBalanceRequest_IndexKey.toJSON(message.from);
        }
        if (message.to !== undefined) {
            obj.to = exports.ListBalanceRequest_IndexKey.toJSON(message.to);
        }
        return obj;
    },
    create(base) {
        return exports.ListBalanceRequest_RangeQuery.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListBalanceRequest_RangeQuery();
        message.from = (object.from !== undefined && object.from !== null)
            ? exports.ListBalanceRequest_IndexKey.fromPartial(object.from)
            : undefined;
        message.to = (object.to !== undefined && object.to !== null)
            ? exports.ListBalanceRequest_IndexKey.fromPartial(object.to)
            : undefined;
        return message;
    },
};
function createBaseListBalanceResponse() {
    return { values: [], pagination: undefined };
}
exports.ListBalanceResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.values) {
            bank_1.Balance.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListBalanceResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.values.push(bank_1.Balance.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            values: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.values) ? object.values.map((e) => bank_1.Balance.fromJSON(e)) : [],
            pagination: isSet(object.pagination) ? pagination_1.PageResponse.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.values) === null || _a === void 0 ? void 0 : _a.length) {
            obj.values = message.values.map((e) => bank_1.Balance.toJSON(e));
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListBalanceResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListBalanceResponse();
        message.values = ((_a = object.values) === null || _a === void 0 ? void 0 : _a.map((e) => bank_1.Balance.fromPartial(e))) || [];
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageResponse.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
function createBaseGetSupplyRequest() {
    return { denom: "" };
}
exports.GetSupplyRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.denom !== "") {
            writer.uint32(10).string(message.denom);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetSupplyRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.denom = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { denom: isSet(object.denom) ? globalThis.String(object.denom) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.denom !== "") {
            obj.denom = message.denom;
        }
        return obj;
    },
    create(base) {
        return exports.GetSupplyRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetSupplyRequest();
        message.denom = (_a = object.denom) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseGetSupplyResponse() {
    return { value: undefined };
}
exports.GetSupplyResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== undefined) {
            bank_1.Supply.encode(message.value, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetSupplyResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = bank_1.Supply.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { value: isSet(object.value) ? bank_1.Supply.fromJSON(object.value) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.value !== undefined) {
            obj.value = bank_1.Supply.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.GetSupplyResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGetSupplyResponse();
        message.value = (object.value !== undefined && object.value !== null)
            ? bank_1.Supply.fromPartial(object.value)
            : undefined;
        return message;
    },
};
function createBaseListSupplyRequest() {
    return { prefixQuery: undefined, rangeQuery: undefined, pagination: undefined };
}
exports.ListSupplyRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.prefixQuery !== undefined) {
            exports.ListSupplyRequest_IndexKey.encode(message.prefixQuery, writer.uint32(10).fork()).ldelim();
        }
        if (message.rangeQuery !== undefined) {
            exports.ListSupplyRequest_RangeQuery.encode(message.rangeQuery, writer.uint32(18).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListSupplyRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.prefixQuery = exports.ListSupplyRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.rangeQuery = exports.ListSupplyRequest_RangeQuery.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            prefixQuery: isSet(object.prefixQuery) ? exports.ListSupplyRequest_IndexKey.fromJSON(object.prefixQuery) : undefined,
            rangeQuery: isSet(object.rangeQuery) ? exports.ListSupplyRequest_RangeQuery.fromJSON(object.rangeQuery) : undefined,
            pagination: isSet(object.pagination) ? pagination_1.PageRequest.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.prefixQuery !== undefined) {
            obj.prefixQuery = exports.ListSupplyRequest_IndexKey.toJSON(message.prefixQuery);
        }
        if (message.rangeQuery !== undefined) {
            obj.rangeQuery = exports.ListSupplyRequest_RangeQuery.toJSON(message.rangeQuery);
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListSupplyRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListSupplyRequest();
        message.prefixQuery = (object.prefixQuery !== undefined && object.prefixQuery !== null)
            ? exports.ListSupplyRequest_IndexKey.fromPartial(object.prefixQuery)
            : undefined;
        message.rangeQuery = (object.rangeQuery !== undefined && object.rangeQuery !== null)
            ? exports.ListSupplyRequest_RangeQuery.fromPartial(object.rangeQuery)
            : undefined;
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageRequest.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
function createBaseListSupplyRequest_IndexKey() {
    return { denom: undefined };
}
exports.ListSupplyRequest_IndexKey = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.denom !== undefined) {
            exports.ListSupplyRequest_IndexKey_Denom.encode(message.denom, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListSupplyRequest_IndexKey();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.denom = exports.ListSupplyRequest_IndexKey_Denom.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { denom: isSet(object.denom) ? exports.ListSupplyRequest_IndexKey_Denom.fromJSON(object.denom) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.denom !== undefined) {
            obj.denom = exports.ListSupplyRequest_IndexKey_Denom.toJSON(message.denom);
        }
        return obj;
    },
    create(base) {
        return exports.ListSupplyRequest_IndexKey.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListSupplyRequest_IndexKey();
        message.denom = (object.denom !== undefined && object.denom !== null)
            ? exports.ListSupplyRequest_IndexKey_Denom.fromPartial(object.denom)
            : undefined;
        return message;
    },
};
function createBaseListSupplyRequest_IndexKey_Denom() {
    return { denom: undefined };
}
exports.ListSupplyRequest_IndexKey_Denom = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.denom !== undefined) {
            writer.uint32(10).string(message.denom);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListSupplyRequest_IndexKey_Denom();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.denom = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { denom: isSet(object.denom) ? globalThis.String(object.denom) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.denom !== undefined) {
            obj.denom = message.denom;
        }
        return obj;
    },
    create(base) {
        return exports.ListSupplyRequest_IndexKey_Denom.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListSupplyRequest_IndexKey_Denom();
        message.denom = (_a = object.denom) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseListSupplyRequest_RangeQuery() {
    return { from: undefined, to: undefined };
}
exports.ListSupplyRequest_RangeQuery = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.from !== undefined) {
            exports.ListSupplyRequest_IndexKey.encode(message.from, writer.uint32(10).fork()).ldelim();
        }
        if (message.to !== undefined) {
            exports.ListSupplyRequest_IndexKey.encode(message.to, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListSupplyRequest_RangeQuery();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.from = exports.ListSupplyRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.to = exports.ListSupplyRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            from: isSet(object.from) ? exports.ListSupplyRequest_IndexKey.fromJSON(object.from) : undefined,
            to: isSet(object.to) ? exports.ListSupplyRequest_IndexKey.fromJSON(object.to) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.from !== undefined) {
            obj.from = exports.ListSupplyRequest_IndexKey.toJSON(message.from);
        }
        if (message.to !== undefined) {
            obj.to = exports.ListSupplyRequest_IndexKey.toJSON(message.to);
        }
        return obj;
    },
    create(base) {
        return exports.ListSupplyRequest_RangeQuery.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListSupplyRequest_RangeQuery();
        message.from = (object.from !== undefined && object.from !== null)
            ? exports.ListSupplyRequest_IndexKey.fromPartial(object.from)
            : undefined;
        message.to = (object.to !== undefined && object.to !== null)
            ? exports.ListSupplyRequest_IndexKey.fromPartial(object.to)
            : undefined;
        return message;
    },
};
function createBaseListSupplyResponse() {
    return { values: [], pagination: undefined };
}
exports.ListSupplyResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.values) {
            bank_1.Supply.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListSupplyResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.values.push(bank_1.Supply.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            values: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.values) ? object.values.map((e) => bank_1.Supply.fromJSON(e)) : [],
            pagination: isSet(object.pagination) ? pagination_1.PageResponse.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.values) === null || _a === void 0 ? void 0 : _a.length) {
            obj.values = message.values.map((e) => bank_1.Supply.toJSON(e));
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListSupplyResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListSupplyResponse();
        message.values = ((_a = object.values) === null || _a === void 0 ? void 0 : _a.map((e) => bank_1.Supply.fromPartial(e))) || [];
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageResponse.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
class BankQueryServiceClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.GetBalance = this.GetBalance.bind(this);
        this.ListBalance = this.ListBalance.bind(this);
        this.GetSupply = this.GetSupply.bind(this);
        this.ListSupply = this.ListSupply.bind(this);
    }
    GetBalance(request, metadata) {
        return this.rpc.unary(exports.BankQueryServiceGetBalanceDesc, exports.GetBalanceRequest.fromPartial(request), metadata);
    }
    ListBalance(request, metadata) {
        return this.rpc.unary(exports.BankQueryServiceListBalanceDesc, exports.ListBalanceRequest.fromPartial(request), metadata);
    }
    GetSupply(request, metadata) {
        return this.rpc.unary(exports.BankQueryServiceGetSupplyDesc, exports.GetSupplyRequest.fromPartial(request), metadata);
    }
    ListSupply(request, metadata) {
        return this.rpc.unary(exports.BankQueryServiceListSupplyDesc, exports.ListSupplyRequest.fromPartial(request), metadata);
    }
}
exports.BankQueryServiceClientImpl = BankQueryServiceClientImpl;
exports.BankQueryServiceDesc = { serviceName: "testpb.BankQueryService" };
exports.BankQueryServiceGetBalanceDesc = {
    methodName: "GetBalance",
    service: exports.BankQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetBalanceRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetBalanceResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.BankQueryServiceListBalanceDesc = {
    methodName: "ListBalance",
    service: exports.BankQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.ListBalanceRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.ListBalanceResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.BankQueryServiceGetSupplyDesc = {
    methodName: "GetSupply",
    service: exports.BankQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetSupplyRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetSupplyResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.BankQueryServiceListSupplyDesc = {
    methodName: "ListSupply",
    service: exports.BankQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.ListSupplyRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.ListSupplyResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
class GrpcWebImpl {
    constructor(host, options) {
        this.host = host;
        this.options = options;
    }
    unary(methodDesc, _request, metadata) {
        var _a;
        const request = Object.assign(Object.assign({}, _request), methodDesc.requestType);
        const maybeCombinedMetadata = metadata && this.options.metadata
            ? new browser_headers_1.BrowserHeaders(Object.assign(Object.assign({}, (_a = this.options) === null || _a === void 0 ? void 0 : _a.metadata.headersMap), metadata === null || metadata === void 0 ? void 0 : metadata.headersMap))
            : metadata !== null && metadata !== void 0 ? metadata : this.options.metadata;
        return new Promise((resolve, reject) => {
            var _a;
            grpc_web_1.grpc.unary(methodDesc, Object.assign(Object.assign({ request, host: this.host, metadata: maybeCombinedMetadata !== null && maybeCombinedMetadata !== void 0 ? maybeCombinedMetadata : {} }, (this.options.transport !== undefined ? { transport: this.options.transport } : {})), { debug: (_a = this.options.debug) !== null && _a !== void 0 ? _a : false, onEnd: function (response) {
                    if (response.status === grpc_web_1.grpc.Code.OK) {
                        resolve(response.message.toObject());
                    }
                    else {
                        const err = new GrpcWebError(response.statusMessage, response.status, response.trailers);
                        reject(err);
                    }
                } }));
        });
    }
}
exports.GrpcWebImpl = GrpcWebImpl;
function isSet(value) {
    return value !== null && value !== undefined;
}
class GrpcWebError extends globalThis.Error {
    constructor(message, code, metadata) {
        super(message);
        this.code = code;
        this.metadata = metadata;
    }
}
exports.GrpcWebError = GrpcWebError;
