// @injectivelabs/core-proto-ts/cjs/testpb/test_schema_query.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestSchemaQueryServiceClientImpl = exports.ListExampleAutoIncFieldNameResponse = exports.ListExampleAutoIncFieldNameRequest_RangeQuery = exports.ListExampleAutoIncFieldNameRequest_IndexKey_Foo = exports.ListExampleAutoIncFieldNameRequest_IndexKey = exports.ListExampleAutoIncFieldNameRequest = exports.GetExampleAutoIncFieldNameResponse = exports.GetExampleAutoIncFieldNameRequest = exports.ListSimpleExampleResponse = exports.ListSimpleExampleRequest_RangeQuery = exports.ListSimpleExampleRequest_IndexKey_Unique = exports.ListSimpleExampleRequest_IndexKey_Name = exports.ListSimpleExampleRequest_IndexKey = exports.ListSimpleExampleRequest = exports.GetSimpleExampleByUniqueResponse = exports.GetSimpleExampleByUniqueRequest = exports.GetSimpleExampleResponse = exports.GetSimpleExampleRequest = exports.ListExampleTimestampResponse = exports.ListExampleTimestampRequest_RangeQuery = exports.ListExampleTimestampRequest_IndexKey_Ts = exports.ListExampleTimestampRequest_IndexKey_Id = exports.ListExampleTimestampRequest_IndexKey = exports.ListExampleTimestampRequest = exports.GetExampleTimestampResponse = exports.GetExampleTimestampRequest = exports.GetExampleSingletonResponse = exports.GetExampleSingletonRequest = exports.ListExampleAutoIncrementTableResponse = exports.ListExampleAutoIncrementTableRequest_RangeQuery = exports.ListExampleAutoIncrementTableRequest_IndexKey_X = exports.ListExampleAutoIncrementTableRequest_IndexKey_Id = exports.ListExampleAutoIncrementTableRequest_IndexKey = exports.ListExampleAutoIncrementTableRequest = exports.GetExampleAutoIncrementTableByXResponse = exports.GetExampleAutoIncrementTableByXRequest = exports.GetExampleAutoIncrementTableResponse = exports.GetExampleAutoIncrementTableRequest = exports.ListExampleTableResponse = exports.ListExampleTableRequest_RangeQuery = exports.ListExampleTableRequest_IndexKey_BzStr = exports.ListExampleTableRequest_IndexKey_StrU32 = exports.ListExampleTableRequest_IndexKey_U64Str = exports.ListExampleTableRequest_IndexKey_U32I64Str = exports.ListExampleTableRequest_IndexKey = exports.ListExampleTableRequest = exports.GetExampleTableByU64StrResponse = exports.GetExampleTableByU64StrRequest = exports.GetExampleTableResponse = exports.GetExampleTableRequest = void 0;
exports.GrpcWebError = exports.GrpcWebImpl = exports.TestSchemaQueryServiceListExampleAutoIncFieldNameDesc = exports.TestSchemaQueryServiceGetExampleAutoIncFieldNameDesc = exports.TestSchemaQueryServiceListSimpleExampleDesc = exports.TestSchemaQueryServiceGetSimpleExampleByUniqueDesc = exports.TestSchemaQueryServiceGetSimpleExampleDesc = exports.TestSchemaQueryServiceListExampleTimestampDesc = exports.TestSchemaQueryServiceGetExampleTimestampDesc = exports.TestSchemaQueryServiceGetExampleSingletonDesc = exports.TestSchemaQueryServiceListExampleAutoIncrementTableDesc = exports.TestSchemaQueryServiceGetExampleAutoIncrementTableByXDesc = exports.TestSchemaQueryServiceGetExampleAutoIncrementTableDesc = exports.TestSchemaQueryServiceListExampleTableDesc = exports.TestSchemaQueryServiceGetExampleTableByU64StrDesc = exports.TestSchemaQueryServiceGetExampleTableDesc = exports.TestSchemaQueryServiceDesc = void 0;
/* eslint-disable */
const grpc_web_1 = require("@injectivelabs/grpc-web");
const browser_headers_1 = require("browser-headers");
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const pagination_1 = require("../cosmos/base/query/v1beta1/pagination");
const timestamp_1 = require("../google/protobuf/timestamp");
const test_schema_1 = require("./test_schema");
function createBaseGetExampleTableRequest() {
    return { u32: 0, i64: "0", str: "" };
}
exports.GetExampleTableRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.u32 !== 0) {
            writer.uint32(8).uint32(message.u32);
        }
        if (message.i64 !== "0") {
            writer.uint32(16).int64(message.i64);
        }
        if (message.str !== "") {
            writer.uint32(26).string(message.str);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleTableRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.u32 = reader.uint32();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.i64 = longToString(reader.int64());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.str = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            u32: isSet(object.u32) ? globalThis.Number(object.u32) : 0,
            i64: isSet(object.i64) ? globalThis.String(object.i64) : "0",
            str: isSet(object.str) ? globalThis.String(object.str) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.u32 !== 0) {
            obj.u32 = Math.round(message.u32);
        }
        if (message.i64 !== "0") {
            obj.i64 = message.i64;
        }
        if (message.str !== "") {
            obj.str = message.str;
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleTableRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseGetExampleTableRequest();
        message.u32 = (_a = object.u32) !== null && _a !== void 0 ? _a : 0;
        message.i64 = (_b = object.i64) !== null && _b !== void 0 ? _b : "0";
        message.str = (_c = object.str) !== null && _c !== void 0 ? _c : "";
        return message;
    },
};
function createBaseGetExampleTableResponse() {
    return { value: undefined };
}
exports.GetExampleTableResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== undefined) {
            test_schema_1.ExampleTable.encode(message.value, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleTableResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = test_schema_1.ExampleTable.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { value: isSet(object.value) ? test_schema_1.ExampleTable.fromJSON(object.value) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.value !== undefined) {
            obj.value = test_schema_1.ExampleTable.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleTableResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGetExampleTableResponse();
        message.value = (object.value !== undefined && object.value !== null)
            ? test_schema_1.ExampleTable.fromPartial(object.value)
            : undefined;
        return message;
    },
};
function createBaseGetExampleTableByU64StrRequest() {
    return { u64: "0", str: "" };
}
exports.GetExampleTableByU64StrRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.u64 !== "0") {
            writer.uint32(8).uint64(message.u64);
        }
        if (message.str !== "") {
            writer.uint32(18).string(message.str);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleTableByU64StrRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.u64 = longToString(reader.uint64());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.str = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            u64: isSet(object.u64) ? globalThis.String(object.u64) : "0",
            str: isSet(object.str) ? globalThis.String(object.str) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.u64 !== "0") {
            obj.u64 = message.u64;
        }
        if (message.str !== "") {
            obj.str = message.str;
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleTableByU64StrRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseGetExampleTableByU64StrRequest();
        message.u64 = (_a = object.u64) !== null && _a !== void 0 ? _a : "0";
        message.str = (_b = object.str) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBaseGetExampleTableByU64StrResponse() {
    return { value: undefined };
}
exports.GetExampleTableByU64StrResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== undefined) {
            test_schema_1.ExampleTable.encode(message.value, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleTableByU64StrResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = test_schema_1.ExampleTable.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { value: isSet(object.value) ? test_schema_1.ExampleTable.fromJSON(object.value) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.value !== undefined) {
            obj.value = test_schema_1.ExampleTable.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleTableByU64StrResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGetExampleTableByU64StrResponse();
        message.value = (object.value !== undefined && object.value !== null)
            ? test_schema_1.ExampleTable.fromPartial(object.value)
            : undefined;
        return message;
    },
};
function createBaseListExampleTableRequest() {
    return { prefixQuery: undefined, rangeQuery: undefined, pagination: undefined };
}
exports.ListExampleTableRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.prefixQuery !== undefined) {
            exports.ListExampleTableRequest_IndexKey.encode(message.prefixQuery, writer.uint32(10).fork()).ldelim();
        }
        if (message.rangeQuery !== undefined) {
            exports.ListExampleTableRequest_RangeQuery.encode(message.rangeQuery, writer.uint32(18).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTableRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.prefixQuery = exports.ListExampleTableRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.rangeQuery = exports.ListExampleTableRequest_RangeQuery.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            prefixQuery: isSet(object.prefixQuery)
                ? exports.ListExampleTableRequest_IndexKey.fromJSON(object.prefixQuery)
                : undefined,
            rangeQuery: isSet(object.rangeQuery) ? exports.ListExampleTableRequest_RangeQuery.fromJSON(object.rangeQuery) : undefined,
            pagination: isSet(object.pagination) ? pagination_1.PageRequest.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.prefixQuery !== undefined) {
            obj.prefixQuery = exports.ListExampleTableRequest_IndexKey.toJSON(message.prefixQuery);
        }
        if (message.rangeQuery !== undefined) {
            obj.rangeQuery = exports.ListExampleTableRequest_RangeQuery.toJSON(message.rangeQuery);
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTableRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListExampleTableRequest();
        message.prefixQuery = (object.prefixQuery !== undefined && object.prefixQuery !== null)
            ? exports.ListExampleTableRequest_IndexKey.fromPartial(object.prefixQuery)
            : undefined;
        message.rangeQuery = (object.rangeQuery !== undefined && object.rangeQuery !== null)
            ? exports.ListExampleTableRequest_RangeQuery.fromPartial(object.rangeQuery)
            : undefined;
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageRequest.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
function createBaseListExampleTableRequest_IndexKey() {
    return { u32I64Str: undefined, u64Str: undefined, strU32: undefined, bzStr: undefined };
}
exports.ListExampleTableRequest_IndexKey = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.u32I64Str !== undefined) {
            exports.ListExampleTableRequest_IndexKey_U32I64Str.encode(message.u32I64Str, writer.uint32(10).fork()).ldelim();
        }
        if (message.u64Str !== undefined) {
            exports.ListExampleTableRequest_IndexKey_U64Str.encode(message.u64Str, writer.uint32(18).fork()).ldelim();
        }
        if (message.strU32 !== undefined) {
            exports.ListExampleTableRequest_IndexKey_StrU32.encode(message.strU32, writer.uint32(26).fork()).ldelim();
        }
        if (message.bzStr !== undefined) {
            exports.ListExampleTableRequest_IndexKey_BzStr.encode(message.bzStr, writer.uint32(34).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTableRequest_IndexKey();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.u32I64Str = exports.ListExampleTableRequest_IndexKey_U32I64Str.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.u64Str = exports.ListExampleTableRequest_IndexKey_U64Str.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.strU32 = exports.ListExampleTableRequest_IndexKey_StrU32.decode(reader, reader.uint32());
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.bzStr = exports.ListExampleTableRequest_IndexKey_BzStr.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            u32I64Str: isSet(object.u32I64Str)
                ? exports.ListExampleTableRequest_IndexKey_U32I64Str.fromJSON(object.u32I64Str)
                : undefined,
            u64Str: isSet(object.u64Str) ? exports.ListExampleTableRequest_IndexKey_U64Str.fromJSON(object.u64Str) : undefined,
            strU32: isSet(object.strU32) ? exports.ListExampleTableRequest_IndexKey_StrU32.fromJSON(object.strU32) : undefined,
            bzStr: isSet(object.bzStr) ? exports.ListExampleTableRequest_IndexKey_BzStr.fromJSON(object.bzStr) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.u32I64Str !== undefined) {
            obj.u32I64Str = exports.ListExampleTableRequest_IndexKey_U32I64Str.toJSON(message.u32I64Str);
        }
        if (message.u64Str !== undefined) {
            obj.u64Str = exports.ListExampleTableRequest_IndexKey_U64Str.toJSON(message.u64Str);
        }
        if (message.strU32 !== undefined) {
            obj.strU32 = exports.ListExampleTableRequest_IndexKey_StrU32.toJSON(message.strU32);
        }
        if (message.bzStr !== undefined) {
            obj.bzStr = exports.ListExampleTableRequest_IndexKey_BzStr.toJSON(message.bzStr);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTableRequest_IndexKey.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListExampleTableRequest_IndexKey();
        message.u32I64Str = (object.u32I64Str !== undefined && object.u32I64Str !== null)
            ? exports.ListExampleTableRequest_IndexKey_U32I64Str.fromPartial(object.u32I64Str)
            : undefined;
        message.u64Str = (object.u64Str !== undefined && object.u64Str !== null)
            ? exports.ListExampleTableRequest_IndexKey_U64Str.fromPartial(object.u64Str)
            : undefined;
        message.strU32 = (object.strU32 !== undefined && object.strU32 !== null)
            ? exports.ListExampleTableRequest_IndexKey_StrU32.fromPartial(object.strU32)
            : undefined;
        message.bzStr = (object.bzStr !== undefined && object.bzStr !== null)
            ? exports.ListExampleTableRequest_IndexKey_BzStr.fromPartial(object.bzStr)
            : undefined;
        return message;
    },
};
function createBaseListExampleTableRequest_IndexKey_U32I64Str() {
    return { u32: undefined, i64: undefined, str: undefined };
}
exports.ListExampleTableRequest_IndexKey_U32I64Str = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.u32 !== undefined) {
            writer.uint32(8).uint32(message.u32);
        }
        if (message.i64 !== undefined) {
            writer.uint32(16).int64(message.i64);
        }
        if (message.str !== undefined) {
            writer.uint32(26).string(message.str);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTableRequest_IndexKey_U32I64Str();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.u32 = reader.uint32();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.i64 = longToString(reader.int64());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.str = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            u32: isSet(object.u32) ? globalThis.Number(object.u32) : undefined,
            i64: isSet(object.i64) ? globalThis.String(object.i64) : undefined,
            str: isSet(object.str) ? globalThis.String(object.str) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.u32 !== undefined) {
            obj.u32 = Math.round(message.u32);
        }
        if (message.i64 !== undefined) {
            obj.i64 = message.i64;
        }
        if (message.str !== undefined) {
            obj.str = message.str;
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTableRequest_IndexKey_U32I64Str.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseListExampleTableRequest_IndexKey_U32I64Str();
        message.u32 = (_a = object.u32) !== null && _a !== void 0 ? _a : undefined;
        message.i64 = (_b = object.i64) !== null && _b !== void 0 ? _b : undefined;
        message.str = (_c = object.str) !== null && _c !== void 0 ? _c : undefined;
        return message;
    },
};
function createBaseListExampleTableRequest_IndexKey_U64Str() {
    return { u64: undefined, str: undefined };
}
exports.ListExampleTableRequest_IndexKey_U64Str = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.u64 !== undefined) {
            writer.uint32(8).uint64(message.u64);
        }
        if (message.str !== undefined) {
            writer.uint32(18).string(message.str);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTableRequest_IndexKey_U64Str();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.u64 = longToString(reader.uint64());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.str = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            u64: isSet(object.u64) ? globalThis.String(object.u64) : undefined,
            str: isSet(object.str) ? globalThis.String(object.str) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.u64 !== undefined) {
            obj.u64 = message.u64;
        }
        if (message.str !== undefined) {
            obj.str = message.str;
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTableRequest_IndexKey_U64Str.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseListExampleTableRequest_IndexKey_U64Str();
        message.u64 = (_a = object.u64) !== null && _a !== void 0 ? _a : undefined;
        message.str = (_b = object.str) !== null && _b !== void 0 ? _b : undefined;
        return message;
    },
};
function createBaseListExampleTableRequest_IndexKey_StrU32() {
    return { str: undefined, u32: undefined };
}
exports.ListExampleTableRequest_IndexKey_StrU32 = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.str !== undefined) {
            writer.uint32(10).string(message.str);
        }
        if (message.u32 !== undefined) {
            writer.uint32(16).uint32(message.u32);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTableRequest_IndexKey_StrU32();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.str = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.u32 = reader.uint32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            str: isSet(object.str) ? globalThis.String(object.str) : undefined,
            u32: isSet(object.u32) ? globalThis.Number(object.u32) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.str !== undefined) {
            obj.str = message.str;
        }
        if (message.u32 !== undefined) {
            obj.u32 = Math.round(message.u32);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTableRequest_IndexKey_StrU32.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseListExampleTableRequest_IndexKey_StrU32();
        message.str = (_a = object.str) !== null && _a !== void 0 ? _a : undefined;
        message.u32 = (_b = object.u32) !== null && _b !== void 0 ? _b : undefined;
        return message;
    },
};
function createBaseListExampleTableRequest_IndexKey_BzStr() {
    return { bz: undefined, str: undefined };
}
exports.ListExampleTableRequest_IndexKey_BzStr = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.bz !== undefined) {
            writer.uint32(10).bytes(message.bz);
        }
        if (message.str !== undefined) {
            writer.uint32(18).string(message.str);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTableRequest_IndexKey_BzStr();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.bz = reader.bytes();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.str = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            bz: isSet(object.bz) ? bytesFromBase64(object.bz) : undefined,
            str: isSet(object.str) ? globalThis.String(object.str) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.bz !== undefined) {
            obj.bz = base64FromBytes(message.bz);
        }
        if (message.str !== undefined) {
            obj.str = message.str;
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTableRequest_IndexKey_BzStr.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseListExampleTableRequest_IndexKey_BzStr();
        message.bz = (_a = object.bz) !== null && _a !== void 0 ? _a : undefined;
        message.str = (_b = object.str) !== null && _b !== void 0 ? _b : undefined;
        return message;
    },
};
function createBaseListExampleTableRequest_RangeQuery() {
    return { from: undefined, to: undefined };
}
exports.ListExampleTableRequest_RangeQuery = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.from !== undefined) {
            exports.ListExampleTableRequest_IndexKey.encode(message.from, writer.uint32(10).fork()).ldelim();
        }
        if (message.to !== undefined) {
            exports.ListExampleTableRequest_IndexKey.encode(message.to, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTableRequest_RangeQuery();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.from = exports.ListExampleTableRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.to = exports.ListExampleTableRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            from: isSet(object.from) ? exports.ListExampleTableRequest_IndexKey.fromJSON(object.from) : undefined,
            to: isSet(object.to) ? exports.ListExampleTableRequest_IndexKey.fromJSON(object.to) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.from !== undefined) {
            obj.from = exports.ListExampleTableRequest_IndexKey.toJSON(message.from);
        }
        if (message.to !== undefined) {
            obj.to = exports.ListExampleTableRequest_IndexKey.toJSON(message.to);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTableRequest_RangeQuery.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListExampleTableRequest_RangeQuery();
        message.from = (object.from !== undefined && object.from !== null)
            ? exports.ListExampleTableRequest_IndexKey.fromPartial(object.from)
            : undefined;
        message.to = (object.to !== undefined && object.to !== null)
            ? exports.ListExampleTableRequest_IndexKey.fromPartial(object.to)
            : undefined;
        return message;
    },
};
function createBaseListExampleTableResponse() {
    return { values: [], pagination: undefined };
}
exports.ListExampleTableResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.values) {
            test_schema_1.ExampleTable.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTableResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.values.push(test_schema_1.ExampleTable.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            values: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.values) ? object.values.map((e) => test_schema_1.ExampleTable.fromJSON(e)) : [],
            pagination: isSet(object.pagination) ? pagination_1.PageResponse.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.values) === null || _a === void 0 ? void 0 : _a.length) {
            obj.values = message.values.map((e) => test_schema_1.ExampleTable.toJSON(e));
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTableResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListExampleTableResponse();
        message.values = ((_a = object.values) === null || _a === void 0 ? void 0 : _a.map((e) => test_schema_1.ExampleTable.fromPartial(e))) || [];
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageResponse.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
function createBaseGetExampleAutoIncrementTableRequest() {
    return { id: "0" };
}
exports.GetExampleAutoIncrementTableRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== "0") {
            writer.uint32(8).uint64(message.id);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleAutoIncrementTableRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.id = longToString(reader.uint64());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { id: isSet(object.id) ? globalThis.String(object.id) : "0" };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== "0") {
            obj.id = message.id;
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleAutoIncrementTableRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetExampleAutoIncrementTableRequest();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : "0";
        return message;
    },
};
function createBaseGetExampleAutoIncrementTableResponse() {
    return { value: undefined };
}
exports.GetExampleAutoIncrementTableResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== undefined) {
            test_schema_1.ExampleAutoIncrementTable.encode(message.value, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleAutoIncrementTableResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = test_schema_1.ExampleAutoIncrementTable.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { value: isSet(object.value) ? test_schema_1.ExampleAutoIncrementTable.fromJSON(object.value) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.value !== undefined) {
            obj.value = test_schema_1.ExampleAutoIncrementTable.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleAutoIncrementTableResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGetExampleAutoIncrementTableResponse();
        message.value = (object.value !== undefined && object.value !== null)
            ? test_schema_1.ExampleAutoIncrementTable.fromPartial(object.value)
            : undefined;
        return message;
    },
};
function createBaseGetExampleAutoIncrementTableByXRequest() {
    return { x: "" };
}
exports.GetExampleAutoIncrementTableByXRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.x !== "") {
            writer.uint32(10).string(message.x);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleAutoIncrementTableByXRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.x = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { x: isSet(object.x) ? globalThis.String(object.x) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.x !== "") {
            obj.x = message.x;
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleAutoIncrementTableByXRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetExampleAutoIncrementTableByXRequest();
        message.x = (_a = object.x) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseGetExampleAutoIncrementTableByXResponse() {
    return { value: undefined };
}
exports.GetExampleAutoIncrementTableByXResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== undefined) {
            test_schema_1.ExampleAutoIncrementTable.encode(message.value, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleAutoIncrementTableByXResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = test_schema_1.ExampleAutoIncrementTable.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { value: isSet(object.value) ? test_schema_1.ExampleAutoIncrementTable.fromJSON(object.value) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.value !== undefined) {
            obj.value = test_schema_1.ExampleAutoIncrementTable.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleAutoIncrementTableByXResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGetExampleAutoIncrementTableByXResponse();
        message.value = (object.value !== undefined && object.value !== null)
            ? test_schema_1.ExampleAutoIncrementTable.fromPartial(object.value)
            : undefined;
        return message;
    },
};
function createBaseListExampleAutoIncrementTableRequest() {
    return { prefixQuery: undefined, rangeQuery: undefined, pagination: undefined };
}
exports.ListExampleAutoIncrementTableRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.prefixQuery !== undefined) {
            exports.ListExampleAutoIncrementTableRequest_IndexKey.encode(message.prefixQuery, writer.uint32(10).fork()).ldelim();
        }
        if (message.rangeQuery !== undefined) {
            exports.ListExampleAutoIncrementTableRequest_RangeQuery.encode(message.rangeQuery, writer.uint32(18).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleAutoIncrementTableRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.prefixQuery = exports.ListExampleAutoIncrementTableRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.rangeQuery = exports.ListExampleAutoIncrementTableRequest_RangeQuery.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            prefixQuery: isSet(object.prefixQuery)
                ? exports.ListExampleAutoIncrementTableRequest_IndexKey.fromJSON(object.prefixQuery)
                : undefined,
            rangeQuery: isSet(object.rangeQuery)
                ? exports.ListExampleAutoIncrementTableRequest_RangeQuery.fromJSON(object.rangeQuery)
                : undefined,
            pagination: isSet(object.pagination) ? pagination_1.PageRequest.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.prefixQuery !== undefined) {
            obj.prefixQuery = exports.ListExampleAutoIncrementTableRequest_IndexKey.toJSON(message.prefixQuery);
        }
        if (message.rangeQuery !== undefined) {
            obj.rangeQuery = exports.ListExampleAutoIncrementTableRequest_RangeQuery.toJSON(message.rangeQuery);
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleAutoIncrementTableRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListExampleAutoIncrementTableRequest();
        message.prefixQuery = (object.prefixQuery !== undefined && object.prefixQuery !== null)
            ? exports.ListExampleAutoIncrementTableRequest_IndexKey.fromPartial(object.prefixQuery)
            : undefined;
        message.rangeQuery = (object.rangeQuery !== undefined && object.rangeQuery !== null)
            ? exports.ListExampleAutoIncrementTableRequest_RangeQuery.fromPartial(object.rangeQuery)
            : undefined;
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageRequest.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
function createBaseListExampleAutoIncrementTableRequest_IndexKey() {
    return { id: undefined, x: undefined };
}
exports.ListExampleAutoIncrementTableRequest_IndexKey = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== undefined) {
            exports.ListExampleAutoIncrementTableRequest_IndexKey_Id.encode(message.id, writer.uint32(10).fork()).ldelim();
        }
        if (message.x !== undefined) {
            exports.ListExampleAutoIncrementTableRequest_IndexKey_X.encode(message.x, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleAutoIncrementTableRequest_IndexKey();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = exports.ListExampleAutoIncrementTableRequest_IndexKey_Id.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.x = exports.ListExampleAutoIncrementTableRequest_IndexKey_X.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? exports.ListExampleAutoIncrementTableRequest_IndexKey_Id.fromJSON(object.id) : undefined,
            x: isSet(object.x) ? exports.ListExampleAutoIncrementTableRequest_IndexKey_X.fromJSON(object.x) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== undefined) {
            obj.id = exports.ListExampleAutoIncrementTableRequest_IndexKey_Id.toJSON(message.id);
        }
        if (message.x !== undefined) {
            obj.x = exports.ListExampleAutoIncrementTableRequest_IndexKey_X.toJSON(message.x);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleAutoIncrementTableRequest_IndexKey.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListExampleAutoIncrementTableRequest_IndexKey();
        message.id = (object.id !== undefined && object.id !== null)
            ? exports.ListExampleAutoIncrementTableRequest_IndexKey_Id.fromPartial(object.id)
            : undefined;
        message.x = (object.x !== undefined && object.x !== null)
            ? exports.ListExampleAutoIncrementTableRequest_IndexKey_X.fromPartial(object.x)
            : undefined;
        return message;
    },
};
function createBaseListExampleAutoIncrementTableRequest_IndexKey_Id() {
    return { id: undefined };
}
exports.ListExampleAutoIncrementTableRequest_IndexKey_Id = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== undefined) {
            writer.uint32(8).uint64(message.id);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleAutoIncrementTableRequest_IndexKey_Id();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.id = longToString(reader.uint64());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { id: isSet(object.id) ? globalThis.String(object.id) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== undefined) {
            obj.id = message.id;
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleAutoIncrementTableRequest_IndexKey_Id.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListExampleAutoIncrementTableRequest_IndexKey_Id();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseListExampleAutoIncrementTableRequest_IndexKey_X() {
    return { x: undefined };
}
exports.ListExampleAutoIncrementTableRequest_IndexKey_X = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.x !== undefined) {
            writer.uint32(10).string(message.x);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleAutoIncrementTableRequest_IndexKey_X();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.x = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { x: isSet(object.x) ? globalThis.String(object.x) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.x !== undefined) {
            obj.x = message.x;
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleAutoIncrementTableRequest_IndexKey_X.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListExampleAutoIncrementTableRequest_IndexKey_X();
        message.x = (_a = object.x) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseListExampleAutoIncrementTableRequest_RangeQuery() {
    return { from: undefined, to: undefined };
}
exports.ListExampleAutoIncrementTableRequest_RangeQuery = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.from !== undefined) {
            exports.ListExampleAutoIncrementTableRequest_IndexKey.encode(message.from, writer.uint32(10).fork()).ldelim();
        }
        if (message.to !== undefined) {
            exports.ListExampleAutoIncrementTableRequest_IndexKey.encode(message.to, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleAutoIncrementTableRequest_RangeQuery();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.from = exports.ListExampleAutoIncrementTableRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.to = exports.ListExampleAutoIncrementTableRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            from: isSet(object.from) ? exports.ListExampleAutoIncrementTableRequest_IndexKey.fromJSON(object.from) : undefined,
            to: isSet(object.to) ? exports.ListExampleAutoIncrementTableRequest_IndexKey.fromJSON(object.to) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.from !== undefined) {
            obj.from = exports.ListExampleAutoIncrementTableRequest_IndexKey.toJSON(message.from);
        }
        if (message.to !== undefined) {
            obj.to = exports.ListExampleAutoIncrementTableRequest_IndexKey.toJSON(message.to);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleAutoIncrementTableRequest_RangeQuery.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListExampleAutoIncrementTableRequest_RangeQuery();
        message.from = (object.from !== undefined && object.from !== null)
            ? exports.ListExampleAutoIncrementTableRequest_IndexKey.fromPartial(object.from)
            : undefined;
        message.to = (object.to !== undefined && object.to !== null)
            ? exports.ListExampleAutoIncrementTableRequest_IndexKey.fromPartial(object.to)
            : undefined;
        return message;
    },
};
function createBaseListExampleAutoIncrementTableResponse() {
    return { values: [], pagination: undefined };
}
exports.ListExampleAutoIncrementTableResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.values) {
            test_schema_1.ExampleAutoIncrementTable.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleAutoIncrementTableResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.values.push(test_schema_1.ExampleAutoIncrementTable.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            values: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.values)
                ? object.values.map((e) => test_schema_1.ExampleAutoIncrementTable.fromJSON(e))
                : [],
            pagination: isSet(object.pagination) ? pagination_1.PageResponse.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.values) === null || _a === void 0 ? void 0 : _a.length) {
            obj.values = message.values.map((e) => test_schema_1.ExampleAutoIncrementTable.toJSON(e));
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleAutoIncrementTableResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListExampleAutoIncrementTableResponse();
        message.values = ((_a = object.values) === null || _a === void 0 ? void 0 : _a.map((e) => test_schema_1.ExampleAutoIncrementTable.fromPartial(e))) || [];
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageResponse.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
function createBaseGetExampleSingletonRequest() {
    return {};
}
exports.GetExampleSingletonRequest = {
    encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleSingletonRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.GetExampleSingletonRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseGetExampleSingletonRequest();
        return message;
    },
};
function createBaseGetExampleSingletonResponse() {
    return { value: undefined };
}
exports.GetExampleSingletonResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== undefined) {
            test_schema_1.ExampleSingleton.encode(message.value, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleSingletonResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = test_schema_1.ExampleSingleton.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { value: isSet(object.value) ? test_schema_1.ExampleSingleton.fromJSON(object.value) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.value !== undefined) {
            obj.value = test_schema_1.ExampleSingleton.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleSingletonResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGetExampleSingletonResponse();
        message.value = (object.value !== undefined && object.value !== null)
            ? test_schema_1.ExampleSingleton.fromPartial(object.value)
            : undefined;
        return message;
    },
};
function createBaseGetExampleTimestampRequest() {
    return { id: "0" };
}
exports.GetExampleTimestampRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== "0") {
            writer.uint32(8).uint64(message.id);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleTimestampRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.id = longToString(reader.uint64());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { id: isSet(object.id) ? globalThis.String(object.id) : "0" };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== "0") {
            obj.id = message.id;
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleTimestampRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetExampleTimestampRequest();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : "0";
        return message;
    },
};
function createBaseGetExampleTimestampResponse() {
    return { value: undefined };
}
exports.GetExampleTimestampResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== undefined) {
            test_schema_1.ExampleTimestamp.encode(message.value, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleTimestampResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = test_schema_1.ExampleTimestamp.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { value: isSet(object.value) ? test_schema_1.ExampleTimestamp.fromJSON(object.value) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.value !== undefined) {
            obj.value = test_schema_1.ExampleTimestamp.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleTimestampResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGetExampleTimestampResponse();
        message.value = (object.value !== undefined && object.value !== null)
            ? test_schema_1.ExampleTimestamp.fromPartial(object.value)
            : undefined;
        return message;
    },
};
function createBaseListExampleTimestampRequest() {
    return { prefixQuery: undefined, rangeQuery: undefined, pagination: undefined };
}
exports.ListExampleTimestampRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.prefixQuery !== undefined) {
            exports.ListExampleTimestampRequest_IndexKey.encode(message.prefixQuery, writer.uint32(10).fork()).ldelim();
        }
        if (message.rangeQuery !== undefined) {
            exports.ListExampleTimestampRequest_RangeQuery.encode(message.rangeQuery, writer.uint32(18).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTimestampRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.prefixQuery = exports.ListExampleTimestampRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.rangeQuery = exports.ListExampleTimestampRequest_RangeQuery.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            prefixQuery: isSet(object.prefixQuery)
                ? exports.ListExampleTimestampRequest_IndexKey.fromJSON(object.prefixQuery)
                : undefined,
            rangeQuery: isSet(object.rangeQuery)
                ? exports.ListExampleTimestampRequest_RangeQuery.fromJSON(object.rangeQuery)
                : undefined,
            pagination: isSet(object.pagination) ? pagination_1.PageRequest.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.prefixQuery !== undefined) {
            obj.prefixQuery = exports.ListExampleTimestampRequest_IndexKey.toJSON(message.prefixQuery);
        }
        if (message.rangeQuery !== undefined) {
            obj.rangeQuery = exports.ListExampleTimestampRequest_RangeQuery.toJSON(message.rangeQuery);
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTimestampRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListExampleTimestampRequest();
        message.prefixQuery = (object.prefixQuery !== undefined && object.prefixQuery !== null)
            ? exports.ListExampleTimestampRequest_IndexKey.fromPartial(object.prefixQuery)
            : undefined;
        message.rangeQuery = (object.rangeQuery !== undefined && object.rangeQuery !== null)
            ? exports.ListExampleTimestampRequest_RangeQuery.fromPartial(object.rangeQuery)
            : undefined;
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageRequest.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
function createBaseListExampleTimestampRequest_IndexKey() {
    return { id: undefined, ts: undefined };
}
exports.ListExampleTimestampRequest_IndexKey = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== undefined) {
            exports.ListExampleTimestampRequest_IndexKey_Id.encode(message.id, writer.uint32(10).fork()).ldelim();
        }
        if (message.ts !== undefined) {
            exports.ListExampleTimestampRequest_IndexKey_Ts.encode(message.ts, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTimestampRequest_IndexKey();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = exports.ListExampleTimestampRequest_IndexKey_Id.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.ts = exports.ListExampleTimestampRequest_IndexKey_Ts.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? exports.ListExampleTimestampRequest_IndexKey_Id.fromJSON(object.id) : undefined,
            ts: isSet(object.ts) ? exports.ListExampleTimestampRequest_IndexKey_Ts.fromJSON(object.ts) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== undefined) {
            obj.id = exports.ListExampleTimestampRequest_IndexKey_Id.toJSON(message.id);
        }
        if (message.ts !== undefined) {
            obj.ts = exports.ListExampleTimestampRequest_IndexKey_Ts.toJSON(message.ts);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTimestampRequest_IndexKey.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListExampleTimestampRequest_IndexKey();
        message.id = (object.id !== undefined && object.id !== null)
            ? exports.ListExampleTimestampRequest_IndexKey_Id.fromPartial(object.id)
            : undefined;
        message.ts = (object.ts !== undefined && object.ts !== null)
            ? exports.ListExampleTimestampRequest_IndexKey_Ts.fromPartial(object.ts)
            : undefined;
        return message;
    },
};
function createBaseListExampleTimestampRequest_IndexKey_Id() {
    return { id: undefined };
}
exports.ListExampleTimestampRequest_IndexKey_Id = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== undefined) {
            writer.uint32(8).uint64(message.id);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTimestampRequest_IndexKey_Id();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.id = longToString(reader.uint64());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { id: isSet(object.id) ? globalThis.String(object.id) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== undefined) {
            obj.id = message.id;
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTimestampRequest_IndexKey_Id.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListExampleTimestampRequest_IndexKey_Id();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseListExampleTimestampRequest_IndexKey_Ts() {
    return { ts: undefined };
}
exports.ListExampleTimestampRequest_IndexKey_Ts = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.ts !== undefined) {
            timestamp_1.Timestamp.encode(toTimestamp(message.ts), writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTimestampRequest_IndexKey_Ts();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.ts = fromTimestamp(timestamp_1.Timestamp.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { ts: isSet(object.ts) ? fromJsonTimestamp(object.ts) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.ts !== undefined) {
            obj.ts = message.ts.toISOString();
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTimestampRequest_IndexKey_Ts.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListExampleTimestampRequest_IndexKey_Ts();
        message.ts = (_a = object.ts) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseListExampleTimestampRequest_RangeQuery() {
    return { from: undefined, to: undefined };
}
exports.ListExampleTimestampRequest_RangeQuery = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.from !== undefined) {
            exports.ListExampleTimestampRequest_IndexKey.encode(message.from, writer.uint32(10).fork()).ldelim();
        }
        if (message.to !== undefined) {
            exports.ListExampleTimestampRequest_IndexKey.encode(message.to, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTimestampRequest_RangeQuery();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.from = exports.ListExampleTimestampRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.to = exports.ListExampleTimestampRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            from: isSet(object.from) ? exports.ListExampleTimestampRequest_IndexKey.fromJSON(object.from) : undefined,
            to: isSet(object.to) ? exports.ListExampleTimestampRequest_IndexKey.fromJSON(object.to) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.from !== undefined) {
            obj.from = exports.ListExampleTimestampRequest_IndexKey.toJSON(message.from);
        }
        if (message.to !== undefined) {
            obj.to = exports.ListExampleTimestampRequest_IndexKey.toJSON(message.to);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTimestampRequest_RangeQuery.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListExampleTimestampRequest_RangeQuery();
        message.from = (object.from !== undefined && object.from !== null)
            ? exports.ListExampleTimestampRequest_IndexKey.fromPartial(object.from)
            : undefined;
        message.to = (object.to !== undefined && object.to !== null)
            ? exports.ListExampleTimestampRequest_IndexKey.fromPartial(object.to)
            : undefined;
        return message;
    },
};
function createBaseListExampleTimestampResponse() {
    return { values: [], pagination: undefined };
}
exports.ListExampleTimestampResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.values) {
            test_schema_1.ExampleTimestamp.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleTimestampResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.values.push(test_schema_1.ExampleTimestamp.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            values: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.values)
                ? object.values.map((e) => test_schema_1.ExampleTimestamp.fromJSON(e))
                : [],
            pagination: isSet(object.pagination) ? pagination_1.PageResponse.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.values) === null || _a === void 0 ? void 0 : _a.length) {
            obj.values = message.values.map((e) => test_schema_1.ExampleTimestamp.toJSON(e));
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleTimestampResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListExampleTimestampResponse();
        message.values = ((_a = object.values) === null || _a === void 0 ? void 0 : _a.map((e) => test_schema_1.ExampleTimestamp.fromPartial(e))) || [];
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageResponse.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
function createBaseGetSimpleExampleRequest() {
    return { name: "" };
}
exports.GetSimpleExampleRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetSimpleExampleRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { name: isSet(object.name) ? globalThis.String(object.name) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== "") {
            obj.name = message.name;
        }
        return obj;
    },
    create(base) {
        return exports.GetSimpleExampleRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetSimpleExampleRequest();
        message.name = (_a = object.name) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseGetSimpleExampleResponse() {
    return { value: undefined };
}
exports.GetSimpleExampleResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== undefined) {
            test_schema_1.SimpleExample.encode(message.value, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetSimpleExampleResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = test_schema_1.SimpleExample.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { value: isSet(object.value) ? test_schema_1.SimpleExample.fromJSON(object.value) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.value !== undefined) {
            obj.value = test_schema_1.SimpleExample.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.GetSimpleExampleResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGetSimpleExampleResponse();
        message.value = (object.value !== undefined && object.value !== null)
            ? test_schema_1.SimpleExample.fromPartial(object.value)
            : undefined;
        return message;
    },
};
function createBaseGetSimpleExampleByUniqueRequest() {
    return { unique: "" };
}
exports.GetSimpleExampleByUniqueRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.unique !== "") {
            writer.uint32(10).string(message.unique);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetSimpleExampleByUniqueRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.unique = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { unique: isSet(object.unique) ? globalThis.String(object.unique) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.unique !== "") {
            obj.unique = message.unique;
        }
        return obj;
    },
    create(base) {
        return exports.GetSimpleExampleByUniqueRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetSimpleExampleByUniqueRequest();
        message.unique = (_a = object.unique) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseGetSimpleExampleByUniqueResponse() {
    return { value: undefined };
}
exports.GetSimpleExampleByUniqueResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== undefined) {
            test_schema_1.SimpleExample.encode(message.value, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetSimpleExampleByUniqueResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = test_schema_1.SimpleExample.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { value: isSet(object.value) ? test_schema_1.SimpleExample.fromJSON(object.value) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.value !== undefined) {
            obj.value = test_schema_1.SimpleExample.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.GetSimpleExampleByUniqueResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGetSimpleExampleByUniqueResponse();
        message.value = (object.value !== undefined && object.value !== null)
            ? test_schema_1.SimpleExample.fromPartial(object.value)
            : undefined;
        return message;
    },
};
function createBaseListSimpleExampleRequest() {
    return { prefixQuery: undefined, rangeQuery: undefined, pagination: undefined };
}
exports.ListSimpleExampleRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.prefixQuery !== undefined) {
            exports.ListSimpleExampleRequest_IndexKey.encode(message.prefixQuery, writer.uint32(10).fork()).ldelim();
        }
        if (message.rangeQuery !== undefined) {
            exports.ListSimpleExampleRequest_RangeQuery.encode(message.rangeQuery, writer.uint32(18).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListSimpleExampleRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.prefixQuery = exports.ListSimpleExampleRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.rangeQuery = exports.ListSimpleExampleRequest_RangeQuery.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            prefixQuery: isSet(object.prefixQuery)
                ? exports.ListSimpleExampleRequest_IndexKey.fromJSON(object.prefixQuery)
                : undefined,
            rangeQuery: isSet(object.rangeQuery)
                ? exports.ListSimpleExampleRequest_RangeQuery.fromJSON(object.rangeQuery)
                : undefined,
            pagination: isSet(object.pagination) ? pagination_1.PageRequest.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.prefixQuery !== undefined) {
            obj.prefixQuery = exports.ListSimpleExampleRequest_IndexKey.toJSON(message.prefixQuery);
        }
        if (message.rangeQuery !== undefined) {
            obj.rangeQuery = exports.ListSimpleExampleRequest_RangeQuery.toJSON(message.rangeQuery);
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListSimpleExampleRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListSimpleExampleRequest();
        message.prefixQuery = (object.prefixQuery !== undefined && object.prefixQuery !== null)
            ? exports.ListSimpleExampleRequest_IndexKey.fromPartial(object.prefixQuery)
            : undefined;
        message.rangeQuery = (object.rangeQuery !== undefined && object.rangeQuery !== null)
            ? exports.ListSimpleExampleRequest_RangeQuery.fromPartial(object.rangeQuery)
            : undefined;
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageRequest.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
function createBaseListSimpleExampleRequest_IndexKey() {
    return { name: undefined, unique: undefined };
}
exports.ListSimpleExampleRequest_IndexKey = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.name !== undefined) {
            exports.ListSimpleExampleRequest_IndexKey_Name.encode(message.name, writer.uint32(10).fork()).ldelim();
        }
        if (message.unique !== undefined) {
            exports.ListSimpleExampleRequest_IndexKey_Unique.encode(message.unique, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListSimpleExampleRequest_IndexKey();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.name = exports.ListSimpleExampleRequest_IndexKey_Name.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.unique = exports.ListSimpleExampleRequest_IndexKey_Unique.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            name: isSet(object.name) ? exports.ListSimpleExampleRequest_IndexKey_Name.fromJSON(object.name) : undefined,
            unique: isSet(object.unique) ? exports.ListSimpleExampleRequest_IndexKey_Unique.fromJSON(object.unique) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== undefined) {
            obj.name = exports.ListSimpleExampleRequest_IndexKey_Name.toJSON(message.name);
        }
        if (message.unique !== undefined) {
            obj.unique = exports.ListSimpleExampleRequest_IndexKey_Unique.toJSON(message.unique);
        }
        return obj;
    },
    create(base) {
        return exports.ListSimpleExampleRequest_IndexKey.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListSimpleExampleRequest_IndexKey();
        message.name = (object.name !== undefined && object.name !== null)
            ? exports.ListSimpleExampleRequest_IndexKey_Name.fromPartial(object.name)
            : undefined;
        message.unique = (object.unique !== undefined && object.unique !== null)
            ? exports.ListSimpleExampleRequest_IndexKey_Unique.fromPartial(object.unique)
            : undefined;
        return message;
    },
};
function createBaseListSimpleExampleRequest_IndexKey_Name() {
    return { name: undefined };
}
exports.ListSimpleExampleRequest_IndexKey_Name = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.name !== undefined) {
            writer.uint32(10).string(message.name);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListSimpleExampleRequest_IndexKey_Name();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { name: isSet(object.name) ? globalThis.String(object.name) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== undefined) {
            obj.name = message.name;
        }
        return obj;
    },
    create(base) {
        return exports.ListSimpleExampleRequest_IndexKey_Name.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListSimpleExampleRequest_IndexKey_Name();
        message.name = (_a = object.name) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseListSimpleExampleRequest_IndexKey_Unique() {
    return { unique: undefined };
}
exports.ListSimpleExampleRequest_IndexKey_Unique = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.unique !== undefined) {
            writer.uint32(10).string(message.unique);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListSimpleExampleRequest_IndexKey_Unique();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.unique = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { unique: isSet(object.unique) ? globalThis.String(object.unique) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.unique !== undefined) {
            obj.unique = message.unique;
        }
        return obj;
    },
    create(base) {
        return exports.ListSimpleExampleRequest_IndexKey_Unique.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListSimpleExampleRequest_IndexKey_Unique();
        message.unique = (_a = object.unique) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseListSimpleExampleRequest_RangeQuery() {
    return { from: undefined, to: undefined };
}
exports.ListSimpleExampleRequest_RangeQuery = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.from !== undefined) {
            exports.ListSimpleExampleRequest_IndexKey.encode(message.from, writer.uint32(10).fork()).ldelim();
        }
        if (message.to !== undefined) {
            exports.ListSimpleExampleRequest_IndexKey.encode(message.to, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListSimpleExampleRequest_RangeQuery();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.from = exports.ListSimpleExampleRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.to = exports.ListSimpleExampleRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            from: isSet(object.from) ? exports.ListSimpleExampleRequest_IndexKey.fromJSON(object.from) : undefined,
            to: isSet(object.to) ? exports.ListSimpleExampleRequest_IndexKey.fromJSON(object.to) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.from !== undefined) {
            obj.from = exports.ListSimpleExampleRequest_IndexKey.toJSON(message.from);
        }
        if (message.to !== undefined) {
            obj.to = exports.ListSimpleExampleRequest_IndexKey.toJSON(message.to);
        }
        return obj;
    },
    create(base) {
        return exports.ListSimpleExampleRequest_RangeQuery.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListSimpleExampleRequest_RangeQuery();
        message.from = (object.from !== undefined && object.from !== null)
            ? exports.ListSimpleExampleRequest_IndexKey.fromPartial(object.from)
            : undefined;
        message.to = (object.to !== undefined && object.to !== null)
            ? exports.ListSimpleExampleRequest_IndexKey.fromPartial(object.to)
            : undefined;
        return message;
    },
};
function createBaseListSimpleExampleResponse() {
    return { values: [], pagination: undefined };
}
exports.ListSimpleExampleResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.values) {
            test_schema_1.SimpleExample.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListSimpleExampleResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.values.push(test_schema_1.SimpleExample.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            values: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.values) ? object.values.map((e) => test_schema_1.SimpleExample.fromJSON(e)) : [],
            pagination: isSet(object.pagination) ? pagination_1.PageResponse.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.values) === null || _a === void 0 ? void 0 : _a.length) {
            obj.values = message.values.map((e) => test_schema_1.SimpleExample.toJSON(e));
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListSimpleExampleResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListSimpleExampleResponse();
        message.values = ((_a = object.values) === null || _a === void 0 ? void 0 : _a.map((e) => test_schema_1.SimpleExample.fromPartial(e))) || [];
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageResponse.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
function createBaseGetExampleAutoIncFieldNameRequest() {
    return { foo: "0" };
}
exports.GetExampleAutoIncFieldNameRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.foo !== "0") {
            writer.uint32(8).uint64(message.foo);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleAutoIncFieldNameRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.foo = longToString(reader.uint64());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { foo: isSet(object.foo) ? globalThis.String(object.foo) : "0" };
    },
    toJSON(message) {
        const obj = {};
        if (message.foo !== "0") {
            obj.foo = message.foo;
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleAutoIncFieldNameRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetExampleAutoIncFieldNameRequest();
        message.foo = (_a = object.foo) !== null && _a !== void 0 ? _a : "0";
        return message;
    },
};
function createBaseGetExampleAutoIncFieldNameResponse() {
    return { value: undefined };
}
exports.GetExampleAutoIncFieldNameResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== undefined) {
            test_schema_1.ExampleAutoIncFieldName.encode(message.value, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExampleAutoIncFieldNameResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = test_schema_1.ExampleAutoIncFieldName.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { value: isSet(object.value) ? test_schema_1.ExampleAutoIncFieldName.fromJSON(object.value) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.value !== undefined) {
            obj.value = test_schema_1.ExampleAutoIncFieldName.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.GetExampleAutoIncFieldNameResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseGetExampleAutoIncFieldNameResponse();
        message.value = (object.value !== undefined && object.value !== null)
            ? test_schema_1.ExampleAutoIncFieldName.fromPartial(object.value)
            : undefined;
        return message;
    },
};
function createBaseListExampleAutoIncFieldNameRequest() {
    return { prefixQuery: undefined, rangeQuery: undefined, pagination: undefined };
}
exports.ListExampleAutoIncFieldNameRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.prefixQuery !== undefined) {
            exports.ListExampleAutoIncFieldNameRequest_IndexKey.encode(message.prefixQuery, writer.uint32(10).fork()).ldelim();
        }
        if (message.rangeQuery !== undefined) {
            exports.ListExampleAutoIncFieldNameRequest_RangeQuery.encode(message.rangeQuery, writer.uint32(18).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleAutoIncFieldNameRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.prefixQuery = exports.ListExampleAutoIncFieldNameRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.rangeQuery = exports.ListExampleAutoIncFieldNameRequest_RangeQuery.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            prefixQuery: isSet(object.prefixQuery)
                ? exports.ListExampleAutoIncFieldNameRequest_IndexKey.fromJSON(object.prefixQuery)
                : undefined,
            rangeQuery: isSet(object.rangeQuery)
                ? exports.ListExampleAutoIncFieldNameRequest_RangeQuery.fromJSON(object.rangeQuery)
                : undefined,
            pagination: isSet(object.pagination) ? pagination_1.PageRequest.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.prefixQuery !== undefined) {
            obj.prefixQuery = exports.ListExampleAutoIncFieldNameRequest_IndexKey.toJSON(message.prefixQuery);
        }
        if (message.rangeQuery !== undefined) {
            obj.rangeQuery = exports.ListExampleAutoIncFieldNameRequest_RangeQuery.toJSON(message.rangeQuery);
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleAutoIncFieldNameRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListExampleAutoIncFieldNameRequest();
        message.prefixQuery = (object.prefixQuery !== undefined && object.prefixQuery !== null)
            ? exports.ListExampleAutoIncFieldNameRequest_IndexKey.fromPartial(object.prefixQuery)
            : undefined;
        message.rangeQuery = (object.rangeQuery !== undefined && object.rangeQuery !== null)
            ? exports.ListExampleAutoIncFieldNameRequest_RangeQuery.fromPartial(object.rangeQuery)
            : undefined;
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageRequest.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
function createBaseListExampleAutoIncFieldNameRequest_IndexKey() {
    return { foo: undefined };
}
exports.ListExampleAutoIncFieldNameRequest_IndexKey = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.foo !== undefined) {
            exports.ListExampleAutoIncFieldNameRequest_IndexKey_Foo.encode(message.foo, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleAutoIncFieldNameRequest_IndexKey();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.foo = exports.ListExampleAutoIncFieldNameRequest_IndexKey_Foo.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            foo: isSet(object.foo) ? exports.ListExampleAutoIncFieldNameRequest_IndexKey_Foo.fromJSON(object.foo) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.foo !== undefined) {
            obj.foo = exports.ListExampleAutoIncFieldNameRequest_IndexKey_Foo.toJSON(message.foo);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleAutoIncFieldNameRequest_IndexKey.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListExampleAutoIncFieldNameRequest_IndexKey();
        message.foo = (object.foo !== undefined && object.foo !== null)
            ? exports.ListExampleAutoIncFieldNameRequest_IndexKey_Foo.fromPartial(object.foo)
            : undefined;
        return message;
    },
};
function createBaseListExampleAutoIncFieldNameRequest_IndexKey_Foo() {
    return { foo: undefined };
}
exports.ListExampleAutoIncFieldNameRequest_IndexKey_Foo = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.foo !== undefined) {
            writer.uint32(8).uint64(message.foo);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleAutoIncFieldNameRequest_IndexKey_Foo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.foo = longToString(reader.uint64());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { foo: isSet(object.foo) ? globalThis.String(object.foo) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.foo !== undefined) {
            obj.foo = message.foo;
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleAutoIncFieldNameRequest_IndexKey_Foo.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListExampleAutoIncFieldNameRequest_IndexKey_Foo();
        message.foo = (_a = object.foo) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseListExampleAutoIncFieldNameRequest_RangeQuery() {
    return { from: undefined, to: undefined };
}
exports.ListExampleAutoIncFieldNameRequest_RangeQuery = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.from !== undefined) {
            exports.ListExampleAutoIncFieldNameRequest_IndexKey.encode(message.from, writer.uint32(10).fork()).ldelim();
        }
        if (message.to !== undefined) {
            exports.ListExampleAutoIncFieldNameRequest_IndexKey.encode(message.to, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleAutoIncFieldNameRequest_RangeQuery();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.from = exports.ListExampleAutoIncFieldNameRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.to = exports.ListExampleAutoIncFieldNameRequest_IndexKey.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            from: isSet(object.from) ? exports.ListExampleAutoIncFieldNameRequest_IndexKey.fromJSON(object.from) : undefined,
            to: isSet(object.to) ? exports.ListExampleAutoIncFieldNameRequest_IndexKey.fromJSON(object.to) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.from !== undefined) {
            obj.from = exports.ListExampleAutoIncFieldNameRequest_IndexKey.toJSON(message.from);
        }
        if (message.to !== undefined) {
            obj.to = exports.ListExampleAutoIncFieldNameRequest_IndexKey.toJSON(message.to);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleAutoIncFieldNameRequest_RangeQuery.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseListExampleAutoIncFieldNameRequest_RangeQuery();
        message.from = (object.from !== undefined && object.from !== null)
            ? exports.ListExampleAutoIncFieldNameRequest_IndexKey.fromPartial(object.from)
            : undefined;
        message.to = (object.to !== undefined && object.to !== null)
            ? exports.ListExampleAutoIncFieldNameRequest_IndexKey.fromPartial(object.to)
            : undefined;
        return message;
    },
};
function createBaseListExampleAutoIncFieldNameResponse() {
    return { values: [], pagination: undefined };
}
exports.ListExampleAutoIncFieldNameResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.values) {
            test_schema_1.ExampleAutoIncFieldName.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListExampleAutoIncFieldNameResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.values.push(test_schema_1.ExampleAutoIncFieldName.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            values: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.values)
                ? object.values.map((e) => test_schema_1.ExampleAutoIncFieldName.fromJSON(e))
                : [],
            pagination: isSet(object.pagination) ? pagination_1.PageResponse.fromJSON(object.pagination) : undefined,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.values) === null || _a === void 0 ? void 0 : _a.length) {
            obj.values = message.values.map((e) => test_schema_1.ExampleAutoIncFieldName.toJSON(e));
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.ListExampleAutoIncFieldNameResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListExampleAutoIncFieldNameResponse();
        message.values = ((_a = object.values) === null || _a === void 0 ? void 0 : _a.map((e) => test_schema_1.ExampleAutoIncFieldName.fromPartial(e))) || [];
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageResponse.fromPartial(object.pagination)
            : undefined;
        return message;
    },
};
class TestSchemaQueryServiceClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.GetExampleTable = this.GetExampleTable.bind(this);
        this.GetExampleTableByU64Str = this.GetExampleTableByU64Str.bind(this);
        this.ListExampleTable = this.ListExampleTable.bind(this);
        this.GetExampleAutoIncrementTable = this.GetExampleAutoIncrementTable.bind(this);
        this.GetExampleAutoIncrementTableByX = this.GetExampleAutoIncrementTableByX.bind(this);
        this.ListExampleAutoIncrementTable = this.ListExampleAutoIncrementTable.bind(this);
        this.GetExampleSingleton = this.GetExampleSingleton.bind(this);
        this.GetExampleTimestamp = this.GetExampleTimestamp.bind(this);
        this.ListExampleTimestamp = this.ListExampleTimestamp.bind(this);
        this.GetSimpleExample = this.GetSimpleExample.bind(this);
        this.GetSimpleExampleByUnique = this.GetSimpleExampleByUnique.bind(this);
        this.ListSimpleExample = this.ListSimpleExample.bind(this);
        this.GetExampleAutoIncFieldName = this.GetExampleAutoIncFieldName.bind(this);
        this.ListExampleAutoIncFieldName = this.ListExampleAutoIncFieldName.bind(this);
    }
    GetExampleTable(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceGetExampleTableDesc, exports.GetExampleTableRequest.fromPartial(request), metadata);
    }
    GetExampleTableByU64Str(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceGetExampleTableByU64StrDesc, exports.GetExampleTableByU64StrRequest.fromPartial(request), metadata);
    }
    ListExampleTable(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceListExampleTableDesc, exports.ListExampleTableRequest.fromPartial(request), metadata);
    }
    GetExampleAutoIncrementTable(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceGetExampleAutoIncrementTableDesc, exports.GetExampleAutoIncrementTableRequest.fromPartial(request), metadata);
    }
    GetExampleAutoIncrementTableByX(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceGetExampleAutoIncrementTableByXDesc, exports.GetExampleAutoIncrementTableByXRequest.fromPartial(request), metadata);
    }
    ListExampleAutoIncrementTable(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceListExampleAutoIncrementTableDesc, exports.ListExampleAutoIncrementTableRequest.fromPartial(request), metadata);
    }
    GetExampleSingleton(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceGetExampleSingletonDesc, exports.GetExampleSingletonRequest.fromPartial(request), metadata);
    }
    GetExampleTimestamp(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceGetExampleTimestampDesc, exports.GetExampleTimestampRequest.fromPartial(request), metadata);
    }
    ListExampleTimestamp(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceListExampleTimestampDesc, exports.ListExampleTimestampRequest.fromPartial(request), metadata);
    }
    GetSimpleExample(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceGetSimpleExampleDesc, exports.GetSimpleExampleRequest.fromPartial(request), metadata);
    }
    GetSimpleExampleByUnique(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceGetSimpleExampleByUniqueDesc, exports.GetSimpleExampleByUniqueRequest.fromPartial(request), metadata);
    }
    ListSimpleExample(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceListSimpleExampleDesc, exports.ListSimpleExampleRequest.fromPartial(request), metadata);
    }
    GetExampleAutoIncFieldName(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceGetExampleAutoIncFieldNameDesc, exports.GetExampleAutoIncFieldNameRequest.fromPartial(request), metadata);
    }
    ListExampleAutoIncFieldName(request, metadata) {
        return this.rpc.unary(exports.TestSchemaQueryServiceListExampleAutoIncFieldNameDesc, exports.ListExampleAutoIncFieldNameRequest.fromPartial(request), metadata);
    }
}
exports.TestSchemaQueryServiceClientImpl = TestSchemaQueryServiceClientImpl;
exports.TestSchemaQueryServiceDesc = { serviceName: "testpb.TestSchemaQueryService" };
exports.TestSchemaQueryServiceGetExampleTableDesc = {
    methodName: "GetExampleTable",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetExampleTableRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetExampleTableResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceGetExampleTableByU64StrDesc = {
    methodName: "GetExampleTableByU64Str",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetExampleTableByU64StrRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetExampleTableByU64StrResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceListExampleTableDesc = {
    methodName: "ListExampleTable",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.ListExampleTableRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.ListExampleTableResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceGetExampleAutoIncrementTableDesc = {
    methodName: "GetExampleAutoIncrementTable",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetExampleAutoIncrementTableRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetExampleAutoIncrementTableResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceGetExampleAutoIncrementTableByXDesc = {
    methodName: "GetExampleAutoIncrementTableByX",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetExampleAutoIncrementTableByXRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetExampleAutoIncrementTableByXResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceListExampleAutoIncrementTableDesc = {
    methodName: "ListExampleAutoIncrementTable",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.ListExampleAutoIncrementTableRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.ListExampleAutoIncrementTableResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceGetExampleSingletonDesc = {
    methodName: "GetExampleSingleton",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetExampleSingletonRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetExampleSingletonResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceGetExampleTimestampDesc = {
    methodName: "GetExampleTimestamp",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetExampleTimestampRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetExampleTimestampResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceListExampleTimestampDesc = {
    methodName: "ListExampleTimestamp",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.ListExampleTimestampRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.ListExampleTimestampResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceGetSimpleExampleDesc = {
    methodName: "GetSimpleExample",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetSimpleExampleRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetSimpleExampleResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceGetSimpleExampleByUniqueDesc = {
    methodName: "GetSimpleExampleByUnique",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetSimpleExampleByUniqueRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetSimpleExampleByUniqueResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceListSimpleExampleDesc = {
    methodName: "ListSimpleExample",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.ListSimpleExampleRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.ListSimpleExampleResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceGetExampleAutoIncFieldNameDesc = {
    methodName: "GetExampleAutoIncFieldName",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.GetExampleAutoIncFieldNameRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.GetExampleAutoIncFieldNameResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
exports.TestSchemaQueryServiceListExampleAutoIncFieldNameDesc = {
    methodName: "ListExampleAutoIncFieldName",
    service: exports.TestSchemaQueryServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.ListExampleAutoIncFieldNameRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.ListExampleAutoIncFieldNameResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
class GrpcWebImpl {
    constructor(host, options) {
        this.host = host;
        this.options = options;
    }
    unary(methodDesc, _request, metadata) {
        var _a;
        const request = Object.assign(Object.assign({}, _request), methodDesc.requestType);
        const maybeCombinedMetadata = metadata && this.options.metadata
            ? new browser_headers_1.BrowserHeaders(Object.assign(Object.assign({}, (_a = this.options) === null || _a === void 0 ? void 0 : _a.metadata.headersMap), metadata === null || metadata === void 0 ? void 0 : metadata.headersMap))
            : metadata !== null && metadata !== void 0 ? metadata : this.options.metadata;
        return new Promise((resolve, reject) => {
            var _a;
            grpc_web_1.grpc.unary(methodDesc, Object.assign(Object.assign({ request, host: this.host, metadata: maybeCombinedMetadata !== null && maybeCombinedMetadata !== void 0 ? maybeCombinedMetadata : {} }, (this.options.transport !== undefined ? { transport: this.options.transport } : {})), { debug: (_a = this.options.debug) !== null && _a !== void 0 ? _a : false, onEnd: function (response) {
                    if (response.status === grpc_web_1.grpc.Code.OK) {
                        resolve(response.message.toObject());
                    }
                    else {
                        const err = new GrpcWebError(response.statusMessage, response.status, response.trailers);
                        reject(err);
                    }
                } }));
        });
    }
}
exports.GrpcWebImpl = GrpcWebImpl;
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function toTimestamp(date) {
    const seconds = Math.trunc(date.getTime() / 1000).toString();
    const nanos = (date.getTime() % 1000) * 1000000;
    return { seconds, nanos };
}
function fromTimestamp(t) {
    let millis = (globalThis.Number(t.seconds) || 0) * 1000;
    millis += (t.nanos || 0) / 1000000;
    return new globalThis.Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof globalThis.Date) {
        return o;
    }
    else if (typeof o === "string") {
        return new globalThis.Date(o);
    }
    else {
        return fromTimestamp(timestamp_1.Timestamp.fromJSON(o));
    }
}
function longToString(long) {
    return long.toString();
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
class GrpcWebError extends globalThis.Error {
    constructor(message, code, metadata) {
        super(message);
        this.code = code;
        this.metadata = metadata;
    }
}
exports.GrpcWebError = GrpcWebError;
