// @injectivelabs/indexer-proto-ts/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InjectiveCampaignRpc = exports.InjectiveTradingRpc = exports.InjectiveSpotExchangeRpc = exports.InjectivePortfolioRpc = exports.InjectiveOracleRpc = exports.InjectiveMetaRpc = exports.InjectiveInsuranceRpc = exports.InjectiveExplorerRpc = exports.InjectiveExchangeRpc = exports.InjectiveDerivativeExchangeRpc = exports.InjectiveAuctionRpc = exports.InjectiveArchiverRPC = exports.InjectiveAccountRpc = exports.EventProviderApi = void 0;
exports.EventProviderApi = __importStar(require("./event_provider_api"));
exports.InjectiveAccountRpc = __importStar(require("./injective_accounts_rpc"));
exports.InjectiveArchiverRPC = __importStar(require("./injective_archiver_rpc"));
;
exports.InjectiveAuctionRpc = __importStar(require("./injective_auction_rpc"));
exports.InjectiveDerivativeExchangeRpc = __importStar(require("./injective_derivative_exchange_rpc"));
exports.InjectiveExchangeRpc = __importStar(require("./injective_exchange_rpc"));
exports.InjectiveExplorerRpc = __importStar(require("./injective_explorer_rpc"));
exports.InjectiveInsuranceRpc = __importStar(require("./injective_insurance_rpc"));
exports.InjectiveMetaRpc = __importStar(require("./injective_meta_rpc"));
exports.InjectiveOracleRpc = __importStar(require("./injective_oracle_rpc"));
exports.InjectivePortfolioRpc = __importStar(require("./injective_portfolio_rpc"));
exports.InjectiveSpotExchangeRpc = __importStar(require("./injective_spot_exchange_rpc"));
exports.InjectiveTradingRpc = __importStar(require("./injective_trading_rpc"));
exports.InjectiveCampaignRpc = __importStar(require("./injective_campaign_rpc"));
