// @wormhole-foundation/sdk-connect/dist/esm/routes/request.js
import { amount } from "@wormhole-foundation/sdk-base";
import { getTokenDetails } from "./token.js";
export class RouteTransferRequest {
    source;
    destination;
    fromChain;
    toChain;
    constructor(fromChain, toChain, source, destination) {
        this.fromChain = fromChain;
        this.toChain = toChain;
        this.source = source;
        this.destination = destination;
    }
    parseAmount(amt) {
        return amount.parse(amt, this.source.decimals);
    }
    amountFromBaseUnits(amt) {
        return amount.fromBaseUnits(amt, this.source.decimals);
    }
    async displayQuote(quote, params, details) {
        let dq = {
            success: true,
            sourceToken: {
                token: quote.sourceToken.token,
                amount: amount.fromBaseUnits(quote.sourceToken.amount, this.source.decimals),
            },
            destinationToken: {
                token: quote.destinationToken.token,
                amount: amount.fromBaseUnits(quote.destinationToken.amount, this.destination.decimals),
            },
            params,
        };
        if (quote.relayFee) {
            dq.relayFee = {
                token: quote.relayFee.token,
                amount: amount.fromBaseUnits(quote.relayFee.amount, this.source.decimals),
            };
        }
        if (quote.destinationNativeGas) {
            const dstDecimals = await this.toChain.getDecimals("native");
            dq.destinationNativeGas = amount.fromBaseUnits(quote.destinationNativeGas, dstDecimals);
        }
        if (details) {
            dq.details = details;
        }
        return dq;
    }
    static async create(wh, params, fromChain, toChain) {
        fromChain = fromChain ?? wh.getChain(params.source.chain);
        toChain = toChain ?? wh.getChain(params.destination.chain);
        const sourceDetails = await getTokenDetails(fromChain, params.source);
        const destDetails = await getTokenDetails(toChain, params.destination);
        const rtr = new RouteTransferRequest(fromChain, toChain, sourceDetails, destDetails);
        return rtr;
    }
}
