// @injectivelabs/core-proto-ts/cjs/packetforward/v1/genesis.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InFlightPacket = exports.Params = exports.GenesisState_InFlightPacketsEntry = exports.GenesisState = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
function createBaseGenesisState() {
    return { params: undefined, inFlightPackets: {} };
}
exports.GenesisState = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.params !== undefined) {
            exports.Params.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        Object.entries(message.inFlightPackets).forEach(([key, value]) => {
            exports.GenesisState_InFlightPacketsEntry.encode({ key: key, value }, writer.uint32(18).fork()).ldelim();
        });
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.params = exports.Params.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    const entry2 = exports.GenesisState_InFlightPacketsEntry.decode(reader, reader.uint32());
                    if (entry2.value !== undefined) {
                        message.inFlightPackets[entry2.key] = entry2.value;
                    }
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            params: isSet(object.params) ? exports.Params.fromJSON(object.params) : undefined,
            inFlightPackets: isObject(object.inFlightPackets)
                ? Object.entries(object.inFlightPackets).reduce((acc, [key, value]) => {
                    acc[key] = exports.InFlightPacket.fromJSON(value);
                    return acc;
                }, {})
                : {},
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.params !== undefined) {
            obj.params = exports.Params.toJSON(message.params);
        }
        if (message.inFlightPackets) {
            const entries = Object.entries(message.inFlightPackets);
            if (entries.length > 0) {
                obj.inFlightPackets = {};
                entries.forEach(([k, v]) => {
                    obj.inFlightPackets[k] = exports.InFlightPacket.toJSON(v);
                });
            }
        }
        return obj;
    },
    create(base) {
        return exports.GenesisState.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGenesisState();
        message.params = (object.params !== undefined && object.params !== null)
            ? exports.Params.fromPartial(object.params)
            : undefined;
        message.inFlightPackets = Object.entries((_a = object.inFlightPackets) !== null && _a !== void 0 ? _a : {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[key] = exports.InFlightPacket.fromPartial(value);
            }
            return acc;
        }, {});
        return message;
    },
};
function createBaseGenesisState_InFlightPacketsEntry() {
    return { key: "", value: undefined };
}
exports.GenesisState_InFlightPacketsEntry = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== undefined) {
            exports.InFlightPacket.encode(message.value, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState_InFlightPacketsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.value = exports.InFlightPacket.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object.value) ? exports.InFlightPacket.fromJSON(object.value) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.value !== undefined) {
            obj.value = exports.InFlightPacket.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.GenesisState_InFlightPacketsEntry.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGenesisState_InFlightPacketsEntry();
        message.key = (_a = object.key) !== null && _a !== void 0 ? _a : "";
        message.value = (object.value !== undefined && object.value !== null)
            ? exports.InFlightPacket.fromPartial(object.value)
            : undefined;
        return message;
    },
};
function createBaseParams() {
    return { feePercentage: "" };
}
exports.Params = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.feePercentage !== "") {
            writer.uint32(10).string(message.feePercentage);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.feePercentage = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { feePercentage: isSet(object.feePercentage) ? globalThis.String(object.feePercentage) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.feePercentage !== "") {
            obj.feePercentage = message.feePercentage;
        }
        return obj;
    },
    create(base) {
        return exports.Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseParams();
        message.feePercentage = (_a = object.feePercentage) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseInFlightPacket() {
    return {
        originalSenderAddress: "",
        refundChannelId: "",
        refundPortId: "",
        packetSrcChannelId: "",
        packetSrcPortId: "",
        packetTimeoutTimestamp: "0",
        packetTimeoutHeight: "",
        packetData: new Uint8Array(0),
        refundSequence: "0",
        retriesRemaining: 0,
        timeout: "0",
        nonrefundable: false,
    };
}
exports.InFlightPacket = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.originalSenderAddress !== "") {
            writer.uint32(10).string(message.originalSenderAddress);
        }
        if (message.refundChannelId !== "") {
            writer.uint32(18).string(message.refundChannelId);
        }
        if (message.refundPortId !== "") {
            writer.uint32(26).string(message.refundPortId);
        }
        if (message.packetSrcChannelId !== "") {
            writer.uint32(34).string(message.packetSrcChannelId);
        }
        if (message.packetSrcPortId !== "") {
            writer.uint32(42).string(message.packetSrcPortId);
        }
        if (message.packetTimeoutTimestamp !== "0") {
            writer.uint32(48).uint64(message.packetTimeoutTimestamp);
        }
        if (message.packetTimeoutHeight !== "") {
            writer.uint32(58).string(message.packetTimeoutHeight);
        }
        if (message.packetData.length !== 0) {
            writer.uint32(66).bytes(message.packetData);
        }
        if (message.refundSequence !== "0") {
            writer.uint32(72).uint64(message.refundSequence);
        }
        if (message.retriesRemaining !== 0) {
            writer.uint32(80).int32(message.retriesRemaining);
        }
        if (message.timeout !== "0") {
            writer.uint32(88).uint64(message.timeout);
        }
        if (message.nonrefundable === true) {
            writer.uint32(96).bool(message.nonrefundable);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInFlightPacket();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.originalSenderAddress = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.refundChannelId = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.refundPortId = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.packetSrcChannelId = reader.string();
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.packetSrcPortId = reader.string();
                    continue;
                case 6:
                    if (tag !== 48) {
                        break;
                    }
                    message.packetTimeoutTimestamp = longToString(reader.uint64());
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.packetTimeoutHeight = reader.string();
                    continue;
                case 8:
                    if (tag !== 66) {
                        break;
                    }
                    message.packetData = reader.bytes();
                    continue;
                case 9:
                    if (tag !== 72) {
                        break;
                    }
                    message.refundSequence = longToString(reader.uint64());
                    continue;
                case 10:
                    if (tag !== 80) {
                        break;
                    }
                    message.retriesRemaining = reader.int32();
                    continue;
                case 11:
                    if (tag !== 88) {
                        break;
                    }
                    message.timeout = longToString(reader.uint64());
                    continue;
                case 12:
                    if (tag !== 96) {
                        break;
                    }
                    message.nonrefundable = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            originalSenderAddress: isSet(object.originalSenderAddress) ? globalThis.String(object.originalSenderAddress) : "",
            refundChannelId: isSet(object.refundChannelId) ? globalThis.String(object.refundChannelId) : "",
            refundPortId: isSet(object.refundPortId) ? globalThis.String(object.refundPortId) : "",
            packetSrcChannelId: isSet(object.packetSrcChannelId) ? globalThis.String(object.packetSrcChannelId) : "",
            packetSrcPortId: isSet(object.packetSrcPortId) ? globalThis.String(object.packetSrcPortId) : "",
            packetTimeoutTimestamp: isSet(object.packetTimeoutTimestamp)
                ? globalThis.String(object.packetTimeoutTimestamp)
                : "0",
            packetTimeoutHeight: isSet(object.packetTimeoutHeight) ? globalThis.String(object.packetTimeoutHeight) : "",
            packetData: isSet(object.packetData) ? bytesFromBase64(object.packetData) : new Uint8Array(0),
            refundSequence: isSet(object.refundSequence) ? globalThis.String(object.refundSequence) : "0",
            retriesRemaining: isSet(object.retriesRemaining) ? globalThis.Number(object.retriesRemaining) : 0,
            timeout: isSet(object.timeout) ? globalThis.String(object.timeout) : "0",
            nonrefundable: isSet(object.nonrefundable) ? globalThis.Boolean(object.nonrefundable) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.originalSenderAddress !== "") {
            obj.originalSenderAddress = message.originalSenderAddress;
        }
        if (message.refundChannelId !== "") {
            obj.refundChannelId = message.refundChannelId;
        }
        if (message.refundPortId !== "") {
            obj.refundPortId = message.refundPortId;
        }
        if (message.packetSrcChannelId !== "") {
            obj.packetSrcChannelId = message.packetSrcChannelId;
        }
        if (message.packetSrcPortId !== "") {
            obj.packetSrcPortId = message.packetSrcPortId;
        }
        if (message.packetTimeoutTimestamp !== "0") {
            obj.packetTimeoutTimestamp = message.packetTimeoutTimestamp;
        }
        if (message.packetTimeoutHeight !== "") {
            obj.packetTimeoutHeight = message.packetTimeoutHeight;
        }
        if (message.packetData.length !== 0) {
            obj.packetData = base64FromBytes(message.packetData);
        }
        if (message.refundSequence !== "0") {
            obj.refundSequence = message.refundSequence;
        }
        if (message.retriesRemaining !== 0) {
            obj.retriesRemaining = Math.round(message.retriesRemaining);
        }
        if (message.timeout !== "0") {
            obj.timeout = message.timeout;
        }
        if (message.nonrefundable === true) {
            obj.nonrefundable = message.nonrefundable;
        }
        return obj;
    },
    create(base) {
        return exports.InFlightPacket.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
        const message = createBaseInFlightPacket();
        message.originalSenderAddress = (_a = object.originalSenderAddress) !== null && _a !== void 0 ? _a : "";
        message.refundChannelId = (_b = object.refundChannelId) !== null && _b !== void 0 ? _b : "";
        message.refundPortId = (_c = object.refundPortId) !== null && _c !== void 0 ? _c : "";
        message.packetSrcChannelId = (_d = object.packetSrcChannelId) !== null && _d !== void 0 ? _d : "";
        message.packetSrcPortId = (_e = object.packetSrcPortId) !== null && _e !== void 0 ? _e : "";
        message.packetTimeoutTimestamp = (_f = object.packetTimeoutTimestamp) !== null && _f !== void 0 ? _f : "0";
        message.packetTimeoutHeight = (_g = object.packetTimeoutHeight) !== null && _g !== void 0 ? _g : "";
        message.packetData = (_h = object.packetData) !== null && _h !== void 0 ? _h : new Uint8Array(0);
        message.refundSequence = (_j = object.refundSequence) !== null && _j !== void 0 ? _j : "0";
        message.retriesRemaining = (_k = object.retriesRemaining) !== null && _k !== void 0 ? _k : 0;
        message.timeout = (_l = object.timeout) !== null && _l !== void 0 ? _l : "0";
        message.nonrefundable = (_m = object.nonrefundable) !== null && _m !== void 0 ? _m : false;
        return message;
    },
};
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function longToString(long) {
    return long.toString();
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
