// @injectivelabs/core-proto-ts/cjs/injective/stream/v1beta1/query.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcWebError = exports.GrpcWebImpl = exports.StreamStreamDesc = exports.StreamDesc = exports.StreamClientImpl = exports.OraclePriceFilter = exports.SubaccountDepositsFilter = exports.BankBalancesFilter = exports.OrderbookFilter = exports.OrdersFilter = exports.PositionsFilter = exports.TradesFilter = exports.DerivativeTrade = exports.SpotTrade = exports.OraclePrice = exports.Position = exports.DerivativeOrder = exports.DerivativeOrderUpdate = exports.SpotOrder = exports.SpotOrderUpdate = exports.SubaccountDeposit = exports.SubaccountDeposits = exports.BankBalance = exports.Orderbook = exports.OrderbookUpdate = exports.StreamResponse = exports.StreamRequest = exports.orderUpdateStatusToJSON = exports.orderUpdateStatusFromJSON = exports.OrderUpdateStatus = void 0;
/* eslint-disable */
const grpc_web_1 = require("@injectivelabs/grpc-web");
const browser_headers_1 = require("browser-headers");
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const coin_1 = require("../../../cosmos/base/v1beta1/coin");
const exchange_1 = require("../../exchange/v1beta1/exchange");
var OrderUpdateStatus;
(function (OrderUpdateStatus) {
    OrderUpdateStatus[OrderUpdateStatus["Unspecified"] = 0] = "Unspecified";
    OrderUpdateStatus[OrderUpdateStatus["Booked"] = 1] = "Booked";
    OrderUpdateStatus[OrderUpdateStatus["Matched"] = 2] = "Matched";
    OrderUpdateStatus[OrderUpdateStatus["Cancelled"] = 3] = "Cancelled";
    OrderUpdateStatus[OrderUpdateStatus["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(OrderUpdateStatus = exports.OrderUpdateStatus || (exports.OrderUpdateStatus = {}));
function orderUpdateStatusFromJSON(object) {
    switch (object) {
        case 0:
        case "Unspecified":
            return OrderUpdateStatus.Unspecified;
        case 1:
        case "Booked":
            return OrderUpdateStatus.Booked;
        case 2:
        case "Matched":
            return OrderUpdateStatus.Matched;
        case 3:
        case "Cancelled":
            return OrderUpdateStatus.Cancelled;
        case -1:
        case "UNRECOGNIZED":
        default:
            return OrderUpdateStatus.UNRECOGNIZED;
    }
}
exports.orderUpdateStatusFromJSON = orderUpdateStatusFromJSON;
function orderUpdateStatusToJSON(object) {
    switch (object) {
        case OrderUpdateStatus.Unspecified:
            return "Unspecified";
        case OrderUpdateStatus.Booked:
            return "Booked";
        case OrderUpdateStatus.Matched:
            return "Matched";
        case OrderUpdateStatus.Cancelled:
            return "Cancelled";
        case OrderUpdateStatus.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
exports.orderUpdateStatusToJSON = orderUpdateStatusToJSON;
function createBaseStreamRequest() {
    return {
        bankBalancesFilter: undefined,
        subaccountDepositsFilter: undefined,
        spotTradesFilter: undefined,
        derivativeTradesFilter: undefined,
        spotOrdersFilter: undefined,
        derivativeOrdersFilter: undefined,
        spotOrderbooksFilter: undefined,
        derivativeOrderbooksFilter: undefined,
        positionsFilter: undefined,
        oraclePriceFilter: undefined,
    };
}
exports.StreamRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.bankBalancesFilter !== undefined) {
            exports.BankBalancesFilter.encode(message.bankBalancesFilter, writer.uint32(10).fork()).ldelim();
        }
        if (message.subaccountDepositsFilter !== undefined) {
            exports.SubaccountDepositsFilter.encode(message.subaccountDepositsFilter, writer.uint32(18).fork()).ldelim();
        }
        if (message.spotTradesFilter !== undefined) {
            exports.TradesFilter.encode(message.spotTradesFilter, writer.uint32(26).fork()).ldelim();
        }
        if (message.derivativeTradesFilter !== undefined) {
            exports.TradesFilter.encode(message.derivativeTradesFilter, writer.uint32(34).fork()).ldelim();
        }
        if (message.spotOrdersFilter !== undefined) {
            exports.OrdersFilter.encode(message.spotOrdersFilter, writer.uint32(42).fork()).ldelim();
        }
        if (message.derivativeOrdersFilter !== undefined) {
            exports.OrdersFilter.encode(message.derivativeOrdersFilter, writer.uint32(50).fork()).ldelim();
        }
        if (message.spotOrderbooksFilter !== undefined) {
            exports.OrderbookFilter.encode(message.spotOrderbooksFilter, writer.uint32(58).fork()).ldelim();
        }
        if (message.derivativeOrderbooksFilter !== undefined) {
            exports.OrderbookFilter.encode(message.derivativeOrderbooksFilter, writer.uint32(66).fork()).ldelim();
        }
        if (message.positionsFilter !== undefined) {
            exports.PositionsFilter.encode(message.positionsFilter, writer.uint32(74).fork()).ldelim();
        }
        if (message.oraclePriceFilter !== undefined) {
            exports.OraclePriceFilter.encode(message.oraclePriceFilter, writer.uint32(82).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStreamRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.bankBalancesFilter = exports.BankBalancesFilter.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.subaccountDepositsFilter = exports.SubaccountDepositsFilter.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.spotTradesFilter = exports.TradesFilter.decode(reader, reader.uint32());
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.derivativeTradesFilter = exports.TradesFilter.decode(reader, reader.uint32());
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.spotOrdersFilter = exports.OrdersFilter.decode(reader, reader.uint32());
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.derivativeOrdersFilter = exports.OrdersFilter.decode(reader, reader.uint32());
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.spotOrderbooksFilter = exports.OrderbookFilter.decode(reader, reader.uint32());
                    continue;
                case 8:
                    if (tag !== 66) {
                        break;
                    }
                    message.derivativeOrderbooksFilter = exports.OrderbookFilter.decode(reader, reader.uint32());
                    continue;
                case 9:
                    if (tag !== 74) {
                        break;
                    }
                    message.positionsFilter = exports.PositionsFilter.decode(reader, reader.uint32());
                    continue;
                case 10:
                    if (tag !== 82) {
                        break;
                    }
                    message.oraclePriceFilter = exports.OraclePriceFilter.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            bankBalancesFilter: isSet(object.bankBalancesFilter)
                ? exports.BankBalancesFilter.fromJSON(object.bankBalancesFilter)
                : undefined,
            subaccountDepositsFilter: isSet(object.subaccountDepositsFilter)
                ? exports.SubaccountDepositsFilter.fromJSON(object.subaccountDepositsFilter)
                : undefined,
            spotTradesFilter: isSet(object.spotTradesFilter) ? exports.TradesFilter.fromJSON(object.spotTradesFilter) : undefined,
            derivativeTradesFilter: isSet(object.derivativeTradesFilter)
                ? exports.TradesFilter.fromJSON(object.derivativeTradesFilter)
                : undefined,
            spotOrdersFilter: isSet(object.spotOrdersFilter) ? exports.OrdersFilter.fromJSON(object.spotOrdersFilter) : undefined,
            derivativeOrdersFilter: isSet(object.derivativeOrdersFilter)
                ? exports.OrdersFilter.fromJSON(object.derivativeOrdersFilter)
                : undefined,
            spotOrderbooksFilter: isSet(object.spotOrderbooksFilter)
                ? exports.OrderbookFilter.fromJSON(object.spotOrderbooksFilter)
                : undefined,
            derivativeOrderbooksFilter: isSet(object.derivativeOrderbooksFilter)
                ? exports.OrderbookFilter.fromJSON(object.derivativeOrderbooksFilter)
                : undefined,
            positionsFilter: isSet(object.positionsFilter) ? exports.PositionsFilter.fromJSON(object.positionsFilter) : undefined,
            oraclePriceFilter: isSet(object.oraclePriceFilter)
                ? exports.OraclePriceFilter.fromJSON(object.oraclePriceFilter)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.bankBalancesFilter !== undefined) {
            obj.bankBalancesFilter = exports.BankBalancesFilter.toJSON(message.bankBalancesFilter);
        }
        if (message.subaccountDepositsFilter !== undefined) {
            obj.subaccountDepositsFilter = exports.SubaccountDepositsFilter.toJSON(message.subaccountDepositsFilter);
        }
        if (message.spotTradesFilter !== undefined) {
            obj.spotTradesFilter = exports.TradesFilter.toJSON(message.spotTradesFilter);
        }
        if (message.derivativeTradesFilter !== undefined) {
            obj.derivativeTradesFilter = exports.TradesFilter.toJSON(message.derivativeTradesFilter);
        }
        if (message.spotOrdersFilter !== undefined) {
            obj.spotOrdersFilter = exports.OrdersFilter.toJSON(message.spotOrdersFilter);
        }
        if (message.derivativeOrdersFilter !== undefined) {
            obj.derivativeOrdersFilter = exports.OrdersFilter.toJSON(message.derivativeOrdersFilter);
        }
        if (message.spotOrderbooksFilter !== undefined) {
            obj.spotOrderbooksFilter = exports.OrderbookFilter.toJSON(message.spotOrderbooksFilter);
        }
        if (message.derivativeOrderbooksFilter !== undefined) {
            obj.derivativeOrderbooksFilter = exports.OrderbookFilter.toJSON(message.derivativeOrderbooksFilter);
        }
        if (message.positionsFilter !== undefined) {
            obj.positionsFilter = exports.PositionsFilter.toJSON(message.positionsFilter);
        }
        if (message.oraclePriceFilter !== undefined) {
            obj.oraclePriceFilter = exports.OraclePriceFilter.toJSON(message.oraclePriceFilter);
        }
        return obj;
    },
    create(base) {
        return exports.StreamRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseStreamRequest();
        message.bankBalancesFilter = (object.bankBalancesFilter !== undefined && object.bankBalancesFilter !== null)
            ? exports.BankBalancesFilter.fromPartial(object.bankBalancesFilter)
            : undefined;
        message.subaccountDepositsFilter =
            (object.subaccountDepositsFilter !== undefined && object.subaccountDepositsFilter !== null)
                ? exports.SubaccountDepositsFilter.fromPartial(object.subaccountDepositsFilter)
                : undefined;
        message.spotTradesFilter = (object.spotTradesFilter !== undefined && object.spotTradesFilter !== null)
            ? exports.TradesFilter.fromPartial(object.spotTradesFilter)
            : undefined;
        message.derivativeTradesFilter =
            (object.derivativeTradesFilter !== undefined && object.derivativeTradesFilter !== null)
                ? exports.TradesFilter.fromPartial(object.derivativeTradesFilter)
                : undefined;
        message.spotOrdersFilter = (object.spotOrdersFilter !== undefined && object.spotOrdersFilter !== null)
            ? exports.OrdersFilter.fromPartial(object.spotOrdersFilter)
            : undefined;
        message.derivativeOrdersFilter =
            (object.derivativeOrdersFilter !== undefined && object.derivativeOrdersFilter !== null)
                ? exports.OrdersFilter.fromPartial(object.derivativeOrdersFilter)
                : undefined;
        message.spotOrderbooksFilter = (object.spotOrderbooksFilter !== undefined && object.spotOrderbooksFilter !== null)
            ? exports.OrderbookFilter.fromPartial(object.spotOrderbooksFilter)
            : undefined;
        message.derivativeOrderbooksFilter =
            (object.derivativeOrderbooksFilter !== undefined && object.derivativeOrderbooksFilter !== null)
                ? exports.OrderbookFilter.fromPartial(object.derivativeOrderbooksFilter)
                : undefined;
        message.positionsFilter = (object.positionsFilter !== undefined && object.positionsFilter !== null)
            ? exports.PositionsFilter.fromPartial(object.positionsFilter)
            : undefined;
        message.oraclePriceFilter = (object.oraclePriceFilter !== undefined && object.oraclePriceFilter !== null)
            ? exports.OraclePriceFilter.fromPartial(object.oraclePriceFilter)
            : undefined;
        return message;
    },
};
function createBaseStreamResponse() {
    return {
        blockHeight: "0",
        blockTime: "0",
        bankBalances: [],
        subaccountDeposits: [],
        spotTrades: [],
        derivativeTrades: [],
        spotOrders: [],
        derivativeOrders: [],
        spotOrderbookUpdates: [],
        derivativeOrderbookUpdates: [],
        positions: [],
        oraclePrices: [],
    };
}
exports.StreamResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.blockHeight !== "0") {
            writer.uint32(8).uint64(message.blockHeight);
        }
        if (message.blockTime !== "0") {
            writer.uint32(16).int64(message.blockTime);
        }
        for (const v of message.bankBalances) {
            exports.BankBalance.encode(v, writer.uint32(26).fork()).ldelim();
        }
        for (const v of message.subaccountDeposits) {
            exports.SubaccountDeposits.encode(v, writer.uint32(34).fork()).ldelim();
        }
        for (const v of message.spotTrades) {
            exports.SpotTrade.encode(v, writer.uint32(42).fork()).ldelim();
        }
        for (const v of message.derivativeTrades) {
            exports.DerivativeTrade.encode(v, writer.uint32(50).fork()).ldelim();
        }
        for (const v of message.spotOrders) {
            exports.SpotOrderUpdate.encode(v, writer.uint32(58).fork()).ldelim();
        }
        for (const v of message.derivativeOrders) {
            exports.DerivativeOrderUpdate.encode(v, writer.uint32(66).fork()).ldelim();
        }
        for (const v of message.spotOrderbookUpdates) {
            exports.OrderbookUpdate.encode(v, writer.uint32(74).fork()).ldelim();
        }
        for (const v of message.derivativeOrderbookUpdates) {
            exports.OrderbookUpdate.encode(v, writer.uint32(82).fork()).ldelim();
        }
        for (const v of message.positions) {
            exports.Position.encode(v, writer.uint32(90).fork()).ldelim();
        }
        for (const v of message.oraclePrices) {
            exports.OraclePrice.encode(v, writer.uint32(98).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStreamResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.blockHeight = longToString(reader.uint64());
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.blockTime = longToString(reader.int64());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.bankBalances.push(exports.BankBalance.decode(reader, reader.uint32()));
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.subaccountDeposits.push(exports.SubaccountDeposits.decode(reader, reader.uint32()));
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.spotTrades.push(exports.SpotTrade.decode(reader, reader.uint32()));
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.derivativeTrades.push(exports.DerivativeTrade.decode(reader, reader.uint32()));
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.spotOrders.push(exports.SpotOrderUpdate.decode(reader, reader.uint32()));
                    continue;
                case 8:
                    if (tag !== 66) {
                        break;
                    }
                    message.derivativeOrders.push(exports.DerivativeOrderUpdate.decode(reader, reader.uint32()));
                    continue;
                case 9:
                    if (tag !== 74) {
                        break;
                    }
                    message.spotOrderbookUpdates.push(exports.OrderbookUpdate.decode(reader, reader.uint32()));
                    continue;
                case 10:
                    if (tag !== 82) {
                        break;
                    }
                    message.derivativeOrderbookUpdates.push(exports.OrderbookUpdate.decode(reader, reader.uint32()));
                    continue;
                case 11:
                    if (tag !== 90) {
                        break;
                    }
                    message.positions.push(exports.Position.decode(reader, reader.uint32()));
                    continue;
                case 12:
                    if (tag !== 98) {
                        break;
                    }
                    message.oraclePrices.push(exports.OraclePrice.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            blockHeight: isSet(object.blockHeight) ? globalThis.String(object.blockHeight) : "0",
            blockTime: isSet(object.blockTime) ? globalThis.String(object.blockTime) : "0",
            bankBalances: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.bankBalances)
                ? object.bankBalances.map((e) => exports.BankBalance.fromJSON(e))
                : [],
            subaccountDeposits: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.subaccountDeposits)
                ? object.subaccountDeposits.map((e) => exports.SubaccountDeposits.fromJSON(e))
                : [],
            spotTrades: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.spotTrades)
                ? object.spotTrades.map((e) => exports.SpotTrade.fromJSON(e))
                : [],
            derivativeTrades: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.derivativeTrades)
                ? object.derivativeTrades.map((e) => exports.DerivativeTrade.fromJSON(e))
                : [],
            spotOrders: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.spotOrders)
                ? object.spotOrders.map((e) => exports.SpotOrderUpdate.fromJSON(e))
                : [],
            derivativeOrders: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.derivativeOrders)
                ? object.derivativeOrders.map((e) => exports.DerivativeOrderUpdate.fromJSON(e))
                : [],
            spotOrderbookUpdates: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.spotOrderbookUpdates)
                ? object.spotOrderbookUpdates.map((e) => exports.OrderbookUpdate.fromJSON(e))
                : [],
            derivativeOrderbookUpdates: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.derivativeOrderbookUpdates)
                ? object.derivativeOrderbookUpdates.map((e) => exports.OrderbookUpdate.fromJSON(e))
                : [],
            positions: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.positions)
                ? object.positions.map((e) => exports.Position.fromJSON(e))
                : [],
            oraclePrices: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.oraclePrices)
                ? object.oraclePrices.map((e) => exports.OraclePrice.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        const obj = {};
        if (message.blockHeight !== "0") {
            obj.blockHeight = message.blockHeight;
        }
        if (message.blockTime !== "0") {
            obj.blockTime = message.blockTime;
        }
        if ((_a = message.bankBalances) === null || _a === void 0 ? void 0 : _a.length) {
            obj.bankBalances = message.bankBalances.map((e) => exports.BankBalance.toJSON(e));
        }
        if ((_b = message.subaccountDeposits) === null || _b === void 0 ? void 0 : _b.length) {
            obj.subaccountDeposits = message.subaccountDeposits.map((e) => exports.SubaccountDeposits.toJSON(e));
        }
        if ((_c = message.spotTrades) === null || _c === void 0 ? void 0 : _c.length) {
            obj.spotTrades = message.spotTrades.map((e) => exports.SpotTrade.toJSON(e));
        }
        if ((_d = message.derivativeTrades) === null || _d === void 0 ? void 0 : _d.length) {
            obj.derivativeTrades = message.derivativeTrades.map((e) => exports.DerivativeTrade.toJSON(e));
        }
        if ((_e = message.spotOrders) === null || _e === void 0 ? void 0 : _e.length) {
            obj.spotOrders = message.spotOrders.map((e) => exports.SpotOrderUpdate.toJSON(e));
        }
        if ((_f = message.derivativeOrders) === null || _f === void 0 ? void 0 : _f.length) {
            obj.derivativeOrders = message.derivativeOrders.map((e) => exports.DerivativeOrderUpdate.toJSON(e));
        }
        if ((_g = message.spotOrderbookUpdates) === null || _g === void 0 ? void 0 : _g.length) {
            obj.spotOrderbookUpdates = message.spotOrderbookUpdates.map((e) => exports.OrderbookUpdate.toJSON(e));
        }
        if ((_h = message.derivativeOrderbookUpdates) === null || _h === void 0 ? void 0 : _h.length) {
            obj.derivativeOrderbookUpdates = message.derivativeOrderbookUpdates.map((e) => exports.OrderbookUpdate.toJSON(e));
        }
        if ((_j = message.positions) === null || _j === void 0 ? void 0 : _j.length) {
            obj.positions = message.positions.map((e) => exports.Position.toJSON(e));
        }
        if ((_k = message.oraclePrices) === null || _k === void 0 ? void 0 : _k.length) {
            obj.oraclePrices = message.oraclePrices.map((e) => exports.OraclePrice.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.StreamResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
        const message = createBaseStreamResponse();
        message.blockHeight = (_a = object.blockHeight) !== null && _a !== void 0 ? _a : "0";
        message.blockTime = (_b = object.blockTime) !== null && _b !== void 0 ? _b : "0";
        message.bankBalances = ((_c = object.bankBalances) === null || _c === void 0 ? void 0 : _c.map((e) => exports.BankBalance.fromPartial(e))) || [];
        message.subaccountDeposits = ((_d = object.subaccountDeposits) === null || _d === void 0 ? void 0 : _d.map((e) => exports.SubaccountDeposits.fromPartial(e))) || [];
        message.spotTrades = ((_e = object.spotTrades) === null || _e === void 0 ? void 0 : _e.map((e) => exports.SpotTrade.fromPartial(e))) || [];
        message.derivativeTrades = ((_f = object.derivativeTrades) === null || _f === void 0 ? void 0 : _f.map((e) => exports.DerivativeTrade.fromPartial(e))) || [];
        message.spotOrders = ((_g = object.spotOrders) === null || _g === void 0 ? void 0 : _g.map((e) => exports.SpotOrderUpdate.fromPartial(e))) || [];
        message.derivativeOrders = ((_h = object.derivativeOrders) === null || _h === void 0 ? void 0 : _h.map((e) => exports.DerivativeOrderUpdate.fromPartial(e))) || [];
        message.spotOrderbookUpdates = ((_j = object.spotOrderbookUpdates) === null || _j === void 0 ? void 0 : _j.map((e) => exports.OrderbookUpdate.fromPartial(e))) || [];
        message.derivativeOrderbookUpdates =
            ((_k = object.derivativeOrderbookUpdates) === null || _k === void 0 ? void 0 : _k.map((e) => exports.OrderbookUpdate.fromPartial(e))) || [];
        message.positions = ((_l = object.positions) === null || _l === void 0 ? void 0 : _l.map((e) => exports.Position.fromPartial(e))) || [];
        message.oraclePrices = ((_m = object.oraclePrices) === null || _m === void 0 ? void 0 : _m.map((e) => exports.OraclePrice.fromPartial(e))) || [];
        return message;
    },
};
function createBaseOrderbookUpdate() {
    return { seq: "0", orderbook: undefined };
}
exports.OrderbookUpdate = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.seq !== "0") {
            writer.uint32(8).uint64(message.seq);
        }
        if (message.orderbook !== undefined) {
            exports.Orderbook.encode(message.orderbook, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOrderbookUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.seq = longToString(reader.uint64());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.orderbook = exports.Orderbook.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            seq: isSet(object.seq) ? globalThis.String(object.seq) : "0",
            orderbook: isSet(object.orderbook) ? exports.Orderbook.fromJSON(object.orderbook) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.seq !== "0") {
            obj.seq = message.seq;
        }
        if (message.orderbook !== undefined) {
            obj.orderbook = exports.Orderbook.toJSON(message.orderbook);
        }
        return obj;
    },
    create(base) {
        return exports.OrderbookUpdate.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseOrderbookUpdate();
        message.seq = (_a = object.seq) !== null && _a !== void 0 ? _a : "0";
        message.orderbook = (object.orderbook !== undefined && object.orderbook !== null)
            ? exports.Orderbook.fromPartial(object.orderbook)
            : undefined;
        return message;
    },
};
function createBaseOrderbook() {
    return { marketId: "", buyLevels: [], sellLevels: [] };
}
exports.Orderbook = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.marketId !== "") {
            writer.uint32(10).string(message.marketId);
        }
        for (const v of message.buyLevels) {
            exchange_1.Level.encode(v, writer.uint32(18).fork()).ldelim();
        }
        for (const v of message.sellLevels) {
            exchange_1.Level.encode(v, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOrderbook();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.marketId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.buyLevels.push(exchange_1.Level.decode(reader, reader.uint32()));
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.sellLevels.push(exchange_1.Level.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            marketId: isSet(object.marketId) ? globalThis.String(object.marketId) : "",
            buyLevels: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.buyLevels) ? object.buyLevels.map((e) => exchange_1.Level.fromJSON(e)) : [],
            sellLevels: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.sellLevels)
                ? object.sellLevels.map((e) => exchange_1.Level.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a, _b;
        const obj = {};
        if (message.marketId !== "") {
            obj.marketId = message.marketId;
        }
        if ((_a = message.buyLevels) === null || _a === void 0 ? void 0 : _a.length) {
            obj.buyLevels = message.buyLevels.map((e) => exchange_1.Level.toJSON(e));
        }
        if ((_b = message.sellLevels) === null || _b === void 0 ? void 0 : _b.length) {
            obj.sellLevels = message.sellLevels.map((e) => exchange_1.Level.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.Orderbook.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseOrderbook();
        message.marketId = (_a = object.marketId) !== null && _a !== void 0 ? _a : "";
        message.buyLevels = ((_b = object.buyLevels) === null || _b === void 0 ? void 0 : _b.map((e) => exchange_1.Level.fromPartial(e))) || [];
        message.sellLevels = ((_c = object.sellLevels) === null || _c === void 0 ? void 0 : _c.map((e) => exchange_1.Level.fromPartial(e))) || [];
        return message;
    },
};
function createBaseBankBalance() {
    return { account: "", balances: [] };
}
exports.BankBalance = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.account !== "") {
            writer.uint32(10).string(message.account);
        }
        for (const v of message.balances) {
            coin_1.Coin.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseBankBalance();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.account = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.balances.push(coin_1.Coin.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            account: isSet(object.account) ? globalThis.String(object.account) : "",
            balances: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.balances) ? object.balances.map((e) => coin_1.Coin.fromJSON(e)) : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.account !== "") {
            obj.account = message.account;
        }
        if ((_a = message.balances) === null || _a === void 0 ? void 0 : _a.length) {
            obj.balances = message.balances.map((e) => coin_1.Coin.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.BankBalance.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseBankBalance();
        message.account = (_a = object.account) !== null && _a !== void 0 ? _a : "";
        message.balances = ((_b = object.balances) === null || _b === void 0 ? void 0 : _b.map((e) => coin_1.Coin.fromPartial(e))) || [];
        return message;
    },
};
function createBaseSubaccountDeposits() {
    return { subaccountId: "", deposits: [] };
}
exports.SubaccountDeposits = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.subaccountId !== "") {
            writer.uint32(10).string(message.subaccountId);
        }
        for (const v of message.deposits) {
            exports.SubaccountDeposit.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSubaccountDeposits();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.subaccountId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.deposits.push(exports.SubaccountDeposit.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            subaccountId: isSet(object.subaccountId) ? globalThis.String(object.subaccountId) : "",
            deposits: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.deposits)
                ? object.deposits.map((e) => exports.SubaccountDeposit.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.subaccountId !== "") {
            obj.subaccountId = message.subaccountId;
        }
        if ((_a = message.deposits) === null || _a === void 0 ? void 0 : _a.length) {
            obj.deposits = message.deposits.map((e) => exports.SubaccountDeposit.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.SubaccountDeposits.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseSubaccountDeposits();
        message.subaccountId = (_a = object.subaccountId) !== null && _a !== void 0 ? _a : "";
        message.deposits = ((_b = object.deposits) === null || _b === void 0 ? void 0 : _b.map((e) => exports.SubaccountDeposit.fromPartial(e))) || [];
        return message;
    },
};
function createBaseSubaccountDeposit() {
    return { denom: "", deposit: undefined };
}
exports.SubaccountDeposit = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.denom !== "") {
            writer.uint32(10).string(message.denom);
        }
        if (message.deposit !== undefined) {
            exchange_1.Deposit.encode(message.deposit, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSubaccountDeposit();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.denom = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.deposit = exchange_1.Deposit.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            denom: isSet(object.denom) ? globalThis.String(object.denom) : "",
            deposit: isSet(object.deposit) ? exchange_1.Deposit.fromJSON(object.deposit) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.denom !== "") {
            obj.denom = message.denom;
        }
        if (message.deposit !== undefined) {
            obj.deposit = exchange_1.Deposit.toJSON(message.deposit);
        }
        return obj;
    },
    create(base) {
        return exports.SubaccountDeposit.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseSubaccountDeposit();
        message.denom = (_a = object.denom) !== null && _a !== void 0 ? _a : "";
        message.deposit = (object.deposit !== undefined && object.deposit !== null)
            ? exchange_1.Deposit.fromPartial(object.deposit)
            : undefined;
        return message;
    },
};
function createBaseSpotOrderUpdate() {
    return { status: 0, orderHash: new Uint8Array(0), cid: "", order: undefined };
}
exports.SpotOrderUpdate = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.status !== 0) {
            writer.uint32(8).int32(message.status);
        }
        if (message.orderHash.length !== 0) {
            writer.uint32(18).bytes(message.orderHash);
        }
        if (message.cid !== "") {
            writer.uint32(26).string(message.cid);
        }
        if (message.order !== undefined) {
            exports.SpotOrder.encode(message.order, writer.uint32(34).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSpotOrderUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.status = reader.int32();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.orderHash = reader.bytes();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.cid = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.order = exports.SpotOrder.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            status: isSet(object.status) ? orderUpdateStatusFromJSON(object.status) : 0,
            orderHash: isSet(object.orderHash) ? bytesFromBase64(object.orderHash) : new Uint8Array(0),
            cid: isSet(object.cid) ? globalThis.String(object.cid) : "",
            order: isSet(object.order) ? exports.SpotOrder.fromJSON(object.order) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.status !== 0) {
            obj.status = orderUpdateStatusToJSON(message.status);
        }
        if (message.orderHash.length !== 0) {
            obj.orderHash = base64FromBytes(message.orderHash);
        }
        if (message.cid !== "") {
            obj.cid = message.cid;
        }
        if (message.order !== undefined) {
            obj.order = exports.SpotOrder.toJSON(message.order);
        }
        return obj;
    },
    create(base) {
        return exports.SpotOrderUpdate.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseSpotOrderUpdate();
        message.status = (_a = object.status) !== null && _a !== void 0 ? _a : 0;
        message.orderHash = (_b = object.orderHash) !== null && _b !== void 0 ? _b : new Uint8Array(0);
        message.cid = (_c = object.cid) !== null && _c !== void 0 ? _c : "";
        message.order = (object.order !== undefined && object.order !== null)
            ? exports.SpotOrder.fromPartial(object.order)
            : undefined;
        return message;
    },
};
function createBaseSpotOrder() {
    return { marketId: "", order: undefined };
}
exports.SpotOrder = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.marketId !== "") {
            writer.uint32(10).string(message.marketId);
        }
        if (message.order !== undefined) {
            exchange_1.SpotLimitOrder.encode(message.order, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSpotOrder();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.marketId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.order = exchange_1.SpotLimitOrder.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            marketId: isSet(object.marketId) ? globalThis.String(object.marketId) : "",
            order: isSet(object.order) ? exchange_1.SpotLimitOrder.fromJSON(object.order) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.marketId !== "") {
            obj.marketId = message.marketId;
        }
        if (message.order !== undefined) {
            obj.order = exchange_1.SpotLimitOrder.toJSON(message.order);
        }
        return obj;
    },
    create(base) {
        return exports.SpotOrder.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseSpotOrder();
        message.marketId = (_a = object.marketId) !== null && _a !== void 0 ? _a : "";
        message.order = (object.order !== undefined && object.order !== null)
            ? exchange_1.SpotLimitOrder.fromPartial(object.order)
            : undefined;
        return message;
    },
};
function createBaseDerivativeOrderUpdate() {
    return { status: 0, orderHash: new Uint8Array(0), cid: "", order: undefined };
}
exports.DerivativeOrderUpdate = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.status !== 0) {
            writer.uint32(8).int32(message.status);
        }
        if (message.orderHash.length !== 0) {
            writer.uint32(18).bytes(message.orderHash);
        }
        if (message.cid !== "") {
            writer.uint32(26).string(message.cid);
        }
        if (message.order !== undefined) {
            exports.DerivativeOrder.encode(message.order, writer.uint32(34).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDerivativeOrderUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.status = reader.int32();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.orderHash = reader.bytes();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.cid = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.order = exports.DerivativeOrder.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            status: isSet(object.status) ? orderUpdateStatusFromJSON(object.status) : 0,
            orderHash: isSet(object.orderHash) ? bytesFromBase64(object.orderHash) : new Uint8Array(0),
            cid: isSet(object.cid) ? globalThis.String(object.cid) : "",
            order: isSet(object.order) ? exports.DerivativeOrder.fromJSON(object.order) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.status !== 0) {
            obj.status = orderUpdateStatusToJSON(message.status);
        }
        if (message.orderHash.length !== 0) {
            obj.orderHash = base64FromBytes(message.orderHash);
        }
        if (message.cid !== "") {
            obj.cid = message.cid;
        }
        if (message.order !== undefined) {
            obj.order = exports.DerivativeOrder.toJSON(message.order);
        }
        return obj;
    },
    create(base) {
        return exports.DerivativeOrderUpdate.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseDerivativeOrderUpdate();
        message.status = (_a = object.status) !== null && _a !== void 0 ? _a : 0;
        message.orderHash = (_b = object.orderHash) !== null && _b !== void 0 ? _b : new Uint8Array(0);
        message.cid = (_c = object.cid) !== null && _c !== void 0 ? _c : "";
        message.order = (object.order !== undefined && object.order !== null)
            ? exports.DerivativeOrder.fromPartial(object.order)
            : undefined;
        return message;
    },
};
function createBaseDerivativeOrder() {
    return { marketId: "", order: undefined, isMarket: false };
}
exports.DerivativeOrder = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.marketId !== "") {
            writer.uint32(10).string(message.marketId);
        }
        if (message.order !== undefined) {
            exchange_1.DerivativeLimitOrder.encode(message.order, writer.uint32(18).fork()).ldelim();
        }
        if (message.isMarket === true) {
            writer.uint32(24).bool(message.isMarket);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDerivativeOrder();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.marketId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.order = exchange_1.DerivativeLimitOrder.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.isMarket = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            marketId: isSet(object.marketId) ? globalThis.String(object.marketId) : "",
            order: isSet(object.order) ? exchange_1.DerivativeLimitOrder.fromJSON(object.order) : undefined,
            isMarket: isSet(object.isMarket) ? globalThis.Boolean(object.isMarket) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.marketId !== "") {
            obj.marketId = message.marketId;
        }
        if (message.order !== undefined) {
            obj.order = exchange_1.DerivativeLimitOrder.toJSON(message.order);
        }
        if (message.isMarket === true) {
            obj.isMarket = message.isMarket;
        }
        return obj;
    },
    create(base) {
        return exports.DerivativeOrder.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseDerivativeOrder();
        message.marketId = (_a = object.marketId) !== null && _a !== void 0 ? _a : "";
        message.order = (object.order !== undefined && object.order !== null)
            ? exchange_1.DerivativeLimitOrder.fromPartial(object.order)
            : undefined;
        message.isMarket = (_b = object.isMarket) !== null && _b !== void 0 ? _b : false;
        return message;
    },
};
function createBasePosition() {
    return {
        marketId: "",
        subaccountId: "",
        isLong: false,
        quantity: "",
        entryPrice: "",
        margin: "",
        cumulativeFundingEntry: "",
    };
}
exports.Position = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.marketId !== "") {
            writer.uint32(10).string(message.marketId);
        }
        if (message.subaccountId !== "") {
            writer.uint32(18).string(message.subaccountId);
        }
        if (message.isLong === true) {
            writer.uint32(24).bool(message.isLong);
        }
        if (message.quantity !== "") {
            writer.uint32(34).string(message.quantity);
        }
        if (message.entryPrice !== "") {
            writer.uint32(42).string(message.entryPrice);
        }
        if (message.margin !== "") {
            writer.uint32(50).string(message.margin);
        }
        if (message.cumulativeFundingEntry !== "") {
            writer.uint32(58).string(message.cumulativeFundingEntry);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePosition();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.marketId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.subaccountId = reader.string();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.isLong = reader.bool();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.quantity = reader.string();
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.entryPrice = reader.string();
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.margin = reader.string();
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.cumulativeFundingEntry = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            marketId: isSet(object.marketId) ? globalThis.String(object.marketId) : "",
            subaccountId: isSet(object.subaccountId) ? globalThis.String(object.subaccountId) : "",
            isLong: isSet(object.isLong) ? globalThis.Boolean(object.isLong) : false,
            quantity: isSet(object.quantity) ? globalThis.String(object.quantity) : "",
            entryPrice: isSet(object.entryPrice) ? globalThis.String(object.entryPrice) : "",
            margin: isSet(object.margin) ? globalThis.String(object.margin) : "",
            cumulativeFundingEntry: isSet(object.cumulativeFundingEntry)
                ? globalThis.String(object.cumulativeFundingEntry)
                : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.marketId !== "") {
            obj.marketId = message.marketId;
        }
        if (message.subaccountId !== "") {
            obj.subaccountId = message.subaccountId;
        }
        if (message.isLong === true) {
            obj.isLong = message.isLong;
        }
        if (message.quantity !== "") {
            obj.quantity = message.quantity;
        }
        if (message.entryPrice !== "") {
            obj.entryPrice = message.entryPrice;
        }
        if (message.margin !== "") {
            obj.margin = message.margin;
        }
        if (message.cumulativeFundingEntry !== "") {
            obj.cumulativeFundingEntry = message.cumulativeFundingEntry;
        }
        return obj;
    },
    create(base) {
        return exports.Position.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g;
        const message = createBasePosition();
        message.marketId = (_a = object.marketId) !== null && _a !== void 0 ? _a : "";
        message.subaccountId = (_b = object.subaccountId) !== null && _b !== void 0 ? _b : "";
        message.isLong = (_c = object.isLong) !== null && _c !== void 0 ? _c : false;
        message.quantity = (_d = object.quantity) !== null && _d !== void 0 ? _d : "";
        message.entryPrice = (_e = object.entryPrice) !== null && _e !== void 0 ? _e : "";
        message.margin = (_f = object.margin) !== null && _f !== void 0 ? _f : "";
        message.cumulativeFundingEntry = (_g = object.cumulativeFundingEntry) !== null && _g !== void 0 ? _g : "";
        return message;
    },
};
function createBaseOraclePrice() {
    return { symbol: "", price: "", type: "" };
}
exports.OraclePrice = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.symbol !== "") {
            writer.uint32(10).string(message.symbol);
        }
        if (message.price !== "") {
            writer.uint32(18).string(message.price);
        }
        if (message.type !== "") {
            writer.uint32(26).string(message.type);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOraclePrice();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.symbol = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.price = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.type = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            symbol: isSet(object.symbol) ? globalThis.String(object.symbol) : "",
            price: isSet(object.price) ? globalThis.String(object.price) : "",
            type: isSet(object.type) ? globalThis.String(object.type) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.symbol !== "") {
            obj.symbol = message.symbol;
        }
        if (message.price !== "") {
            obj.price = message.price;
        }
        if (message.type !== "") {
            obj.type = message.type;
        }
        return obj;
    },
    create(base) {
        return exports.OraclePrice.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseOraclePrice();
        message.symbol = (_a = object.symbol) !== null && _a !== void 0 ? _a : "";
        message.price = (_b = object.price) !== null && _b !== void 0 ? _b : "";
        message.type = (_c = object.type) !== null && _c !== void 0 ? _c : "";
        return message;
    },
};
function createBaseSpotTrade() {
    return {
        marketId: "",
        isBuy: false,
        executionType: "",
        quantity: "",
        price: "",
        subaccountId: "",
        fee: "",
        orderHash: new Uint8Array(0),
        feeRecipientAddress: "",
        cid: "",
        tradeId: "",
    };
}
exports.SpotTrade = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.marketId !== "") {
            writer.uint32(10).string(message.marketId);
        }
        if (message.isBuy === true) {
            writer.uint32(16).bool(message.isBuy);
        }
        if (message.executionType !== "") {
            writer.uint32(26).string(message.executionType);
        }
        if (message.quantity !== "") {
            writer.uint32(34).string(message.quantity);
        }
        if (message.price !== "") {
            writer.uint32(42).string(message.price);
        }
        if (message.subaccountId !== "") {
            writer.uint32(50).string(message.subaccountId);
        }
        if (message.fee !== "") {
            writer.uint32(58).string(message.fee);
        }
        if (message.orderHash.length !== 0) {
            writer.uint32(66).bytes(message.orderHash);
        }
        if (message.feeRecipientAddress !== "") {
            writer.uint32(74).string(message.feeRecipientAddress);
        }
        if (message.cid !== "") {
            writer.uint32(82).string(message.cid);
        }
        if (message.tradeId !== "") {
            writer.uint32(90).string(message.tradeId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSpotTrade();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.marketId = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.isBuy = reader.bool();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.executionType = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.quantity = reader.string();
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.price = reader.string();
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.subaccountId = reader.string();
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.fee = reader.string();
                    continue;
                case 8:
                    if (tag !== 66) {
                        break;
                    }
                    message.orderHash = reader.bytes();
                    continue;
                case 9:
                    if (tag !== 74) {
                        break;
                    }
                    message.feeRecipientAddress = reader.string();
                    continue;
                case 10:
                    if (tag !== 82) {
                        break;
                    }
                    message.cid = reader.string();
                    continue;
                case 11:
                    if (tag !== 90) {
                        break;
                    }
                    message.tradeId = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            marketId: isSet(object.marketId) ? globalThis.String(object.marketId) : "",
            isBuy: isSet(object.isBuy) ? globalThis.Boolean(object.isBuy) : false,
            executionType: isSet(object.executionType) ? globalThis.String(object.executionType) : "",
            quantity: isSet(object.quantity) ? globalThis.String(object.quantity) : "",
            price: isSet(object.price) ? globalThis.String(object.price) : "",
            subaccountId: isSet(object.subaccountId) ? globalThis.String(object.subaccountId) : "",
            fee: isSet(object.fee) ? globalThis.String(object.fee) : "",
            orderHash: isSet(object.orderHash) ? bytesFromBase64(object.orderHash) : new Uint8Array(0),
            feeRecipientAddress: isSet(object.feeRecipientAddress) ? globalThis.String(object.feeRecipientAddress) : "",
            cid: isSet(object.cid) ? globalThis.String(object.cid) : "",
            tradeId: isSet(object.tradeId) ? globalThis.String(object.tradeId) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.marketId !== "") {
            obj.marketId = message.marketId;
        }
        if (message.isBuy === true) {
            obj.isBuy = message.isBuy;
        }
        if (message.executionType !== "") {
            obj.executionType = message.executionType;
        }
        if (message.quantity !== "") {
            obj.quantity = message.quantity;
        }
        if (message.price !== "") {
            obj.price = message.price;
        }
        if (message.subaccountId !== "") {
            obj.subaccountId = message.subaccountId;
        }
        if (message.fee !== "") {
            obj.fee = message.fee;
        }
        if (message.orderHash.length !== 0) {
            obj.orderHash = base64FromBytes(message.orderHash);
        }
        if (message.feeRecipientAddress !== "") {
            obj.feeRecipientAddress = message.feeRecipientAddress;
        }
        if (message.cid !== "") {
            obj.cid = message.cid;
        }
        if (message.tradeId !== "") {
            obj.tradeId = message.tradeId;
        }
        return obj;
    },
    create(base) {
        return exports.SpotTrade.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
        const message = createBaseSpotTrade();
        message.marketId = (_a = object.marketId) !== null && _a !== void 0 ? _a : "";
        message.isBuy = (_b = object.isBuy) !== null && _b !== void 0 ? _b : false;
        message.executionType = (_c = object.executionType) !== null && _c !== void 0 ? _c : "";
        message.quantity = (_d = object.quantity) !== null && _d !== void 0 ? _d : "";
        message.price = (_e = object.price) !== null && _e !== void 0 ? _e : "";
        message.subaccountId = (_f = object.subaccountId) !== null && _f !== void 0 ? _f : "";
        message.fee = (_g = object.fee) !== null && _g !== void 0 ? _g : "";
        message.orderHash = (_h = object.orderHash) !== null && _h !== void 0 ? _h : new Uint8Array(0);
        message.feeRecipientAddress = (_j = object.feeRecipientAddress) !== null && _j !== void 0 ? _j : "";
        message.cid = (_k = object.cid) !== null && _k !== void 0 ? _k : "";
        message.tradeId = (_l = object.tradeId) !== null && _l !== void 0 ? _l : "";
        return message;
    },
};
function createBaseDerivativeTrade() {
    return {
        marketId: "",
        isBuy: false,
        executionType: "",
        subaccountId: "",
        positionDelta: undefined,
        payout: "",
        fee: "",
        orderHash: "",
        feeRecipientAddress: "",
        cid: "",
        tradeId: "",
    };
}
exports.DerivativeTrade = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.marketId !== "") {
            writer.uint32(10).string(message.marketId);
        }
        if (message.isBuy === true) {
            writer.uint32(16).bool(message.isBuy);
        }
        if (message.executionType !== "") {
            writer.uint32(26).string(message.executionType);
        }
        if (message.subaccountId !== "") {
            writer.uint32(34).string(message.subaccountId);
        }
        if (message.positionDelta !== undefined) {
            exchange_1.PositionDelta.encode(message.positionDelta, writer.uint32(42).fork()).ldelim();
        }
        if (message.payout !== "") {
            writer.uint32(50).string(message.payout);
        }
        if (message.fee !== "") {
            writer.uint32(58).string(message.fee);
        }
        if (message.orderHash !== "") {
            writer.uint32(66).string(message.orderHash);
        }
        if (message.feeRecipientAddress !== "") {
            writer.uint32(74).string(message.feeRecipientAddress);
        }
        if (message.cid !== "") {
            writer.uint32(82).string(message.cid);
        }
        if (message.tradeId !== "") {
            writer.uint32(90).string(message.tradeId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDerivativeTrade();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.marketId = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.isBuy = reader.bool();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.executionType = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.subaccountId = reader.string();
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.positionDelta = exchange_1.PositionDelta.decode(reader, reader.uint32());
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.payout = reader.string();
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.fee = reader.string();
                    continue;
                case 8:
                    if (tag !== 66) {
                        break;
                    }
                    message.orderHash = reader.string();
                    continue;
                case 9:
                    if (tag !== 74) {
                        break;
                    }
                    message.feeRecipientAddress = reader.string();
                    continue;
                case 10:
                    if (tag !== 82) {
                        break;
                    }
                    message.cid = reader.string();
                    continue;
                case 11:
                    if (tag !== 90) {
                        break;
                    }
                    message.tradeId = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            marketId: isSet(object.marketId) ? globalThis.String(object.marketId) : "",
            isBuy: isSet(object.isBuy) ? globalThis.Boolean(object.isBuy) : false,
            executionType: isSet(object.executionType) ? globalThis.String(object.executionType) : "",
            subaccountId: isSet(object.subaccountId) ? globalThis.String(object.subaccountId) : "",
            positionDelta: isSet(object.positionDelta) ? exchange_1.PositionDelta.fromJSON(object.positionDelta) : undefined,
            payout: isSet(object.payout) ? globalThis.String(object.payout) : "",
            fee: isSet(object.fee) ? globalThis.String(object.fee) : "",
            orderHash: isSet(object.orderHash) ? globalThis.String(object.orderHash) : "",
            feeRecipientAddress: isSet(object.feeRecipientAddress) ? globalThis.String(object.feeRecipientAddress) : "",
            cid: isSet(object.cid) ? globalThis.String(object.cid) : "",
            tradeId: isSet(object.tradeId) ? globalThis.String(object.tradeId) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.marketId !== "") {
            obj.marketId = message.marketId;
        }
        if (message.isBuy === true) {
            obj.isBuy = message.isBuy;
        }
        if (message.executionType !== "") {
            obj.executionType = message.executionType;
        }
        if (message.subaccountId !== "") {
            obj.subaccountId = message.subaccountId;
        }
        if (message.positionDelta !== undefined) {
            obj.positionDelta = exchange_1.PositionDelta.toJSON(message.positionDelta);
        }
        if (message.payout !== "") {
            obj.payout = message.payout;
        }
        if (message.fee !== "") {
            obj.fee = message.fee;
        }
        if (message.orderHash !== "") {
            obj.orderHash = message.orderHash;
        }
        if (message.feeRecipientAddress !== "") {
            obj.feeRecipientAddress = message.feeRecipientAddress;
        }
        if (message.cid !== "") {
            obj.cid = message.cid;
        }
        if (message.tradeId !== "") {
            obj.tradeId = message.tradeId;
        }
        return obj;
    },
    create(base) {
        return exports.DerivativeTrade.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        const message = createBaseDerivativeTrade();
        message.marketId = (_a = object.marketId) !== null && _a !== void 0 ? _a : "";
        message.isBuy = (_b = object.isBuy) !== null && _b !== void 0 ? _b : false;
        message.executionType = (_c = object.executionType) !== null && _c !== void 0 ? _c : "";
        message.subaccountId = (_d = object.subaccountId) !== null && _d !== void 0 ? _d : "";
        message.positionDelta = (object.positionDelta !== undefined && object.positionDelta !== null)
            ? exchange_1.PositionDelta.fromPartial(object.positionDelta)
            : undefined;
        message.payout = (_e = object.payout) !== null && _e !== void 0 ? _e : "";
        message.fee = (_f = object.fee) !== null && _f !== void 0 ? _f : "";
        message.orderHash = (_g = object.orderHash) !== null && _g !== void 0 ? _g : "";
        message.feeRecipientAddress = (_h = object.feeRecipientAddress) !== null && _h !== void 0 ? _h : "";
        message.cid = (_j = object.cid) !== null && _j !== void 0 ? _j : "";
        message.tradeId = (_k = object.tradeId) !== null && _k !== void 0 ? _k : "";
        return message;
    },
};
function createBaseTradesFilter() {
    return { subaccountIds: [], marketIds: [] };
}
exports.TradesFilter = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.subaccountIds) {
            writer.uint32(10).string(v);
        }
        for (const v of message.marketIds) {
            writer.uint32(18).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTradesFilter();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.subaccountIds.push(reader.string());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.marketIds.push(reader.string());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            subaccountIds: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.subaccountIds)
                ? object.subaccountIds.map((e) => globalThis.String(e))
                : [],
            marketIds: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.marketIds)
                ? object.marketIds.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        var _a, _b;
        const obj = {};
        if ((_a = message.subaccountIds) === null || _a === void 0 ? void 0 : _a.length) {
            obj.subaccountIds = message.subaccountIds;
        }
        if ((_b = message.marketIds) === null || _b === void 0 ? void 0 : _b.length) {
            obj.marketIds = message.marketIds;
        }
        return obj;
    },
    create(base) {
        return exports.TradesFilter.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseTradesFilter();
        message.subaccountIds = ((_a = object.subaccountIds) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        message.marketIds = ((_b = object.marketIds) === null || _b === void 0 ? void 0 : _b.map((e) => e)) || [];
        return message;
    },
};
function createBasePositionsFilter() {
    return { subaccountIds: [], marketIds: [] };
}
exports.PositionsFilter = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.subaccountIds) {
            writer.uint32(10).string(v);
        }
        for (const v of message.marketIds) {
            writer.uint32(18).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePositionsFilter();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.subaccountIds.push(reader.string());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.marketIds.push(reader.string());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            subaccountIds: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.subaccountIds)
                ? object.subaccountIds.map((e) => globalThis.String(e))
                : [],
            marketIds: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.marketIds)
                ? object.marketIds.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        var _a, _b;
        const obj = {};
        if ((_a = message.subaccountIds) === null || _a === void 0 ? void 0 : _a.length) {
            obj.subaccountIds = message.subaccountIds;
        }
        if ((_b = message.marketIds) === null || _b === void 0 ? void 0 : _b.length) {
            obj.marketIds = message.marketIds;
        }
        return obj;
    },
    create(base) {
        return exports.PositionsFilter.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBasePositionsFilter();
        message.subaccountIds = ((_a = object.subaccountIds) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        message.marketIds = ((_b = object.marketIds) === null || _b === void 0 ? void 0 : _b.map((e) => e)) || [];
        return message;
    },
};
function createBaseOrdersFilter() {
    return { subaccountIds: [], marketIds: [] };
}
exports.OrdersFilter = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.subaccountIds) {
            writer.uint32(10).string(v);
        }
        for (const v of message.marketIds) {
            writer.uint32(18).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOrdersFilter();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.subaccountIds.push(reader.string());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.marketIds.push(reader.string());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            subaccountIds: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.subaccountIds)
                ? object.subaccountIds.map((e) => globalThis.String(e))
                : [],
            marketIds: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.marketIds)
                ? object.marketIds.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        var _a, _b;
        const obj = {};
        if ((_a = message.subaccountIds) === null || _a === void 0 ? void 0 : _a.length) {
            obj.subaccountIds = message.subaccountIds;
        }
        if ((_b = message.marketIds) === null || _b === void 0 ? void 0 : _b.length) {
            obj.marketIds = message.marketIds;
        }
        return obj;
    },
    create(base) {
        return exports.OrdersFilter.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseOrdersFilter();
        message.subaccountIds = ((_a = object.subaccountIds) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        message.marketIds = ((_b = object.marketIds) === null || _b === void 0 ? void 0 : _b.map((e) => e)) || [];
        return message;
    },
};
function createBaseOrderbookFilter() {
    return { marketIds: [] };
}
exports.OrderbookFilter = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.marketIds) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOrderbookFilter();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.marketIds.push(reader.string());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            marketIds: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.marketIds)
                ? object.marketIds.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.marketIds) === null || _a === void 0 ? void 0 : _a.length) {
            obj.marketIds = message.marketIds;
        }
        return obj;
    },
    create(base) {
        return exports.OrderbookFilter.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseOrderbookFilter();
        message.marketIds = ((_a = object.marketIds) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        return message;
    },
};
function createBaseBankBalancesFilter() {
    return { accounts: [] };
}
exports.BankBalancesFilter = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.accounts) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseBankBalancesFilter();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.accounts.push(reader.string());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            accounts: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.accounts) ? object.accounts.map((e) => globalThis.String(e)) : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.accounts) === null || _a === void 0 ? void 0 : _a.length) {
            obj.accounts = message.accounts;
        }
        return obj;
    },
    create(base) {
        return exports.BankBalancesFilter.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseBankBalancesFilter();
        message.accounts = ((_a = object.accounts) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        return message;
    },
};
function createBaseSubaccountDepositsFilter() {
    return { subaccountIds: [] };
}
exports.SubaccountDepositsFilter = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.subaccountIds) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSubaccountDepositsFilter();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.subaccountIds.push(reader.string());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            subaccountIds: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.subaccountIds)
                ? object.subaccountIds.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.subaccountIds) === null || _a === void 0 ? void 0 : _a.length) {
            obj.subaccountIds = message.subaccountIds;
        }
        return obj;
    },
    create(base) {
        return exports.SubaccountDepositsFilter.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseSubaccountDepositsFilter();
        message.subaccountIds = ((_a = object.subaccountIds) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        return message;
    },
};
function createBaseOraclePriceFilter() {
    return { symbol: [] };
}
exports.OraclePriceFilter = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.symbol) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOraclePriceFilter();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.symbol.push(reader.string());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            symbol: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.symbol) ? object.symbol.map((e) => globalThis.String(e)) : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.symbol) === null || _a === void 0 ? void 0 : _a.length) {
            obj.symbol = message.symbol;
        }
        return obj;
    },
    create(base) {
        return exports.OraclePriceFilter.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseOraclePriceFilter();
        message.symbol = ((_a = object.symbol) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        return message;
    },
};
class StreamClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.Stream = this.Stream.bind(this);
    }
    Stream(request, metadata) {
        return this.rpc.invoke(exports.StreamStreamDesc, exports.StreamRequest.fromPartial(request), metadata);
    }
}
exports.StreamClientImpl = StreamClientImpl;
exports.StreamDesc = { serviceName: "injective.stream.v1beta1.Stream" };
exports.StreamStreamDesc = {
    methodName: "Stream",
    service: exports.StreamDesc,
    requestStream: false,
    responseStream: true,
    requestType: {
        serializeBinary() {
            return exports.StreamRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = exports.StreamResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
class GrpcWebImpl {
    constructor(host, options) {
        this.host = host;
        this.options = options;
    }
    unary(methodDesc, _request, metadata) {
        var _a;
        const request = Object.assign(Object.assign({}, _request), methodDesc.requestType);
        const maybeCombinedMetadata = metadata && this.options.metadata
            ? new browser_headers_1.BrowserHeaders(Object.assign(Object.assign({}, (_a = this.options) === null || _a === void 0 ? void 0 : _a.metadata.headersMap), metadata === null || metadata === void 0 ? void 0 : metadata.headersMap))
            : metadata !== null && metadata !== void 0 ? metadata : this.options.metadata;
        return new Promise((resolve, reject) => {
            var _a;
            grpc_web_1.grpc.unary(methodDesc, Object.assign(Object.assign({ request, host: this.host, metadata: maybeCombinedMetadata !== null && maybeCombinedMetadata !== void 0 ? maybeCombinedMetadata : {} }, (this.options.transport !== undefined ? { transport: this.options.transport } : {})), { debug: (_a = this.options.debug) !== null && _a !== void 0 ? _a : false, onEnd: function (response) {
                    if (response.status === grpc_web_1.grpc.Code.OK) {
                        resolve(response.message.toObject());
                    }
                    else {
                        const err = new GrpcWebError(response.statusMessage, response.status, response.trailers);
                        reject(err);
                    }
                } }));
        });
    }
    invoke(methodDesc, _request, metadata) {
        var _a, _b, _c;
        const upStreamCodes = (_a = this.options.upStreamRetryCodes) !== null && _a !== void 0 ? _a : [];
        const DEFAULT_TIMEOUT_TIME = 3000;
        const request = Object.assign(Object.assign({}, _request), methodDesc.requestType);
        const transport = (_b = this.options.streamingTransport) !== null && _b !== void 0 ? _b : this.options.transport;
        const maybeCombinedMetadata = metadata && this.options.metadata
            ? new browser_headers_1.BrowserHeaders(Object.assign(Object.assign({}, (_c = this.options) === null || _c === void 0 ? void 0 : _c.metadata.headersMap), metadata === null || metadata === void 0 ? void 0 : metadata.headersMap))
            : metadata !== null && metadata !== void 0 ? metadata : this.options.metadata;
        return new rxjs_1.Observable((observer) => {
            const upStream = () => {
                var _a;
                const client = grpc_web_1.grpc.invoke(methodDesc, Object.assign(Object.assign({ host: this.host, request }, (transport !== undefined ? { transport } : {})), { metadata: maybeCombinedMetadata !== null && maybeCombinedMetadata !== void 0 ? maybeCombinedMetadata : {}, debug: (_a = this.options.debug) !== null && _a !== void 0 ? _a : false, onMessage: (next) => observer.next(next), onEnd: (code, message, trailers) => {
                        if (code === 0) {
                            observer.complete();
                        }
                        else if (upStreamCodes.includes(code)) {
                            setTimeout(upStream, DEFAULT_TIMEOUT_TIME);
                        }
                        else {
                            const err = new Error(message);
                            err.code = code;
                            err.metadata = trailers;
                            observer.error(err);
                        }
                    } }));
                observer.add(() => client.close());
            };
            upStream();
        }).pipe((0, operators_1.share)());
    }
}
exports.GrpcWebImpl = GrpcWebImpl;
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function longToString(long) {
    return long.toString();
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
class GrpcWebError extends globalThis.Error {
    constructor(message, code, metadata) {
        super(message);
        this.code = code;
        this.metadata = metadata;
    }
}
exports.GrpcWebError = GrpcWebError;
