// @wormhole-foundation/sdk-definitions/dist/cjs/protocols/tokenBridge/tokenBridgeLayout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tokenBridgeNamedPayloads = exports.transferWithPayloadLayout = void 0;
const sdk_base_1 = require("@wormhole-foundation/sdk-base");
const index_js_1 = require("./../../layout-items/index.js");
const index_js_2 = require("./../../vaa/index.js");
const fixedLengthStringItem = {
    binary: "bytes",
    size: 32,
    custom: {
        to: (val) => (0, sdk_base_1.range)(val.byteLength)
            .map((i) => String.fromCharCode(val[i]))
            .join(""),
        from: (str) => new Uint8Array(str.split("").map((c) => c.charCodeAt(0))),
    },
};
const transferCommonLayout = [
    {
        name: "token",
        binary: "bytes",
        layout: [
            { name: "amount", ...index_js_1.amountItem },
            { name: "address", ...index_js_1.universalAddressItem },
            { name: "chain", ...(0, index_js_1.chainItem)() },
        ],
    },
    {
        name: "to",
        binary: "bytes",
        layout: [
            { name: "address", ...index_js_1.universalAddressItem },
            { name: "chain", ...(0, index_js_1.chainItem)() },
        ],
    },
];
const transferWithPayloadLayout = (customPayload) => [
    (0, index_js_1.payloadIdItem)(3),
    ...transferCommonLayout,
    { name: "from", ...index_js_1.universalAddressItem },
    (0, sdk_base_1.customizableBytes)({ name: "payload" }, customPayload),
];
exports.transferWithPayloadLayout = transferWithPayloadLayout;
exports.tokenBridgeNamedPayloads = [
    [
        "AttestMeta",
        [
            (0, index_js_1.payloadIdItem)(2),
            {
                name: "token",
                binary: "bytes",
                layout: [
                    { name: "address", ...index_js_1.universalAddressItem },
                    { name: "chain", ...(0, index_js_1.chainItem)() },
                ],
            },
            { name: "decimals", binary: "uint", size: 1 },
            { name: "symbol", ...fixedLengthStringItem },
            { name: "name", ...fixedLengthStringItem },
        ],
    ],
    ["Transfer", [(0, index_js_1.payloadIdItem)(1), ...transferCommonLayout, { name: "fee", ...index_js_1.amountItem }]],
    ["TransferWithPayload", (0, exports.transferWithPayloadLayout)()],
];
(0, index_js_2.registerPayloadTypes)("TokenBridge", exports.tokenBridgeNamedPayloads);
