// @wormhole-foundation/sdk-definitions/dist/cjs/contracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContracts = void 0;
const sdk_base_1 = require("@wormhole-foundation/sdk-base");
/**
 *
 * Get the contracts for a given network and chain
 *
 * @param n the network to get contracts for
 * @param c the chain to get contracts for
 * @returns the contracts for the given network and chain
 */
function getContracts(n, c) {
    const ct = {
        coreBridge: sdk_base_1.contracts.coreBridge.get(n, c),
        tokenBridge: sdk_base_1.contracts.tokenBridge.get(n, c),
        nftBridge: sdk_base_1.contracts.nftBridge.get(n, c),
        relayer: sdk_base_1.contracts.relayer.get(n, c),
        tokenBridgeRelayer: sdk_base_1.contracts.tokenBridgeRelayer.get(n, c),
    };
    if (sdk_base_1.contracts.circleContracts.has(n, c)) {
        ct.cctp = sdk_base_1.contracts.circleContracts.get(n, c);
    }
    if (sdk_base_1.contracts.gateway.has(n, c)) {
        ct.gateway = sdk_base_1.contracts.gateway.get(n, c);
    }
    if (sdk_base_1.contracts.translator.has(n, c)) {
        ct.translator = sdk_base_1.contracts.translator.get(n, c);
    }
    if (sdk_base_1.contracts.portico.has(n, c)) {
        ct.portico = sdk_base_1.contracts.portico.get(n, c);
    }
    return ct;
}
exports.getContracts = getContracts;
