// @wormhole-foundation/sdk-definitions/dist/cjs/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ed25519 = exports.secp256k1 = exports.sha512_256 = exports.sha256 = exports.sha3_256 = exports.keccak256 = void 0;
const sha256_1 = require("@noble/hashes/sha256");
Object.defineProperty(exports, "sha256", { enumerable: true, get: function () { return sha256_1.sha256; } });
const sha3_1 = require("@noble/hashes/sha3");
Object.defineProperty(exports, "keccak256", { enumerable: true, get: function () { return sha3_1.keccak_256; } });
Object.defineProperty(exports, "sha3_256", { enumerable: true, get: function () { return sha3_1.sha3_256; } });
const sha512_1 = require("@noble/hashes/sha512");
Object.defineProperty(exports, "sha512_256", { enumerable: true, get: function () { return sha512_1.sha512_256; } });
const secp256k1_1 = require("@noble/curves/secp256k1");
Object.defineProperty(exports, "secp256k1", { enumerable: true, get: function () { return secp256k1_1.secp256k1; } });
const ed25519_1 = require("@noble/curves/ed25519");
Object.defineProperty(exports, "ed25519", { enumerable: true, get: function () { return ed25519_1.ed25519; } });
